{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d2e2b9e3-17a8-4cdb-beb2-d04b5e664682",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "STUDY = 'ESCAPE'\n",
    "REDMINE_STUDY = 'ESCAPE'\n",
    "\n",
    "#SPECIAL_SUBJECT_ID = 'patid'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "19b945af-1835-4d01-aec3-bf6fdb999884",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "import redminelib\n",
    "import requests\n",
    "\n",
    "from urllib3.exceptions import InsecureRequestWarning\n",
    "requests.packages.urllib3.disable_warnings(category=InsecureRequestWarning)\n",
    "\n",
    "from sqlalchemy import or_\n",
    "from sqlalchemy import create_engine, MetaData, Table\n",
    "from sqlalchemy.orm import sessionmaker, aliased\n",
    "from sqlalchemy import text\n",
    "\n",
    "DBURL = 'oracle://sapphire_r:sapphire_r0@oar1.bwh.harvard.edu:1521/labvprd'\n",
    "\n",
    "sapphire8_md = MetaData(schema='SAPPHIRE8')\n",
    "engine = create_engine(DBURL)\n",
    "Session = sessionmaker(bind=engine)\n",
    "session = Session()\n",
    "\n",
    "Alias = Table('SDIALIAS',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Batch = Table('S_BATCH', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "Cohort = Table('U_COHORT',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Collection = Table('U_COLLECTION',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Package = Table('S_PACKAGE',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Sample = Table('S_SAMPLE',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "SampleFamily = Table('S_SAMPLEFAMILY',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "SampleMap = Table('S_SAMPLEMAP', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "SpecimenType = Table('U_SPECIMENTYPE', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "Study = Table('S_STUDY',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Subject = Table('S_SUBJECT',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Tissue = Table('S_TISSUE', sapphire8_md, autoload=True, autoload_with=engine)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a1e5a6-7db2-45a0-9b6e-dc88a8d1cd60",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "534c0d28-c12e-49a1-be43-8844e1d6eb8b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_studyid</th>\n",
       "      <th>studydesc</th>\n",
       "      <th>studystatus</th>\n",
       "      <th>studytype</th>\n",
       "      <th>activeflag</th>\n",
       "      <th>defaultglpflag</th>\n",
       "      <th>studyalias</th>\n",
       "      <th>u_studydesign</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>ESCAPE clinical trial for lung cancer screenin...</td>\n",
       "      <td>Active</td>\n",
       "      <td>Human Genetics</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  s_studyid                                          studydesc studystatus  \\\n",
       "0    ESCAPE  ESCAPE clinical trial for lung cancer screenin...      Active   \n",
       "\n",
       "        studytype activeflag defaultglpflag studyalias u_studydesign  \n",
       "0  Human Genetics          Y              Y     ESCAPE          None  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_study = session.query(Study).filter(Study.c.s_studyid == STUDY)\n",
    "study = pd.read_sql(q_study.statement, session.bind)\n",
    "study = study.drop(columns=['estimatedsamples', 'usersequence', 'notes','collectionstartdt','collectionenddt','auditsequence',\n",
    "                            'primaryaffection', 'tracelogid', 'createdt', 'createby','createtool', 'moddt', 'completeddt','cancelledby','cancelleddt',\n",
    "                            'u_verifiedby','u_numberofsites','u_legacyid','securityuser','securitydepartment',\n",
    "                            'modby','modtool','projectid','proposedstartdt','proposedenddt','startdt','clinicalflag','plannedsites','plannedparticipants','completedby',\n",
    "                            'enddt','studyfocus','templateflag','subjectrequiredflag','priority','collectinforequiredflag','protocolname','hipaaflag',\n",
    "                            'departmentid','verifiedby','conservativecocflag','conservativerestrictionsflag',\n",
    "                            'verifiedbyrole'])\n",
    "study"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53dfd030-3e1b-4bf9-853c-627f346db2ce",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Tissue\n",
    "\n",
    "This is not a primary table, but can be merged to see additional details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8fcd94c4-f68a-4e5a-b518-6f1905769ccd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_tissueid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TS-00000001</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TS-00000002</td>\n",
       "      <td>CD4 + TCell - Unstim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TS-00000003</td>\n",
       "      <td>CD4 + TCell - Stim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TS-00000004</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TS-00000009</td>\n",
       "      <td>Dorsal Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>TS-00000218</td>\n",
       "      <td>Colo205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TS-00000219</td>\n",
       "      <td>H1299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TS-00000229</td>\n",
       "      <td>U937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TS-00000231</td>\n",
       "      <td>Adherent Cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TS-00000203</td>\n",
       "      <td>Beas-2B Cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     s_tissueid            tissuedesc\n",
       "0   TS-00000001                 Blood\n",
       "1   TS-00000002  CD4 + TCell - Unstim\n",
       "2   TS-00000003    CD4 + TCell - Stim\n",
       "3   TS-00000004                  Lung\n",
       "4   TS-00000009           Dorsal Lung\n",
       "..          ...                   ...\n",
       "58  TS-00000218               Colo205\n",
       "59  TS-00000219                 H1299\n",
       "60  TS-00000229                  U937\n",
       "61  TS-00000231        Adherent Cells\n",
       "62  TS-00000203         Beas-2B Cells\n",
       "\n",
       "[63 rows x 2 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_tissue = session.query(Tissue.c.s_tissueid, Tissue.c.tissuedesc)\n",
    "tissue = pd.read_sql(q_tissue.statement, session.bind)\n",
    "tissue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80415544-773c-4e09-a24c-edeb10b568b0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d3135912-4bd6-448f-94e5-cd8fb9075698",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>u_cohortid</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>studyid</th>\n",
       "      <th>cohorttype</th>\n",
       "      <th>afflicted</th>\n",
       "      <th>s_subjectid</th>\n",
       "      <th>birthdt</th>\n",
       "      <th>genderflag</th>\n",
       "      <th>u_dummyflag</th>\n",
       "      <th>u_motherid</th>\n",
       "      <th>u_fatherid</th>\n",
       "      <th>u_pedigreeid</th>\n",
       "      <th>u_race</th>\n",
       "      <th>u_ethnicity</th>\n",
       "      <th>u_legacyid</th>\n",
       "      <th>u_fail</th>\n",
       "      <th>u_deceasedflag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHO-00136809</td>\n",
       "      <td>ST-00044831</td>\n",
       "      <td>CRIMSN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00044831</td>\n",
       "      <td>1965-07-25</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00031004</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>141902</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHO-00136810</td>\n",
       "      <td>ST-00044832</td>\n",
       "      <td>CRIMSN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00044832</td>\n",
       "      <td>1965-06-14</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00031098</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>142002</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHO-00136811</td>\n",
       "      <td>ST-00044833</td>\n",
       "      <td>CRIMSN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00044833</td>\n",
       "      <td>1965-03-16</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00031415</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>142798</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHO-00136812</td>\n",
       "      <td>ST-00044834</td>\n",
       "      <td>CRIMSN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00044834</td>\n",
       "      <td>1965-12-07</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00032077</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>143987</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHO-00136813</td>\n",
       "      <td>ST-00044835</td>\n",
       "      <td>CRIMSN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00044835</td>\n",
       "      <td>1983-06-17</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00032100</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>144010</td>\n",
       "      <td>Y</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116532</th>\n",
       "      <td>CHO-00099105</td>\n",
       "      <td>ST-00007273</td>\n",
       "      <td>CCRN</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Affected</td>\n",
       "      <td>ST-00007273</td>\n",
       "      <td>NaT</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00043715</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>164132</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116533</th>\n",
       "      <td>CHO-00099112</td>\n",
       "      <td>ST-00007280</td>\n",
       "      <td>PILOTS</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Affected</td>\n",
       "      <td>ST-00007280</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00043791</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>164333</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116534</th>\n",
       "      <td>CHO-00099119</td>\n",
       "      <td>ST-00007287</td>\n",
       "      <td>PILOTS</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Affected</td>\n",
       "      <td>ST-00007287</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00043798</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>164340</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116535</th>\n",
       "      <td>CHO-00100084</td>\n",
       "      <td>ST-00008252</td>\n",
       "      <td>GECOPD</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00008252</td>\n",
       "      <td>NaT</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00046468</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>None</td>\n",
       "      <td>167286</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116536</th>\n",
       "      <td>CHO-00096924</td>\n",
       "      <td>ST-00005092</td>\n",
       "      <td>NWCOPD</td>\n",
       "      <td>Proband</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>ST-00005092</td>\n",
       "      <td>1952-02-07</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-00039122</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>157701</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116537 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          u_cohortid    subjectid studyid cohorttype     afflicted  \\\n",
       "0       CHO-00136809  ST-00044831  CRIMSN    Proband  Undetermined   \n",
       "1       CHO-00136810  ST-00044832  CRIMSN    Proband  Undetermined   \n",
       "2       CHO-00136811  ST-00044833  CRIMSN    Proband  Undetermined   \n",
       "3       CHO-00136812  ST-00044834  CRIMSN    Proband  Undetermined   \n",
       "4       CHO-00136813  ST-00044835  CRIMSN    Proband  Undetermined   \n",
       "...              ...          ...     ...        ...           ...   \n",
       "116532  CHO-00099105  ST-00007273    CCRN    Proband      Affected   \n",
       "116533  CHO-00099112  ST-00007280  PILOTS    Proband      Affected   \n",
       "116534  CHO-00099119  ST-00007287  PILOTS    Proband      Affected   \n",
       "116535  CHO-00100084  ST-00008252  GECOPD    Proband  Undetermined   \n",
       "116536  CHO-00096924  ST-00005092  NWCOPD    Proband  Undetermined   \n",
       "\n",
       "        s_subjectid    birthdt genderflag u_dummyflag u_motherid u_fatherid  \\\n",
       "0       ST-00044831 1965-07-25          F           N       None       None   \n",
       "1       ST-00044832 1965-06-14          F           N       None       None   \n",
       "2       ST-00044833 1965-03-16          F           N       None       None   \n",
       "3       ST-00044834 1965-12-07          F           N       None       None   \n",
       "4       ST-00044835 1983-06-17          M           N       None       None   \n",
       "...             ...        ...        ...         ...        ...        ...   \n",
       "116532  ST-00007273        NaT          F           N       None       None   \n",
       "116533  ST-00007280        NaT          U           N       None       None   \n",
       "116534  ST-00007287        NaT          U           N       None       None   \n",
       "116535  ST-00008252        NaT          M           N       None       None   \n",
       "116536  ST-00005092 1952-02-07          M           N       None       None   \n",
       "\n",
       "       u_pedigreeid     u_race u_ethnicity u_legacyid u_fail u_deceasedflag  \n",
       "0       PG-00031004  Caucasian        None     141902   None           None  \n",
       "1       PG-00031098  Caucasian        None     142002   None           None  \n",
       "2       PG-00031415  Caucasian        None     142798   None           None  \n",
       "3       PG-00032077  Caucasian        None     143987   None           None  \n",
       "4       PG-00032100  Caucasian        None     144010      Y           None  \n",
       "...             ...        ...         ...        ...    ...            ...  \n",
       "116532  PG-00043715  Caucasian        None     164132   None           None  \n",
       "116533  PG-00043791       None        None     164333   None           None  \n",
       "116534  PG-00043798       None        None     164340   None           None  \n",
       "116535  PG-00046468  Caucasian        None     167286   None           None  \n",
       "116536  PG-00039122       None        None     157701   None           None  \n",
       "\n",
       "[116537 rows x 17 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_cohort = session.query(Cohort, Subject).filter(Subject.c.s_subjectid==Cohort.c.subjectid)\n",
    "cohort = pd.read_sql(q_cohort.statement, session.bind)\n",
    "cohort.drop(columns=['usersequence', 'auditsequence', 'tracelogid', 'templateflag', 'notes', 'createtool', 'modtool',\n",
    "                    'moddt', 'cohortdesc', 'createdt', 'createby', 'modby', 'activeflag', 'subjecttype', 'subjectdesc',\n",
    "                    'speciesid', 'strainid', 'securityuser', 'securitydepartment', 'activeflag_1', 'auditsequence_1',\n",
    "                    'createby_1', 'createdt_1', 'createtool_1', 'modby_1', 'moddt_1', 'modtool_1', 'templateflag_1',\n",
    "                     'tracelogid_1', 'usersequence_1', 'notes_1', 'u_cageid', 'u_country', 'affliction', 'severity',\n",
    "                    ], inplace=True)\n",
    "\n",
    "cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f3203d66-f5e3-43c9-87a7-cfa039f79a48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cohorttype\n",
       "Proband        56592\n",
       "NaN            30560\n",
       "Non-Proband    29385\n",
       "dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('cohorttype', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e89e9b3-1ec3-444c-ba32-d60949ef69e8",
   "metadata": {},
   "source": [
    "### Distribution of Participant Date-of-Birth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d9204431-fc36-4d33-ae5c-1e5c44f13f02",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjYAAAGxCAYAAABx6/zIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAA9hAAAPYQGoP6dpAABFsElEQVR4nO3dfVxUZf7/8ffIPQijoDCQinfoeluprXclljd5g9baZoVLuplZpq6p22blirU/NCutzTS70zLNrW/qmrWkbkr6FW+yKE2zMsubQFzFARVB5fr90dezDqAyiKKH1/PxmMfunPM5Z64z19C8vc51zjiMMUYAAAA2UK2yGwAAAFBRCDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDZAGc2bN08Oh8N6+Pr6qk6dOvrjH/+o/fv3V+hrpaSkaOnSpSWWr1mzRg6HQ2vWrPFqf2fa/tNPP1VI+y7W9u3blZycXOb2FH/vAwMD5XK5dPPNN2vKlCnKzs6+bG25VA4fPqy7775bkZGRcjgcuv3220utK/5eOBwO1a5dW127dtXy5ctL1DscDiUnJ5epDevXr1dycrKOHDlSYl39+vWVkJDgxREBlYNgA3hp7ty5Sk9P18qVKzVs2DC9++67uummm3Ts2LEKe41zBZs2bdooPT1dbdq08Wp/ffv2VXp6uqKjoyuohRdn+/btmjx5stdh4uz3/uWXX9Z1112nZ555Rs2aNdOqVasua1sq2tNPP60lS5ZoxowZSk9P17Rp085bf+a9WL9+vV599VX5+PioX79++vDDDz3q0tPTdf/995epDevXr9fkyZNLDTbA1cK3shsAXG1atmypdu3aSZJuvvlmnT59Wk8//bSWLl2qQYMGXdS+8/PzFRQUdM71YWFh6tChg9f7rV27tmrXrn0xTbsinP3eS9Idd9yhRx55RDfeeKMGDBig77//XlFRUZXYwvLbtm2bGjVqVObPUPH3olevXqpZs6beffdd9evXz1pels9Lfn6+AgMDvW80cAVixAa4SGe+OH7++WdJ0uTJk9W+fXuFh4crLCxMbdq00RtvvKHivzd7Zmh/8eLFuv766xUYGKjJkyfL4XDo2LFjeuutt6xTDV27dpV07lNRGzduVL9+/RQREaHAwEA1atRIY8aMsdaXdiqqa9euatmypdauXasOHTooKChI11xzjSZOnKjTp0977N/bY0pNTVWbNm0UFBSk3/zmN3rzzTc92nLnnXdK+jUYnjnGefPmefvWS5Lq1aun559/Xnl5eZozZ461/PPPP9fdd9+t+vXrKygoSPXr19c999xj9VNZ27Jq1Sp169ZNYWFhCg4OVufOnfXvf/+7zO07fPiwRowYoWuuuUb+/v5q2LChnnjiCRUUFEiSfvrpJzkcDq1atUo7duyw2uDt6cbAwED5+/vLz8/PY3nxU1FnPgsrVqzQfffdp9q1ays4OFgTJkzQn//8Z0lSgwYNztmO8/UtcCVgxAa4SD/88IMkWSMiP/30k4YPH6569epJkjZs2KBRo0Zp//79+utf/+qx7RdffKEdO3boySefVIMGDRQSEqLbb79dt9xyi26++WZNnDhR0q8jNefyySefqF+/fmrWrJmmT5+uevXq6aefftKKFSsu2PasrCzdfffdeuyxx/TUU0/po48+0t/+9jfl5ORo5syZVp03x/TVV19p3LhxeuyxxxQVFaXXX39dQ4cOVePGjdWlSxf17dtXKSkpevzxx/Xyyy9bp9UaNWp0wfaeS58+feTj46PPPvvMo81NmzbV3XffrfDwcGVmZmr27Nm64YYbtH37dtWqVeuCbXnnnXd077336rbbbtNbb70lPz8/zZkzR7feeqs++eQTdevW7bztOnHihG6++Wbt2rVLkydPVuvWrbV27VpNmTJFGRkZ+uijjxQdHa309HSNGDFCbrdbCxYskCQ1b978vPs+ffq0Tp06JWOMDhw4oGeffVbHjh1TYmJimd6z++67T3379tX8+fN17NgxtWvXTsePH9dLL72kxYsXW6ctz27HhfoWuCIYAGUyd+5cI8ls2LDBnDx50uTl5Znly5eb2rVrm9DQUJOVlVVim9OnT5uTJ0+ap556ykRERJiioiJrXWxsrPHx8TE7d+4ssV1ISIgZPHhwieWrV682kszq1autZY0aNTKNGjUy+fn5F2z77t27rWXx8fFGkvnnP//pUTts2DBTrVo18/PPP5e6rwsdU2BgoMe2+fn5Jjw83AwfPtxa9v7775c4jvM50/7NmzefsyYqKso0a9bsnOtPnTpljh49akJCQsyLL754wbYcO3bMhIeHm379+nksP336tLn22mvNb3/72wu2+5VXXjGSzHvvveex/JlnnjGSzIoVK6xl8fHxpkWLFhfc55n3ovgjICDAzJo1q0S9JDNp0qQS2997770lap999tkSn5Mzytq3QGXjVBTgpQ4dOsjPz0+hoaFKSEiQy+XSv/71L2tux6effqru3bvL6XTKx8dHfn5++utf/6pDhw6VuHqndevWatKkSbnb8t1332nXrl0aOnRoueZIhIaGqn///h7LEhMTVVRU5DH64c0xXXfdddbIjvTrKZImTZp4nAK6FEyx02JHjx7VX/7yFzVu3Fi+vr7y9fVV9erVdezYMe3YseOC+1u/fr0OHz6swYMH69SpU9ajqKhIvXr10ubNm60J42evPzOKIv36voWEhOj3v/+9x76HDBkiSec9pVVUVOSxz+KnB99++21t3rxZmzdv1r/+9S8NHjxYDz/8sMdI2/nccccdZao7W2X1LeANTkUBXnr77bfVrFkz+fr6KioqyuNKo02bNqlnz57q2rWrXnvtNdWpU0f+/v5aunSp/t//+3/Kz8/32NfFXqV08OBBSVKdOnXKtX1pE21dLpck6dChQ5K8P6aIiIgS+wwICChRV5GOHTumQ4cOqVWrVtayxMRE/fvf/9bEiRN1ww03KCwsTA6HQ3369ClTWw4cOCBJJULJ2Q4fPqyDBw+qQYMGHstXr16trl276tChQ3K5XHI4HB7rIyMj5evra73Hpbnvvvv01ltvWc/j4+M95rs0a9asxOThn3/+WY8++qj+8Ic/qEaNGuc9vvJ89iqjbwFvEWwALxX/QjnbokWL5Ofnp+XLl3uMoJR26bakEl943jozr2ffvn3l2v7Ml/fZsrKyJP33S8zbY6oMH330kU6fPm1Nsna73Vq+fLkmTZqkxx57zKorKCjQ4cOHy7TPWrVqSZJeeumlc15ZdCYYbt682WN506ZNJf36Hm7cuFHGGI++zs7O1qlTp6zXKE1ycrJGjhxpPQ8NDb1gm1u3bq1PPvlE3333nX7729+et/ZiP3vAlYpgA1SgMzfu8/HxsZbl5+dr/vz5Xu2nrP8KbtKkiRo1aqQ333xTY8eOVUBAgFevk5eXp2XLlnmcjlq4cKGqVatmTQatqGM625l2VsS/9Pfs2aPx48fL6XRq+PDhkn5tszGmxPvx+uuvlzilc662dO7cWTVq1ND27ds9AkZpzhV0u3Xrpvfee09Lly7V7373O2v522+/ba0/l/r166t+/frnfd3iMjIyJKncl/ZXZL8AlYVgA1Sgvn37avr06UpMTNQDDzygQ4cO6bnnnvM6cLRq1Upr1qzRhx9+qOjoaIWGhlqjAMW9/PLL6tevnzp06KBHHnlE9erV0549e/TJJ59YV9icS0REhB566CHt2bNHTZo00ccff6zXXntNDz30kDWXoqKO6WwtW7aUJL366qsKDQ1VYGCgGjRoUOqpjrNt27bNmnOSnZ2ttWvXau7cufLx8dGSJUusL/SwsDB16dJFzz77rGrVqqX69esrLS1Nb7zxRolTNOdry0svvaTBgwfr8OHD+v3vf6/IyEgdPHhQX331lQ4ePKjZs2eft7333nuvXn75ZQ0ePFg//fSTWrVqpXXr1iklJUV9+vRR9+7dy/kO/ve9kH49bbh48WKtXLlSv/vd70qcGiurM6fyXnzxRQ0ePFh+fn5q2rRpmUaLgCtG5c5dBq4eZbkyxxhj3nzzTdO0aVMTEBBgGjZsaKZMmWLeeOONElebxMbGmr59+5a6j4yMDNO5c2cTHBxsJJn4+HhjTOlXRRljTHp6uundu7dxOp0mICDANGrUyDzyyCMl2l78qqgWLVqYNWvWmHbt2pmAgAATHR1tHn/8cXPy5MkKPab4+HjrGM544YUXTIMGDYyPj4+RZObOnXvO97T4lUD+/v4mMjLSxMfHm5SUFJOdnV1im3379pk77rjD1KxZ04SGhppevXqZbdu2mdjY2BJXnJ2vLWlpaaZv374mPDzc+Pn5mWuuucb07dvXvP/+++ds79kOHTpkHnzwQRMdHW18fX1NbGysmTBhgjlx4kSJ96i8V0U5nU5z3XXXmenTp5fYr85xVdS5PscTJkwwMTExplq1ah6fNW/6FqhMDmOKXUoAoEro2rWr/vOf/2jbtm2V3RQAqDBc7g0AAGyDYAMAAGyDU1EAAMA2vB6x2b9/v/7whz8oIiJCwcHBuu6667RlyxZrvTFGycnJiomJUVBQkLp27apvvvnGYx8FBQUaNWqUatWqpZCQEPXv37/EfThycnKUlJQkp9Mpp9OppKQkHTlypHxHCQAAqgSvgk1OTo46d+4sPz8//etf/9L27dv1/PPPe1w+OW3aNE2fPl0zZ87U5s2b5XK51KNHD+Xl5Vk1Y8aM0ZIlS7Ro0SKtW7dOR48eVUJCgsf9JRITE5WRkaHU1FSlpqYqIyNDSUlJF3/EAADAtrw6FfXYY4/pf//3f7V27dpS1xtjFBMTozFjxugvf/mLpF9HZ6KiovTMM89o+PDhcrvdql27tubPn6+77rpLkvTLL7+obt26+vjjj3Xrrbdqx44dat68uTZs2KD27dtL+vXXhDt27Khvv/32nPfzAAAAVZtXN+hbtmyZbr31Vt15551KS0vTNddcoxEjRmjYsGGSpN27dysrK0s9e/a0tgkICFB8fLzWr1+v4cOHa8uWLTp58qRHTUxMjFq2bKn169fr1ltvVXp6upxOpxVqpF9/eNDpdGr9+vWlBpuCggIVFBRYz4uKinT48GFFRERw63AAAK4Sxhjl5eUpJiZG1ap5f42TV8Hmxx9/1OzZszV27Fg9/vjj2rRpk0aPHq2AgADde++91m/MFP9hvaioKOvXX7OysuTv76+aNWuWqDmzfVZWliIjI0u8fmRkpFVT3JQpUzR58mRvDgcAAFyh9u7dW64f+PUq2BQVFaldu3ZKSUmRJF1//fX65ptvNHv2bN17771WXfERElPsB+BKU7ymtPrz7WfChAkaO3as9dztdqtevXrau3evwsLCLnxwAACg0uXm5qpu3brl/ikPr4JNdHS0mjdv7rGsWbNm+uCDDyRJLpdL0q8jLtHR0VZNdna2NYrjcrlUWFionJwcj1Gb7OxsderUyaop7VeHDx48WGI06IyAgIBSf7smLCyMYAMAwFWmvNNIvDp51blzZ+3cudNj2XfffafY2FhJUoMGDeRyubRy5UprfWFhodLS0qzQ0rZtW/n5+XnUZGZmatu2bVZNx44d5Xa7tWnTJqtm48aNcrvdVg0AAEBxXo3YPPLII+rUqZNSUlI0cOBAbdq0Sa+++qpeffVVSb+mqzFjxiglJUVxcXGKi4tTSkqKgoODlZiYKElyOp0aOnSoxo0bp4iICIWHh2v8+PFq1aqV9Uu3zZo1U69evTRs2DDNmTNHkvTAAw8oISGBK6IAAMC5efurmR9++KFp2bKlCQgIML/5zW/Mq6++6rG+qKjITJo0ybhcLhMQEGC6dOlitm7d6lGTn59vRo4cacLDw01QUJBJSEgwe/bs8ag5dOiQGTRokAkNDTWhoaFm0KBBJicnp8ztdLvdRpJxu93eHiIAAKgkF/v9bdufVMjNzZXT6ZTb7WaODQBUIGOMTp065XFTVaCsfHx85Ovre845NBf7/e3VqSgAQNVWWFiozMxMHT9+vLKbgqtYcHCwoqOj5e/vX+H7JtgAAMqkqKhIu3fvlo+Pj2JiYuTv788NUOEVY4wKCwt18OBB7d69W3FxceW6Cd/5EGwAAGVSWFiooqIi1a1bV8HBwZXdHFylgoKC5Ofnp59//lmFhYUKDAys0P1XbEwCANheRf8LG1XPpfwM8ekEAAC2QbABAAC2wRwbAMDFW3iZJxEnVv6dSpKTk7V06VJlZGRc9tdes2aNbr75ZuXk5KhGjRqX/fWvZIzYAABsb8iQIXI4HHI4HPLz81PDhg01fvx4HTt2rLKbhgrGiA0AoEro1auX5s6dq5MnT2rt2rW6//77dezYMc2ePduj7uTJk/Lz86ukVuJiMWIDAKgSAgIC5HK5VLduXSUmJmrQoEFaunSpkpOTdd111+nNN99Uw4YNFRAQIGOM9uzZo9tuu03Vq1dXWFiYBg4cqAMHDpTY75w5c6xL4O+8804dOXLEWrd582b16NFDtWrVktPpVHx8vL744guP7R0Oh15//XX97ne/U3BwsOLi4rRs2TKPmo8//lhNmjRRUFCQbr75Zv30008e63/++Wf169dPNWvWVEhIiFq0aKGPP/64wt67qwnBBjiXhY7/PgDYTlBQkE6ePClJ+uGHH/Tee+/pgw8+sObM3H777Tp8+LDS0tK0cuVK7dq1S3fddZfHPs5s9+GHHyo1NVUZGRl6+OGHrfV5eXkaPHiw1q5dqw0bNiguLk59+vRRXl6ex34mT56sgQMH6uuvv1afPn00aNAgHT58WJK0d+9eDRgwQH369FFGRobuv/9+PfbYYx7bP/zwwyooKNBnn32mrVu36plnnlH16tUr+i27KnAqCgBQ5WzatEkLFy5Ut27dJP1688H58+erdu3akqSVK1fq66+/1u7du1W3bl1J0vz589WiRQtt3rxZN9xwgyTpxIkTeuutt1SnTh1J0ksvvaS+ffvq+eefl8vl0i233OLxunPmzFHNmjWVlpamhIQEa/mQIUN0zz33SJJSUlL00ksvadOmTerVq5dmz56thg0basaMGXI4HGratKkVXs7Ys2eP7rjjDrVq1UqS1LBhw0vxtl0VGLEBAFQJy5cvV/Xq1RUYGKiOHTuqS5cueumllyRJsbGxVqiRpB07dqhu3bpWqJGk5s2bq0aNGtqxY4e1rF69elaokaSOHTuqqKhIO3fulCRlZ2frwQcfVJMmTeR0OuV0OnX06FHt2bPHo22tW7e2/n9ISIhCQ0OVnZ1ttaVDhw4eP1/RsWNHj+1Hjx6tv/3tb+rcubMmTZqkr7/+utzv09WOYAMAqBJuvvlmZWRkaOfOnTpx4oQWL16syMhISb+GibMZY0r9HaxzLT/jzLoz/ztkyBBt2bJFL7zwgtavX6+MjAxFRESosLDQY7vik5UdDoeKioqs17yQ+++/Xz/++KOSkpK0detWtWvXzgptVQ3BBgBQJYSEhKhx48aKjY294FVPzZs31549e7R3715r2fbt2+V2u9WsWTNr2Z49e/TLL79Yz9PT01WtWjU1adJEkrR27VqNHj1affr0UYsWLRQQEKD//Oc/XrW7efPm2rBhg8ey4s8lqW7dunrwwQe1ePFijRs3Tq+99ppXr2MXBBsAAIrp3r27WrdurUGDBumLL77Qpk2bdO+99yo+Pl7t2rWz6gIDAzV48GB99dVXVogZOHCgXC6XJKlx48aaP3++duzYoY0bN2rQoEEKCgryqi0PPvigdu3apbFjx2rnzp1auHCh5s2b51EzZswYffLJJ9q9e7e++OILffrppx4BrCph8jAA4OJdAXcCrkgOh0NLly7VqFGj1KVLF1WrVk29evUqcXqncePG1hVLhw8fVp8+fTRr1ixr/ZtvvqkHHnhA119/verVq6eUlBSNHz/eq7bUq1dPH3zwgR555BHNmjVLv/3tb5WSkqL77rvPqjl9+rQefvhh7du3T2FhYerVq5dmzJhxcW/CVcphynLy7iqUm5srp9Mpt9utsLCwym4OrkZnX+Zts/9oA+Vx4sQJ7d69Ww0aNFBgYGBlNwdXsfN9li72+5tTUQAAwDYINgAAwDYINgAAwDYINgAAwDYINgAAwDYINgAAwDYINgAAwDYINgAAwDYINgAAwDb4SQUAwEVbvDPzsr7egKbRl/X1rlQOh0NLlizR7bffftlfu2vXrrruuuv0wgsvXPbXPh9GbAAAtjdkyBA5HA5NnTrVY/nSpUvlcDjOsVXF+Omnn+RwOKxHzZo11aVLF6WlpV3S162qCDYAgCohMDBQzzzzjHJycirl9VetWqXMzEylpaUpLCxMffr00e7du0utPXny5GVunX0QbAAAVUL37t3lcrk0ZcqUc9Z88MEHatGihQICAlS/fn09//zzHuvr169v/bJ2aGio6tWrp1dffbVMrx8RESGXy6XWrVtrzpw5On78uFasWCHp11NKr7zyim677TaFhITob3/7myRp9uzZatSokfz9/dW0aVPNnz+/xH4zMzPVu3dvBQUFqUGDBnr//fc91v/lL39RkyZNFBwcrIYNG2rixIkewSk5OVnXXXed5s+fr/r168vpdOruu+9WXl6eVXPs2DHde++9ql69uqKjo0u8L5I0a9YsxcXFKTAwUFFRUfr9739fpvelohFsAABVgo+Pj1JSUvTSSy9p3759JdZv2bJFAwcO1N13362tW7cqOTlZEydO1Lx58zzqnn/+ebVr105ffvmlRowYoYceekjffvutV20JDg6W5DkyM2nSJN12223aunWr7rvvPi1ZskR/+tOfNG7cOG3btk3Dhw/XH//4R61evdpjXxMnTtQdd9yhr776Sn/4wx90zz33aMeOHdb60NBQzZs3T9u3b9eLL76o1157TTNmzPDYx65du7R06VItX75cy5cvV1pamsdpuz//+c9avXq1lixZohUrVmjNmjXasmWLtf7zzz/X6NGj9dRTT2nnzp1KTU1Vly5dvHpPKgrBBgBQZfzud7/Tddddp0mTJpVYN336dHXr1k0TJ05UkyZNNGTIEI0cOVLPPvusR12fPn00YsQINW7cWH/5y19Uq1YtrVmzpsxtOHbsmCZMmCAfHx/Fx8dbyxMTE3XfffepYcOGio2N1XPPPachQ4ZoxIgRatKkicaOHasBAwboueee89jfnXfeqfvvv19NmjTR008/rXbt2umll16y1j/55JPq1KmT6tevr379+mncuHF67733PPZRVFSkefPmqWXLlrrpppuUlJSkf//735Kko0eP6o033tBzzz2nHj16qFWrVnrrrbd0+vRpa/s9e/YoJCRECQkJio2N1fXXX6/Ro0eX+T2pSAQbAECV8swzz+itt97S9u3bPZbv2LFDnTt39ljWuXNnff/99x5f4q1bt7b+v8PhkMvlUnZ2tiSpd+/eql69uqpXr64WLVp47KtTp06qXr26QkND9eGHH2revHlq1aqVtb5du3Zlas/ZozGS1LFjxxLPz675n//5H914441yuVyqXr26Jk6cqD179nhsU79+fYWGhlrPo6OjrWPatWuXCgsLPV4nPDxcTZs2tZ736NFDsbGxatiwoZKSkrRgwQIdP35clYFgAwCoUrp06aJbb71Vjz/+uMdyY0yJK6SMMSW29/Pz83jucDhUVFQkSXr99deVkZGhjIwMffzxxx51//jHP/TVV1/p4MGD2r9/v/7whz94rA8JCSnxWqW1pyxXcZ2p2bBhg+6++2717t1by5cv15dffqknnnhChYWFZT6m0t6D4kJDQ/XFF1/o3XffVXR0tP7617/q2muv1ZEjRy64bUUj2AAAqpypU6fqww8/1Pr1661lzZs317p16zzq1q9fryZNmsjHx6dM+73mmmvUuHFjNW7cWLGxsR7r6tatq0aNGikiIqJM+2rWrFmp7WnWrJnHsg0bNpR4/pvf/EaS9L//+7+KjY3VE088oXbt2ikuLk4///xzmV7/jMaNG8vPz8/jdXJycvTdd9951Pn6+qp79+6aNm2avv76a/3000/69NNPvXqtisAN+gAAVU6rVq00aNAgj7ko48aN0w033KCnn35ad911l9LT0zVz5kzNmjWrUtr45z//WQMHDlSbNm3UrVs3ffjhh1q8eLFWrVrlUff++++rXbt2uvHGG7VgwQJt2rRJb7zxhqRfQ8mePXu0aNEi3XDDDfroo4+0ZMkSr9pRvXp1DR06VH/+858VERGhqKgoPfHEE6pW7b9jI8uXL9ePP/6oLl26qGbNmvr4449VVFTkcbrqciHYAAAu2tV4J+Cnn37aYxJtmzZt9N577+mvf/2rnn76aUVHR+upp57SkCFDKqV9t99+u1588UU9++yzGj16tBo0aKC5c+eqa9euHnWTJ0/WokWLNGLECLlcLi1YsEDNmzeXJN1222165JFHNHLkSBUUFKhv376aOHGikpOTvWrLs88+q6NHj6p///4KDQ3VuHHj5Ha7rfU1atTQ4sWLlZycrBMnTiguLk7vvvtuiXlGl4PDlOXk2VUoNzdXTqdTbrdbYWFhld0cXI0WnnUeO9GWfyaAV06cOKHdu3erQYMGCgwMrOzm4Cp2vs/SxX5/M8cGAADYBsEGAADYBsEGAADYBsEGAADYBsEGAOAVm15zgsvoUn6GCDYAgDI5c3fayrpVPuzjzGeo+B2PKwL3sQEAlImPj49q1Khh/YZQcHBwmW7vD5xhjNHx48eVnZ2tGjVqlPmOzt4g2AAAyszlckmSFW6A8qhRo4b1WapoBBsAQJk5HA5FR0crMjJSJ0+erOzm4Crk5+d3SUZqziDYAAC85uPjc0m/nIDyYvIwAACwDYINAACwDa+CTXJyshwOh8fj7Mk/xhglJycrJiZGQUFB6tq1q7755huPfRQUFGjUqFGqVauWQkJC1L9/f+3bt8+jJicnR0lJSXI6nXI6nUpKStKRI0fKf5QAAKBK8HrEpkWLFsrMzLQeW7dutdZNmzZN06dP18yZM7V582a5XC716NFDeXl5Vs2YMWO0ZMkSLVq0SOvWrdPRo0eVkJCg06dPWzWJiYnKyMhQamqqUlNTlZGRoaSkpIs8VAAAYHdeTx729fUt9RItY4xeeOEFPfHEExowYIAk6a233lJUVJQWLlyo4cOHy+1264033tD8+fPVvXt3SdI777yjunXratWqVbr11lu1Y8cOpaamasOGDWrfvr0k6bXXXlPHjh21c+dONW3a9GKOFwAA2JjXIzbff/+9YmJi1KBBA91999368ccfJUm7d+9WVlaWevbsadUGBAQoPj5e69evlyRt2bJFJ0+e9KiJiYlRy5YtrZr09HQ5nU4r1EhShw4d5HQ6rZrSFBQUKDc31+MBAACqFq+CTfv27fX222/rk08+0WuvvaasrCx16tRJhw4dUlZWliQpKirKY5uoqChrXVZWlvz9/VWzZs3z1kRGRpZ47cjISKumNFOmTLHm5DidTtWtW9ebQwMAADbgVbDp3bu37rjjDrVq1Urdu3fXRx99JOnXU05nFL+9tjHmgrfcLl5TWv2F9jNhwgS53W7rsXfv3jIdEwAAsI+Lutw7JCRErVq10vfff2/Nuyk+qpKdnW2N4rhcLhUWFionJ+e8NQcOHCjxWgcPHiwxGnS2gIAAhYWFeTwAAEDVclHBpqCgQDt27FB0dLQaNGggl8ullStXWusLCwuVlpamTp06SZLatm0rPz8/j5rMzExt27bNqunYsaPcbrc2bdpk1WzcuFFut9uqAQAAKI1XV0WNHz9e/fr1U7169ZSdna2//e1vys3N1eDBg+VwODRmzBilpKQoLi5OcXFxSklJUXBwsBITEyVJTqdTQ4cO1bhx4xQREaHw8HCNHz/eOrUlSc2aNVOvXr00bNgwzZkzR5L0wAMPKCEhgSuiAADAeXkVbPbt26d77rlH//nPf1S7dm116NBBGzZsUGxsrCTp0UcfVX5+vkaMGKGcnBy1b99eK1asUGhoqLWPGTNmyNfXVwMHDlR+fr66deumefPmefzmyIIFCzR69Gjr6qn+/ftr5syZFXG8AADAxhzGGFPZjbgUcnNz5XQ65Xa7mW+D8ll41mT1RFv+mQDAFediv7/5rSgAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbBBsAAGAbFxVspkyZIofDoTFjxljLjDFKTk5WTEyMgoKC1LVrV33zzTce2xUUFGjUqFGqVauWQkJC1L9/f+3bt8+jJicnR0lJSXI6nXI6nUpKStKRI0cuprkAAMDmyh1sNm/erFdffVWtW7f2WD5t2jRNnz5dM2fO1ObNm+VyudSjRw/l5eVZNWPGjNGSJUu0aNEirVu3TkePHlVCQoJOnz5t1SQmJiojI0OpqalKTU1VRkaGkpKSyttcAABQFZhyyMvLM3FxcWblypUmPj7e/OlPfzLGGFNUVGRcLpeZOnWqVXvixAnjdDrNK6+8Yowx5siRI8bPz88sWrTIqtm/f7+pVq2aSU1NNcYYs337diPJbNiwwapJT083ksy3335bpja63W4jybjd7vIcImDMAv33AQC4LC72+7tcIzYPP/yw+vbtq+7du3ss3717t7KystSzZ09rWUBAgOLj47V+/XpJ0pYtW3Ty5EmPmpiYGLVs2dKqSU9Pl9PpVPv27a2aDh06yOl0WjXFFRQUKDc31+MBAACqFl9vN1i0aJG++OILbd68ucS6rKwsSVJUVJTH8qioKP38889Wjb+/v2rWrFmi5sz2WVlZioyMLLH/yMhIq6a4KVOmaPLkyd4eDgAAsBGvRmz27t2rP/3pT3rnnXcUGBh4zjqHw+Hx3BhTYllxxWtKqz/ffiZMmCC322099u7de97XAwAA9uNVsNmyZYuys7PVtm1b+fr6ytfXV2lpafr73/8uX19fa6Sm+KhKdna2tc7lcqmwsFA5OTnnrTlw4ECJ1z948GCJ0aAzAgICFBYW5vEAAABVi1fBplu3btq6dasyMjKsR7t27TRo0CBlZGSoYcOGcrlcWrlypbVNYWGh0tLS1KlTJ0lS27Zt5efn51GTmZmpbdu2WTUdO3aU2+3Wpk2brJqNGzfK7XZbNQAAAMV5NccmNDRULVu29FgWEhKiiIgIa/mYMWOUkpKiuLg4xcXFKSUlRcHBwUpMTJQkOZ1ODR06VOPGjVNERITCw8M1fvx4tWrVypqM3KxZM/Xq1UvDhg3TnDlzJEkPPPCAEhIS1LRp04s+aAAAYE9eTx6+kEcffVT5+fkaMWKEcnJy1L59e61YsUKhoaFWzYwZM+Tr66uBAwcqPz9f3bp107x58+Tj42PVLFiwQKNHj7aunurfv79mzpxZ0c0FAAA24jDGmMpuxKWQm5srp9Mpt9vNfBuUz8KzJqon2vLPBACuOBf7/c1vRQEAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANsg2AAAANvwKtjMnj1brVu3VlhYmMLCwtSxY0f961//stYbY5ScnKyYmBgFBQWpa9eu+uabbzz2UVBQoFGjRqlWrVoKCQlR//79tW/fPo+anJwcJSUlyel0yul0KikpSUeOHCn/UQIAgCrBq2BTp04dTZ06VZ9//rk+//xz3XLLLbrtttus8DJt2jRNnz5dM2fO1ObNm+VyudSjRw/l5eVZ+xgzZoyWLFmiRYsWad26dTp69KgSEhJ0+vRpqyYxMVEZGRlKTU1VamqqMjIylJSUVEGHDAAAbMtcpJo1a5rXX3/dFBUVGZfLZaZOnWqtO3HihHE6neaVV14xxhhz5MgR4+fnZxYtWmTV7N+/31SrVs2kpqYaY4zZvn27kWQ2bNhg1aSnpxtJ5ttvvy1zu9xut5Fk3G73xR4iqqoF+u8DAHBZXOz3d7nn2Jw+fVqLFi3SsWPH1LFjR+3evVtZWVnq2bOnVRMQEKD4+HitX79ekrRlyxadPHnSoyYmJkYtW7a0atLT0+V0OtW+fXurpkOHDnI6nVZNaQoKCpSbm+vxAAAAVYvXwWbr1q2qXr26AgIC9OCDD2rJkiVq3ry5srKyJElRUVEe9VFRUda6rKws+fv7q2bNmuetiYyMLPG6kZGRVk1ppkyZYs3JcTqdqlu3rreHBgAArnJeB5umTZsqIyNDGzZs0EMPPaTBgwdr+/bt1nqHw+FRb4wpsay44jWl1V9oPxMmTJDb7bYee/fuLeshAQAAm/A62Pj7+6tx48Zq166dpkyZomuvvVYvvviiXC6XJJUYVcnOzrZGcVwulwoLC5WTk3PemgMHDpR43YMHD5YYDTpbQECAdbXWmQcAAKhaLvo+NsYYFRQUqEGDBnK5XFq5cqW1rrCwUGlpaerUqZMkqW3btvLz8/OoyczM1LZt26yajh07yu12a9OmTVbNxo0b5Xa7rRoAAIDS+HpT/Pjjj6t3796qW7eu8vLytGjRIq1Zs0apqalyOBwaM2aMUlJSFBcXp7i4OKWkpCg4OFiJiYmSJKfTqaFDh2rcuHGKiIhQeHi4xo8fr1atWql79+6SpGbNmqlXr14aNmyY5syZI0l64IEHlJCQoKZNm1bw4QMAADvxKtgcOHBASUlJyszMlNPpVOvWrZWamqoePXpIkh599FHl5+drxIgRysnJUfv27bVixQqFhoZa+5gxY4Z8fX01cOBA5efnq1u3bpo3b558fHysmgULFmj06NHW1VP9+/fXzJkzK+J4AQCAjTmMMaayG3Ep5Obmyul0yu12M98G5bPwrMnqibb8MwGAK87Ffn/zW1EAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2CDYAAMA2vAo2U6ZM0Q033KDQ0FBFRkbq9ttv186dOz1qjDFKTk5WTEyMgoKC1LVrV33zzTceNQUFBRo1apRq1aqlkJAQ9e/fX/v27fOoycnJUVJSkpxOp5xOp5KSknTkyJHyHSUAAKgSvAo2aWlpevjhh7VhwwatXLlSp06dUs+ePXXs2DGrZtq0aZo+fbpmzpypzZs3y+VyqUePHsrLy7NqxowZoyVLlmjRokVat26djh49qoSEBJ0+fdqqSUxMVEZGhlJTU5WamqqMjAwlJSVVwCEDAADbMhchOzvbSDJpaWnGGGOKioqMy+UyU6dOtWpOnDhhnE6neeWVV4wxxhw5csT4+fmZRYsWWTX79+831apVM6mpqcYYY7Zv324kmQ0bNlg16enpRpL59ttvy9Q2t9ttJBm3230xh4iqbIH++wAAXBYX+/19UXNs3G63JCk8PFyStHv3bmVlZalnz55WTUBAgOLj47V+/XpJ0pYtW3Ty5EmPmpiYGLVs2dKqSU9Pl9PpVPv27a2aDh06yOl0WjXFFRQUKDc31+MBAACqlnIHG2OMxo4dqxtvvFEtW7aUJGVlZUmSoqKiPGqjoqKsdVlZWfL391fNmjXPWxMZGVniNSMjI62a4qZMmWLNx3E6napbt255Dw0AAFylyh1sRo4cqa+//lrvvvtuiXUOh8PjuTGmxLLiiteUVn++/UyYMEFut9t67N27tyyHAQAAbKRcwWbUqFFatmyZVq9erTp16ljLXS6XJJUYVcnOzrZGcVwulwoLC5WTk3PemgMHDpR43YMHD5YYDTojICBAYWFhHg8AAFC1eBVsjDEaOXKkFi9erE8//VQNGjTwWN+gQQO5XC6tXLnSWlZYWKi0tDR16tRJktS2bVv5+fl51GRmZmrbtm1WTceOHeV2u7Vp0yarZuPGjXK73VYNAABAcb7eFD/88MNauHCh/vnPfyo0NNQamXE6nQoKCpLD4dCYMWOUkpKiuLg4xcXFKSUlRcHBwUpMTLRqhw4dqnHjxikiIkLh4eEaP368WrVqpe7du0uSmjVrpl69emnYsGGaM2eOJOmBBx5QQkKCmjZtWpHHDwAAbMSrYDN79mxJUteuXT2Wz507V0OGDJEkPfroo8rPz9eIESOUk5Oj9u3ba8WKFQoNDbXqZ8yYIV9fXw0cOFD5+fnq1q2b5s2bJx8fH6tmwYIFGj16tHX1VP/+/TVz5szyHCMAAKgiHMYYU9mNuBRyc3PldDrldruZb4PyWXjWRPVEW/6ZAMAV52K/v/mtKAAAYBsEGwAAYBtezbEBAFzFFha7DxinWGFDjNgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADbINgAAADb8K3sBgAAcMktdPz3/yeaymsHLjlGbAAAgG0QbAAAgG1wKgoArhbnOp1y9vLi64AqhhEbAABgGwQbAABgGwQbAABgGwQbAABgGwQbAABgG14Hm88++0z9+vVTTEyMHA6Hli5d6rHeGKPk5GTFxMQoKChIXbt21TfffONRU1BQoFGjRqlWrVoKCQlR//79tW/fPo+anJwcJSUlyel0yul0KikpSUeOHPH6AAEAQNXhdbA5duyYrr32Ws2cObPU9dOmTdP06dM1c+ZMbd68WS6XSz169FBeXp5VM2bMGC1ZskSLFi3SunXrdPToUSUkJOj06dNWTWJiojIyMpSamqrU1FRlZGQoKSmpHIcIAACqCq/vY9O7d2/17t271HXGGL3wwgt64oknNGDAAEnSW2+9paioKC1cuFDDhw+X2+3WG2+8ofnz56t79+6SpHfeeUd169bVqlWrdOutt2rHjh1KTU3Vhg0b1L59e0nSa6+9po4dO2rnzp1q2rRpeY8XAADYWIXOsdm9e7eysrLUs2dPa1lAQIDi4+O1fv16SdKWLVt08uRJj5qYmBi1bNnSqklPT5fT6bRCjSR16NBBTqfTqimuoKBAubm5Hg8AAFC1VGiwycrKkiRFRUV5LI+KirLWZWVlyd/fXzVr1jxvTWRkZIn9R0ZGWjXFTZkyxZqP43Q6Vbdu3Ys+HgAAcHW5JFdFORyet/c2xpRYVlzxmtLqz7efCRMmyO12W4+9e/eWo+UAAOBqVqHBxuVySVKJUZXs7GxrFMflcqmwsFA5OTnnrTlw4ECJ/R88eLDEaNAZAQEBCgsL83gAAICqpUKDTYMGDeRyubRy5UprWWFhodLS0tSpUydJUtu2beXn5+dRk5mZqW3btlk1HTt2lNvt1qZNm6yajRs3yu12WzUAAADFeX1V1NGjR/XDDz9Yz3fv3q2MjAyFh4erXr16GjNmjFJSUhQXF6e4uDilpKQoODhYiYmJkiSn06mhQ4dq3LhxioiIUHh4uMaPH69WrVpZV0k1a9ZMvXr10rBhwzRnzhxJ0gMPPKCEhASuiAIAAOfkdbD5/PPPdfPNN1vPx44dK0kaPHiw5s2bp0cffVT5+fkaMWKEcnJy1L59e61YsUKhoaHWNjNmzJCvr68GDhyo/Px8devWTfPmzZOPj49Vs2DBAo0ePdq6eqp///7nvHcOAFzNFu/MLLFsQNPoSmiJzSw8/9xO2JPDGGMquxGXQm5urpxOp9xuN/NtUD5n/0cx0ZZ/JrhClDnYnOszWfwL/Fyf17LW2cW5go3dj/sqd7Hf316P2AAALk5pQQZAxSDYAMAVqHj4uaynpqrayA5shWADAFeBxTszpba/WM8HVGJbrnoEN1sj2ADAJXYpTj157LPtLxqwJabCXwO4Gl2SOw8DAABUBoINAACwDU5FAUAFq/Srni7XrQq4JQKuQAQbALCBxWdNLJZk7zk3BCqcB6eiAACAbRBsAACAbXAqCgBwaXHqCJcRwQYA7IwfgkQVw6koAABgG4zYAMBFqvTLuwFYGLEBAAC2wYgNAHiB0RngykawAQAbKn7DPqmUm/ZxtRJsiGADACiTEqNVZ4en/1s3oGn0ZWwRUBLBBgDOUvzLu8p8UXNZOGyCYAPg6lL8C/gSn0JhTo13bBUMOVV3VeKqKAAAYBuM2AC4MpxvJIbTJBXiQr8AXtqEY0nW/JlyvebOTI+5OAPKvSegbAg2AFBFnTPIAFcxgg0AW7vQnA/m0KCy2Go+0hWEYAPg8irrhMyLPf10ZntGJYAqhWADoPJUxNwZL69cYYQGsDeCDQD7YJIxUOURbABcWoQNXOn4jNoKwQbAVeNClysDAMEGwFWLy5WvQpf5ztEV5mptdxVEsAFQMS7B7ecJLvZTYtStLBtxqgheINgAAOyBAAQRbABcCmX4gmE0BhWCMINiCDbA/ylxF9BKagcAoPwINgDKz4t/LTNCA+ByqFbZDQAAAKgoBBsAAGAbnIoCUHZM1AR+dQlub4CKQbABcH7lDDPMqUFVU+IChKbRldSSqo1gA6AkRmZwuZxr5IPPIMqJYAOgQr5EGKHBRbtEYaa0zya/M2ZfBBsAQKUhdKCicVUUAACwDUZsgKqEeQtAxTvzd1Vs9InJxJWDYAPYGUEGV6ESvwDOqSl4gWADoFyYLAx4p/gIDi4Ngg0AAF4i2F+5CDaA3XD6CTbDqSl4g2AD2MElDjP86xTA1YJgA1wtGIkBgAu64oPNrFmz9OyzzyozM1MtWrTQCy+8oJtuuqmymwVcHpcpzDDUDzthhLFqu6KDzT/+8Q+NGTNGs2bNUufOnTVnzhz17t1b27dvV7169Sq7ecClcQWMzPDFgCtZRQRxwrx9XdHBZvr06Ro6dKjuv/9+SdILL7ygTz75RLNnz9aUKVMquXVABbnIIEMIQVV3Kf4G+Lu6el2xwaawsFBbtmzRY4895rG8Z8+eWr9+fYn6goICFRQUWM/dbrckKTc399I2FLZx/Giex/Pc42c/uYSfo+OeT5ddv9O77Yu1G4D33mla7O+uEv6ucnNDSi58z/nf/z/QffkaU4nOfG8bYy5QWborNtj85z//0enTpxUVFeWxPCoqSllZWSXqp0yZosmTJ5dYXrdu3UvWRlQhw5wXrqkwTS/jawG4alzW/w5Vvry8PDmd3h/zFRtsznA4PIfpjTEllknShAkTNHbsWOt5UVGRDh8+rIiIiFLrcWnl5uaqbt262rt3r8LCwiq7OVUSfVD56IPKRx9UPm/7wBijvLw8xcSUb97TFRtsatWqJR8fnxKjM9nZ2SVGcSQpICBAAQEBHstq1KhxKZuIMggLC+M/JpWMPqh89EHlow8qnzd9UJ6RmjOqlXvLS8zf319t27bVypUrPZavXLlSnTp1qqRWAQCAK9kVO2IjSWPHjlVSUpLatWunjh076tVXX9WePXv04IMPVnbTAADAFeiKDjZ33XWXDh06pKeeekqZmZlq2bKlPv74Y8XGxlZ203ABAQEBmjRpUonTg7h86IPKRx9UPvqg8l3uPnCY8l5PBQAAcIW5YufYAAAAeItgAwAAbINgAwAAbINgAwAAbINgAwAAbINgg1J99tln6tevn2JiYuRwOLR06VKP9UePHtXIkSNVp04dBQUFqVmzZpo9e7ZHTVZWlpKSkuRyuRQSEqI2bdrof/7nfzxqcnJylJSUJKfTKafTqaSkJB05cuQSH93V4UJ9cODAAQ0ZMkQxMTEKDg5Wr1699P3331vrDx8+rFGjRqlp06YKDg5WvXr1NHr0aOsHYs+gD87tYvvgjPT0dN1yyy0KCQlRjRo11LVrV+Xn51vr6YNzmzJlim644QaFhoYqMjJSt99+u3bu9PzBSmOMkpOTFRMTo6CgIHXt2lXffPONR01BQYFGjRqlWrVqKSQkRP3799e+ffs8auiH0lVUH5xd27t371L/piqiDwg2KNWxY8d07bXXaubMmaWuf+SRR5Samqp33nlHO3bs0COPPKJRo0bpn//8p1WTlJSknTt3atmyZdq6dasGDBigu+66S19++aVVk5iYqIyMDKWmpio1NVUZGRlKSkq65Md3NThfHxhjdPvtt+vHH3/UP//5T3355ZeKjY1V9+7ddezYMUnSL7/8ol9++UXPPfectm7dqnnz5ik1NVVDhw712Bd9cG4X2wfSr6GmV69e6tmzpzZt2qTNmzdr5MiRqlbtv//5pQ/OLS0tTQ8//LA2bNiglStX6tSpU+rZs6fHezxt2jRNnz5dM2fO1ObNm+VyudSjRw/l5f33F7rHjBmjJUuWaNGiRVq3bp2OHj2qhIQEnT592qqhH0pXUX1wxgsvvHDO33CskD4wwAVIMkuWLPFY1qJFC/PUU095LGvTpo158sknrechISHm7bff9qgJDw83r7/+ujHGmO3btxtJZsOGDdb69PR0I8l8++23FXwUV7fifbBz504jyWzbts1adurUKRMeHm5ee+21c+7nvffeM/7+/ubkyZPGGPrAG+Xtg/bt23v8XRRHH3gnOzvbSDJpaWnGGGOKioqMy+UyU6dOtWpOnDhhnE6neeWVV4wxxhw5csT4+fmZRYsWWTX79+831apVM6mpqcYY+sEb5emDMzIyMkydOnVMZmZmib+piuoDRmxQLjfeeKOWLVum/fv3yxij1atX67vvvtOtt97qUfOPf/xDhw8fVlFRkRYtWqSCggJ17dpV0q//knU6nWrfvr21TYcOHeR0OrV+/frLfUhXlYKCAklSYGCgtczHx0f+/v5at27dObdzu90KCwuTr++vNx2nD8qvLH2QnZ2tjRs3KjIyUp06dVJUVJTi4+M9+og+8M6ZU6nh4eGSpN27dysrK0s9e/a0agICAhQfH2+9f1u2bNHJkyc9amJiYtSyZUurhn4ou/L0gSQdP35c99xzj2bOnCmXy1VivxXVBwQblMvf//53NW/eXHXq1JG/v7969eqlWbNm6cYbb7Rq/vGPf+jUqVOKiIhQQECAhg8friVLlqhRo0aSfp2DExkZWWLfkZGRJX7VHZ5+85vfKDY2VhMmTFBOTo4KCws1depUZWVlKTMzs9RtDh06pKefflrDhw+3ltEH5VeWPvjxxx8lScnJyRo2bJhSU1PVpk0bdevWzZqLQx+UnTFGY8eO1Y033qiWLVtKkvUeRUVFedRGRUVZ67KysuTv76+aNWuet4Z+uLDy9oH06xSGTp066bbbbit13xXVB1f0b0XhyvX3v/9dGzZs0LJlyxQbG6vPPvtMI0aMUHR0tLp37y5JevLJJ5WTk6NVq1apVq1aWrp0qe68806tXbtWrVq1kqRSz7MaY855/hW/8vPz0wcffKChQ4cqPDxcPj4+6t69u3r37l1qfW5urvr27avmzZtr0qRJHuvog/IpSx8UFRVJkoYPH64//vGPkqTrr79e//73v/Xmm29qypQpkuiDsho5cqS+/vrrUkcli79XZXn/itfQDxdW3j5YtmyZPv30U485lqWpiD4g2MBr+fn5evzxx7VkyRL17dtXktS6dWtlZGToueeeU/fu3bVr1y7NnDlT27ZtU4sWLSRJ1157rdauXauXX35Zr7zyilwulw4cOFBi/wcPHiyR/FFS27ZtlZGRIbfbrcLCQtWuXVvt27dXu3btPOry8vLUq1cvVa9eXUuWLJGfn5+1jj64OBfqg+joaElS8+bNPbZr1qyZ9uzZI4k+KKtRo0Zp2bJl+uyzz1SnTh1r+ZlTGllZWdb7Lf16GvDM++dyuVRYWKicnByPUZvs7Gx16tTJqqEfzu9i+uDTTz/Vrl27VKNGDY993nHHHbrpppu0Zs2aCusDTkXBaydPntTJkyc9ruqQfp1fcOZfqMePH5ek89Z07NhRbrdbmzZtstZv3LhRbrfb+o8NLszpdKp27dr6/vvv9fnnn3sM8+bm5qpnz57y9/fXsmXLPOaDSPRBRTlXH9SvX18xMTElLo397rvvFBsbK4k+uBBjjEaOHKnFixfr008/VYMGDTzWN2jQQC6XSytXrrSWFRYWKi0tzXr/2rZtKz8/P4+azMxMbdu2zaqhH86tIvrgscce09dff62MjAzrIUkzZszQ3LlzJVVgH5R9HjSqkry8PPPll1+aL7/80kgy06dPN19++aX5+eefjTHGxMfHmxYtWpjVq1ebH3/80cydO9cEBgaaWbNmGWOMKSwsNI0bNzY33XST2bhxo/nhhx/Mc889ZxwOh/noo4+s1+nVq5dp3bq1SU9PN+np6aZVq1YmISGhUo75SnOhPnjvvffM6tWrza5du8zSpUtNbGysGTBggLV9bm6uad++vWnVqpX54YcfTGZmpvU4deqUVUcfnNvF9oExxsyYMcOEhYWZ999/33z//ffmySefNIGBgeaHH36wauiDc3vooYeM0+k0a9as8fgMHz9+3KqZOnWqcTqdZvHixWbr1q3mnnvuMdHR0SY3N9eqefDBB02dOnXMqlWrzBdffGFuueUWc+211/K3UAYV1QfFqZQrbiuiDwg2KNXq1auNpBKPwYMHG2OMyczMNEOGDDExMTEmMDDQNG3a1Dz//POmqKjI2sd3331nBgwYYCIjI01wcLBp3bp1icu/Dx06ZAYNGmRCQ0NNaGioGTRokMnJybmMR3rlulAfvPjii6ZOnTrGz8/P1KtXzzz55JOmoKDggttLMrt377bq6INzu9g+OGPKlCmmTp06Jjg42HTs2NGsXbvWYz19cG7n+gzPnTvXqikqKjKTJk0yLpfLBAQEmC5dupitW7d67Cc/P9+MHDnShIeHm6CgIJOQkGD27NnjUUM/lK6i+qC0/RYPNhXRB47/2zkAAMBVjzk2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANgg2AADANv4/zuWV5ti1qncAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#colors = ['pink', 'orange', 'yellow', 'pink', 'red', 'green']\n",
    "\n",
    "#for (s, st) in enumerate(sample_types):\n",
    "#    x = samples.loc[samples.sampletypeid==st, 'createdt']\n",
    "#    plt.hist(x, alpha=0.5, bins=100, label=str(st), color=colors[s])  #density=True, stacked=True\n",
    "#x = samples.loc[~samples.sampletypeid.isin(sample_types), 'createdt']\n",
    "#plt.hist(x, alpha=0.5, bins=100, label=str('Other'), color='blue')\n",
    "\n",
    "#plt.suptitle('Probability Histogram of Diamond Depths', y=1.05, size=16)\n",
    "#ax.set_xlim(50, 70); ax.set_ylim(0, 1);\n",
    "#plt.tight_layout();\n",
    "\n",
    "#_ = plt.hist(samples.createdt, bins=100)\n",
    "#plt.gca().set(title='Frequency Histogram of Diamond Depths', ylabel='Frequency')\n",
    "#plt.xlim(50,75)\n",
    "#plt.legend()\n",
    "\n",
    "study_design = study.iloc[0]['u_studydesign']\n",
    "if study_design == 'Case Control':\n",
    "    pass\n",
    "else:\n",
    "    probands = cohort.loc[cohort['cohorttype'] == 'Proband', 'birthdt']\n",
    "    non_probands = cohort.loc[cohort['cohorttype'] != 'Proband', 'birthdt']\n",
    "    _ = plt.hist(probands, bins=100, color=\"orange\", label=\"Probands\")\n",
    "    _ = plt.hist(non_probands, bins=100, color=\"lightblue\", label=\"Non-Probands\")\n",
    "    plt.title('Participant Date-of-Birth')\n",
    "    plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "913d49e9-e719-43b8-ba8c-66864526de1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genderflag\n",
       "F    52931\n",
       "M    44553\n",
       "U    18840\n",
       "?        1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('genderflag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2e615bef-4dfa-4244-934b-e4fba865b5d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u_race\n",
       "NaN                 65924\n",
       "Caucasian           38888\n",
       "African American    10040\n",
       "Asian                 900\n",
       "Other                 691\n",
       "Native American        61\n",
       "Pacific Islander       33\n",
       "dtype: int64"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('u_race', dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "978f584b-2549-48a1-8a3b-f660bb72508d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u_ethnicity\n",
       "NaN                                                            100156\n",
       "Hispanic                                                         8439\n",
       "Not Hispanic                                                     7462\n",
       "Not Hispanic or Latino                                            331\n",
       "Hispanic or Latino                                                111\n",
       "European (Non-Jewish), Americal Indian                              2\n",
       "European (non-Jewsih), Hispanic                                     2\n",
       "Mexican                                                             2\n",
       "European (non-jewish) and Black/African American                    2\n",
       "Mexican, Italian                                                    2\n",
       "Hispanic, African American                                          2\n",
       "Mexican, England                                                    1\n",
       "Mexican, French, Irish                                              1\n",
       "Mexican, German, Irish                                              1\n",
       "Mexican/German                                                      1\n",
       "half Mexican, half white                                            1\n",
       "american indian, earopean (non-jewish), hispanic                    1\n",
       "Mexican American, Irish, German, French, Spanish                    1\n",
       "Spanish, German, Dutch                                              1\n",
       "Swedish, Mexican                                                    1\n",
       "Welsh, Mexican, Native American                                     1\n",
       "Mexican, Black                                                      1\n",
       "African-American, Mexican, Japanese                                 1\n",
       "Mexican American                                                    1\n",
       "Mexican America                                                     1\n",
       "Am Indian, Spanish, Irish, German                                   1\n",
       "Irish, Finlander, Mexican                                           1\n",
       "ISREALI                                                             1\n",
       "Greek, Mexican, English                                             1\n",
       "German, Irish, Indian, Mexican                                      1\n",
       "Filipino, Hawaiian, Mexican                                         1\n",
       "Father was white Mother was Black Subj did not know Father          1\n",
       "European (non-Jewsih), Native American                              1\n",
       "European (non-Jewish), Hispanic, Americal Indian                    1\n",
       "European (non Jewish)                                               1\n",
       "Black, white, Indian, Mexican, French, German                       1\n",
       "Italian, Hispanic, European                                         1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('u_ethnicity', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd5c505-3466-4440-b690-0cb4a9c77165",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "712e3eb0-e477-49ce-be70-080f38848c91",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>collectiondesc</th>\n",
       "      <th>createdt</th>\n",
       "      <th>createby</th>\n",
       "      <th>moddt</th>\n",
       "      <th>modby</th>\n",
       "      <th>activeflag</th>\n",
       "      <th>studyid</th>\n",
       "      <th>samplesexpected</th>\n",
       "      <th>securityuser</th>\n",
       "      <th>securitydepartment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL-AACR-0266</td>\n",
       "      <td>ESCAPE Clinical Trial</td>\n",
       "      <td>2020-10-20 17:28:48</td>\n",
       "      <td>resat</td>\n",
       "      <td>2022-02-21 11:47:14</td>\n",
       "      <td>relef</td>\n",
       "      <td>Y</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>resat</td>\n",
       "      <td>SGG LAB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  u_collectionid         collectiondesc            createdt createby  \\\n",
       "0   CL-AACR-0266  ESCAPE Clinical Trial 2020-10-20 17:28:48    resat   \n",
       "\n",
       "                moddt  modby activeflag studyid samplesexpected securityuser  \\\n",
       "0 2022-02-21 11:47:14  relef          Y  ESCAPE            None        resat   \n",
       "\n",
       "  securitydepartment  \n",
       "0            SGG LAB  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_collections = session.query(Collection).filter(Collection.c.studyid==STUDY)\n",
    "collections = pd.read_sql(q_collections.statement, session.bind)\n",
    "collections.drop(columns=['usersequence', 'auditsequence', 'tracelogid', 'timepointname', 'timepointstart', 'timepointend', 'timepointtimeunits', \n",
    "                          'visitstart',  'visitname', 'visittimeunits', 'visitend', 'alternatename', 'receivableflag', 'restrictionclassid',\n",
    "                         'protocolname', 'startdt', 'enddt', 'templateflag', 'notes', 'createtool', 'modtool'], inplace=True)\n",
    "collections"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c27b4f0-42a8-4a2d-9bb9-c7c36cb459d6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sample Families (w/Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "7bd36d82-cd46-4cf9-9b8b-8a2b9f6de997",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_samplefamilyid</th>\n",
       "      <th>studysiteid</th>\n",
       "      <th>collectiondt</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>tissueid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>u_initsampletypeid</th>\n",
       "      <th>u_inittreatmenttypeid</th>\n",
       "      <th>u_initcontainertypeid</th>\n",
       "      <th>u_initenvironmentid</th>\n",
       "      <th>u_specimentypeid</th>\n",
       "      <th>u_packageid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_samplefamilyid, studysiteid, collectiondt, subjectid, tissueid, createdt, u_collectionid, u_initsampletypeid, u_inittreatmenttypeid, u_initcontainertypeid, u_initenvironmentid, u_specimentypeid, u_packageid, tissuedesc]\n",
       "Index: []"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_sf = session.query(SampleFamily, Tissue.c.tissuedesc).filter(SampleFamily.c.sstudyid==STUDY, Tissue.c.s_tissueid==SampleFamily.c.tissueid)\n",
    "sf = pd.read_sql(q_sf.statement, session.bind)\n",
    "sf.drop(columns=['samplefamilydesc', 'initialpackageid', 'initialdepartmentid', 'collectmethodid', 'verifiedby', 'verifieddt',\n",
    "                'u_visitname', 'u_timepointname', 'approvedby', 'approveddt', 'recieveddt', 'cocflag', 'restrictionsflag', 'age',\n",
    "                'ageunits', 'restrictclassid', 'externalsubject', 'animal', 'diseaseid', 'diseaseid', 'metastasisid', 'clinicaldiagid',\n",
    "                'conditionalapprovalreason', 'conditionalapprovalflag', 'specimendefid', 'activeflag', 'auditsequence', 'clinicalevent',\n",
    " 'clinicalprotocolid',\n",
    " 'clinicalprotocolrevision',\n",
    " 'clinicalprotocolversionid',\n",
    "                 'createby',\n",
    " 'createtool',\n",
    " 'deviationdesc',\n",
    " 'eventdefid',\n",
    " 'initialmass',\n",
    " 'initialmassunits',\n",
    " 'initialvolume',\n",
    " 'initialvolumeunits',\n",
    " 'kittrackitem',\n",
    " 'modby',\n",
    " 'moddt',\n",
    " 'modtool',\n",
    " 'notes', 'tracelogid',  'usersequence', 'templateflag', 'participantid', 'u_legacyfamilyid', 'u_collectiontypid', 'sampletypeid',\n",
    "                 'participanteventid', 'sstudyid',\n",
    "                ], inplace=True)\n",
    "sf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b536b713-5e25-4c91-aa45-423aaec7c57d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: int64)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf.value_counts(['u_collectionid', 'tissuedesc'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c395c8-1ba0-4224-929b-f61e8d0a9d00",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9345767c-e1ba-49ab-ac83-0b7abb17af2d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_sampleid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>batchid</th>\n",
       "      <th>samplefamilyid</th>\n",
       "      <th>sstudyid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-001923284</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2021-05-04 08:22:09</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210504-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-001983201</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:04</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-001926215</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-06-11 14:41:26</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-001926207</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2021-06-11 12:00:18</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-001983203</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:06</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>S-001972944</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-08-30 12:50:19</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220830-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>S-001972945</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-08-30 12:50:19</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220830-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>697</th>\n",
       "      <td>S-001970980</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-07-29 11:11:37</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>S-001983234</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:45</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>S-001969298</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2022-07-12 09:07:41</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>700 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      s_sampleid sampletypeid            createdt batchid   samplefamilyid  \\\n",
       "0    S-001923284       Saliva 2021-05-04 08:22:09    None  SF-210504-00002   \n",
       "1    S-001983201          DNA 2022-11-09 15:56:04    None  SF-210611-00002   \n",
       "2    S-001926215          DNA 2021-06-11 14:41:26    None  SF-210611-00002   \n",
       "3    S-001926207       Saliva 2021-06-11 12:00:18    None  SF-210611-00002   \n",
       "4    S-001983203          DNA 2022-11-09 15:56:06    None  SF-210611-00004   \n",
       "..           ...          ...                 ...     ...              ...   \n",
       "695  S-001972944          DNA 2022-08-30 12:50:19    None  SF-220830-00004   \n",
       "696  S-001972945          DNA 2022-08-30 12:50:19    None  SF-220830-00004   \n",
       "697  S-001970980          DNA 2022-07-29 11:11:37    None  SF-220712-00003   \n",
       "698  S-001983234          DNA 2022-11-09 15:56:45    None  SF-220712-00003   \n",
       "699  S-001969298       Saliva 2022-07-12 09:07:41    None  SF-220712-00003   \n",
       "\n",
       "    sstudyid  \n",
       "0     ESCAPE  \n",
       "1     ESCAPE  \n",
       "2     ESCAPE  \n",
       "3     ESCAPE  \n",
       "4     ESCAPE  \n",
       "..       ...  \n",
       "695   ESCAPE  \n",
       "696   ESCAPE  \n",
       "697   ESCAPE  \n",
       "698   ESCAPE  \n",
       "699   ESCAPE  \n",
       "\n",
       "[700 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_samples = session.query(Sample).filter(SampleFamily.c.sstudyid==STUDY, Sample.c.samplefamilyid==SampleFamily.c.s_samplefamilyid)\n",
    "samples = pd.read_sql(q_samples.statement, session.bind)\n",
    "samples.drop(columns=['sampledesc', 'createby', 'createtool', 'submitterid', 'modby', 'submitteddt', 'modtool', 'notes', 'auditsequence', 'tracelogid',\n",
    "                     'usersequence', 'activeflag', 'allocatedforaddressid', 'allocatedforaddresstype', 'allocatedfordepartmentid', 'auditsequence', 'autofinalrptflag',\n",
    "                     'autoreceiveflag', 'batchstageid',  'cancelledby', 'cancelleddt', 'classification', 'cocflag', 'cocrequiredflag', 'collectedby', 'collectiondt',\n",
    "                     'completedt', 'concentration', 'concentrationunits', 'conditionlabel', 'confirmedby', 'confirmeddt', 'controlsubstanceflag','deviations', 'disposaldt',\n",
    "                     'disposalstatus', 'disposaltargetdt', 'disposedby', 'duedt', 'duedtoffset', 'duedtoffsettimeunit', 'duedtoverrideflag', 'eventdt', 'eventnum', 'eventplan',\n",
    "                     'eventplanitem', 'glpflag', 'instrumentid', 'locationid', 'locationpath', 'moddt', 'monitorgroupid', 'numberlabels', 'physicalcondition', 'pooledflag',\n",
    "                     'preptypeid', 'previousstoragestatus', 'priority', 'processinstruction', 'processtype', 'productid', 'productversionid', 'projectid', 'qcsampletype',\n",
    "                     'reagentlotid', 'receivedby', 'receiveddt', 'receiverequiredflag', 'requestid', 'requestitemdetailid', 'requestitemid', 'restrictionsflag',\n",
    "                      'reviewdisposition', 'reviewedby', 'revieweddt', 'reviewremarks', 'reviewrequiredflag', 'samplepointid', 'samplepointinstance', 'samplestatus',\n",
    "                      'schedulerulelabel', 'scheduletemplateflag', 'sdiworkitemcompletionstatus', 'securitydepartment', 'securityuser', 'sourcesdiworkitemid',\n",
    "                      'sourcespid', 'sourcesplevelid', 'sourcespsourcelabel', 'sourcespversionid',    'starttestingdt',\n",
    "                      'storagedisposalstatus', 'studyid', 'samplesubtypeid', 'u_aliquotnumber', 'specimentype', 'treatments',\n",
    "                      'storagestatus',   'templateflag',  'workorderid', 'basedonsampleid', 'u_qcfailureflag', 'u_qctype'], inplace=True)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bd7888d7-0c83-44dd-b502-7d6f8cf11396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sampletypeid\n",
       "Saliva          390\n",
       "DNA             310\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.value_counts(['sampletypeid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cddfe305-a3f2-4e41-9070-db98da6d5ea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_samplefamilyid</th>\n",
       "      <th>studysiteid</th>\n",
       "      <th>collectiondt</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>tissueid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>u_initsampletypeid</th>\n",
       "      <th>u_inittreatmenttypeid</th>\n",
       "      <th>u_initcontainertypeid</th>\n",
       "      <th>u_initenvironmentid</th>\n",
       "      <th>u_specimentypeid</th>\n",
       "      <th>u_packageid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_samplefamilyid, studysiteid, collectiondt, subjectid, tissueid, createdt, u_collectionid, u_initsampletypeid, u_inittreatmenttypeid, u_initcontainertypeid, u_initenvironmentid, u_specimentypeid, u_packageid, tissuedesc]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf[sf['s_samplefamilyid'] == 'SFO-000000876']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "43b7590a-92ff-46c8-a89d-1827d8a08d78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_sampleid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>batchid</th>\n",
       "      <th>samplefamilyid</th>\n",
       "      <th>sstudyid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_sampleid, sampletypeid, createdt, batchid, samplefamilyid, sstudyid]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[samples['samplefamilyid'] == 'SFO-000000876']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5d41f03e-6752-4c2f-af06-9e65eb9449a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Column('sourcesampleid', VARCHAR(length=40), ForeignKey('sapphire8.s_sample.s_sampleid'), table=<S_SAMPLEMAP>, primary_key=True, nullable=False),\n",
       " Column('destsampleid', VARCHAR(length=40), ForeignKey('sapphire8.s_sample.s_sampleid'), table=<S_SAMPLEMAP>, primary_key=True, nullable=False),\n",
       " Column('auditsequence', NUMBER(asdecimal=False), table=<S_SAMPLEMAP>),\n",
       " Column('usersequence', NUMBER(asdecimal=False), table=<S_SAMPLEMAP>),\n",
       " Column('createdt', DATE(), table=<S_SAMPLEMAP>),\n",
       " Column('createby', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('createtool', VARCHAR(length=20), table=<S_SAMPLEMAP>),\n",
       " Column('tracelogid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('moddt', DATE(), table=<S_SAMPLEMAP>),\n",
       " Column('modby', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('modtool', VARCHAR(length=20), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanversionid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanitemid', VARCHAR(length=20), table=<S_SAMPLEMAP>)]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(SampleMap.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4847e681-c8f6-4219-8ae5-3465308a56a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sourcesampleid</th>\n",
       "      <th>destsampleid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>createdt_source</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt_dest</th>\n",
       "      <th>sampletypeid_dest</th>\n",
       "      <th>map_minus_sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160657</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160656</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160658</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160658</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160656</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160657</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592463</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592468</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592464</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592465</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592466</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592467</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592467</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592466</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592465</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592464</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592463</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592468</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>S-000592467</td>\n",
       "      <td>S-000967661</td>\n",
       "      <td>2010-10-09 17:51:31</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>S-000592467</td>\n",
       "      <td>S-000967661</td>\n",
       "      <td>2010-10-09 17:51:31</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>S-000592464</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>S-000592464</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>S-000967661</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>S-000967661</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sourcesampleid destsampleid            createdt createdt_source  \\\n",
       "0     S-000000876  S-000160657 2010-06-22 09:30:12             NaT   \n",
       "1     S-000000876  S-000160656 2010-06-22 09:30:12             NaT   \n",
       "2     S-000000876  S-000160658 2010-06-22 09:30:12             NaT   \n",
       "3     S-000000876  S-000160658 2010-06-22 09:30:12             NaT   \n",
       "4     S-000000876  S-000160656 2010-06-22 09:30:12             NaT   \n",
       "5     S-000000876  S-000160657 2010-06-22 09:30:12             NaT   \n",
       "6     S-000160658  S-000592463 2010-06-22 09:30:12             NaT   \n",
       "7     S-000160658  S-000592468 2010-06-22 09:30:12             NaT   \n",
       "8     S-000160658  S-000592464 2010-06-22 09:30:12             NaT   \n",
       "9     S-000160658  S-000592465 2010-06-22 09:30:12             NaT   \n",
       "10    S-000160658  S-000592466 2010-06-22 09:30:12             NaT   \n",
       "11    S-000160658  S-000592467 2010-06-22 09:30:12             NaT   \n",
       "12    S-000160658  S-000592467 2010-06-22 09:30:12             NaT   \n",
       "13    S-000160658  S-000592466 2010-06-22 09:30:12             NaT   \n",
       "14    S-000160658  S-000592465 2010-06-22 09:30:12             NaT   \n",
       "15    S-000160658  S-000592464 2010-06-22 09:30:12             NaT   \n",
       "16    S-000160658  S-000592463 2010-06-22 09:30:12             NaT   \n",
       "17    S-000160658  S-000592468 2010-06-22 09:30:12             NaT   \n",
       "18    S-000592467  S-000967661 2010-10-09 17:51:31             NaT   \n",
       "19    S-000592467  S-000967661 2010-10-09 17:51:31             NaT   \n",
       "20    S-000592464  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "21    S-000592464  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "22    S-000967661  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "23    S-000967661  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "\n",
       "   sampletypeid createdt_dest sampletypeid_dest map_minus_sample  \n",
       "0           NaN           NaT               NaN              NaT  \n",
       "1           NaN           NaT               NaN              NaT  \n",
       "2           NaN           NaT               NaN              NaT  \n",
       "3           NaN           NaT               NaN              NaT  \n",
       "4           NaN           NaT               NaN              NaT  \n",
       "5           NaN           NaT               NaN              NaT  \n",
       "6           NaN           NaT               NaN              NaT  \n",
       "7           NaN           NaT               NaN              NaT  \n",
       "8           NaN           NaT               NaN              NaT  \n",
       "9           NaN           NaT               NaN              NaT  \n",
       "10          NaN           NaT               NaN              NaT  \n",
       "11          NaN           NaT               NaN              NaT  \n",
       "12          NaN           NaT               NaN              NaT  \n",
       "13          NaN           NaT               NaN              NaT  \n",
       "14          NaN           NaT               NaN              NaT  \n",
       "15          NaN           NaT               NaN              NaT  \n",
       "16          NaN           NaT               NaN              NaT  \n",
       "17          NaN           NaT               NaN              NaT  \n",
       "18          NaN           NaT               NaN              NaT  \n",
       "19          NaN           NaT               NaN              NaT  \n",
       "20          NaN           NaT               NaN              NaT  \n",
       "21          NaN           NaT               NaN              NaT  \n",
       "22          NaN           NaT               NaN              NaT  \n",
       "23          NaN           NaT               NaN              NaT  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q_smap = session.query(SampleMap)\n",
    "q_smap = q_smap.filter(or_(SampleMap.c.destsampleid==Sample.c.s_sampleid, SampleMap.c.sourcesampleid==Sample.c.s_sampleid)).filter(Sample.c.samplefamilyid=='SFO-000000876')\n",
    "q_smap = q_smap.order_by(SampleMap.c.createdt, SampleMap.c.sourcesampleid)\n",
    "\n",
    "sample_map = pd.read_sql(q_smap.statement, session.bind)\n",
    "sample_map.drop(columns=['createby','createtool','tracelogid','moddt','modby','modtool','childsampleplanid','childsampleplanversionid','childsampleplanitemid', \n",
    "                         'auditsequence','usersequence'], inplace=True)\n",
    "sample_info = samples[['s_sampleid', 'createdt', 'sampletypeid']]\n",
    "sample_map = sample_map.merge(sample_info, left_on='sourcesampleid', right_on='s_sampleid', how='left', suffixes=(None, '_source'))\n",
    "sample_map = sample_map.merge(sample_info, left_on='destsampleid', right_on='s_sampleid', how='left', suffixes=(None, '_dest'))\n",
    "sample_map.drop(columns=['s_sampleid', 's_sampleid_dest'], inplace=True)\n",
    "sample_map['map_minus_sample'] = sample_map['createdt'] - sample_map['createdt_dest']\n",
    "sample_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe6d13af-c051-4623-9072-f48c20e733b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = plt.hist(sf.collectiondt, bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0180fd51-e2b0-4159-ac07-f781332386a6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sample_types = ['DNA', 'Serum', 'Plasma', 'totRNA', 'Blood', 'LCL']\n",
    "\n",
    "colors = ['pink', 'orange', 'yellow', 'pink', 'red', 'green']\n",
    "\n",
    "for (s, st) in enumerate(sample_types):\n",
    "    x = samples.loc[samples.sampletypeid==st, 'createdt']\n",
    "    plt.hist(x, alpha=0.5, bins=100, label=str(st), color=colors[s])  #density=True, stacked=True\n",
    "x = samples.loc[~samples.sampletypeid.isin(sample_types), 'createdt']\n",
    "plt.hist(x, alpha=0.5, bins=100, label=str('Other'), color='blue')\n",
    "\n",
    "#plt.suptitle('Probability Histogram of Diamond Depths', y=1.05, size=16)\n",
    "#ax.set_xlim(50, 70); ax.set_ylim(0, 1);\n",
    "#plt.tight_layout();\n",
    "\n",
    "#_ = plt.hist(samples.createdt, bins=100)\n",
    "#plt.gca().set(title='Frequency Histogram of Diamond Depths', ylabel='Frequency')\n",
    "#plt.xlim(50,75)\n",
    "plt.legend()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24d0c0b-e3aa-4026-8a86-02db8109a5a0",
   "metadata": {},
   "source": [
    "### Sankey Diagram of Sample Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20c380aa-8e7d-4b95-b336-c0f2fa41800f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build the data structure -- source,dest,count\n",
    "\n",
    "#1: Find the bases of the tree (samples with no parents) and then group them by collection (count)\n",
    "\n",
    "data = {'studyid': STUDY}\n",
    "statement = text(\"\"\"select s.S_SAMPLEID from sapphire8.s_sample s where s.SSTUDYID='CAMP' AND S_SAMPLEID NOT IN (SELECT DESTSAMPLEID from sapphire8.s_samplemap)\"\"\")\n",
    "base_samples = pd.read_sql(statement, session.bind, params=data)\n",
    "base_samples = base_samples.merge(samples, on='s_sampleid', how='left')\n",
    "base_samples = base_samples.merge(sf, left_on='samplefamilyid', right_on='s_samplefamilyid', how='left')\n",
    "# base_samples\n",
    "\n",
    "base_samples_grouped = base_samples.groupby(['u_collectionid', 'sampletypeid']).agg({'s_sampleid': [list, 'count'],})\n",
    "base_samples_grouped\n",
    "\n",
    "# base_samples.value_counts(['u_collectionid', 'sampletypeid'])\n",
    "\n",
    "# https://plainenglish.io/blog/create-a-sankey-diagram-in-python-e09e23cb1a75\n",
    "#def query_sample_tree():\n",
    "#    data = { \"sampleid\": sid}\n",
    "#\n",
    "#    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "#    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "#    return sample_history\n",
    "#\n",
    "#shistory = query_sample_history('S-001357960')\n",
    "#shistory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5856735d-7da0-4d5c-a4f3-fd3be1bdb48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_next_level(row):\n",
    "    cid = row['u_collectionid']\n",
    "    statement = text(\"\"\"select s.S_SAMPLEID from sapphire8.s_sample s where s.SSTUDYID='CAMP' AND S_SAMPLEID NOT IN (SELECT DESTSAMPLEID from sapphire8.s_samplemap)\"\"\")\n",
    "    base_samples = pd.read_sql(statement, session.bind, params=data)\n",
    "    base_samples = base_samples.merge(samples, on='s_sampleid', how='left')\n",
    "    base_samples = base_samples.merge(sf, left_on='samplefamilyid', right_on='s_samplefamilyid', how='left')\n",
    "    # base_samples\n",
    "\n",
    "    base_samples_grouped = base_samples.groupby(['u_collectionid', 'sampletypeid']).agg({'s_sampleid': [list, 'count'],})\n",
    "    base_samples_grouped\n",
    "\n",
    "    # base_samples.value_counts(['u_collectionid', 'sampletypeid'])\n",
    "\n",
    "    # https://plainenglish.io/blog/create-a-sankey-diagram-in-python-e09e23cb1a75\n",
    "    #def query_sample_tree():\n",
    "    #    data = { \"sampleid\": sid}\n",
    "    #\n",
    "    #    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "    #    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "    #    return sample_history\n",
    "    #\n",
    "    #shistory = query_sample_history('S-001357960')\n",
    "    #shistory    \n",
    "base_samples_grouped = base_samples_grouped.apply(annotate_next_level, axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa115812-e299-4dd9-8cac-b53d64627df1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5825f826-f503-4736-9fcb-d9de9fc37704",
   "metadata": {},
   "outputs": [],
   "source": [
    "q_packages = session.query(Package).filter(Package.c.sstudyid==STUDY)\n",
    "packages = pd.read_sql(q_packages.statement, session.bind)\n",
    "packages.drop(columns=['carrier', 'carriertype', 'trackingnumber', 'createby', 'createtool', 'moddt', 'modby', 'modtool', 'securityuser', 'securitydepartment',\n",
    "                       'usersequence', 'notes', 'auditsequence', 'condition', 'senderaddressid', 'senderaddresstype', 'recipientaddressid', 'recipientaddresstype',\n",
    "                       'tracelogid', 'templateflag', 'packagetype', 'senderdepartmentid', 'expecteddt', ], inplace=True)\n",
    "packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1271e44b-c324-4f2f-ad5c-ad9c41824b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open(DATASETS_PATH) as fh:\n",
    "#    raw = json.load(fh)#\n",
    "#\n",
    "#for ds in raw['datasets']:\n",
    "#    manifest_path = ds['manifest_path']\n",
    "#    manifest_delim = ds.get('manifest_delimiter', ',')\n",
    "#    manifest_columns = ds.get('manifest_columns', None)\n",
    "#    if manifest_columns:\n",
    "#        sample_table = pd.read_csv(manifest_path, delimiter=manifest_delim, names=manifest_columns)\n",
    "#    else:\n",
    "#        sample_table = pd.read_csv(manifest_path, delimiter=manifest_delim)\n",
    "#    print(sample_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3a4667-42d0-4b03-8591-e07ccd0f445e",
   "metadata": {},
   "source": [
    "## Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "76a5748f-362e-407a-a56f-bbfeb4df7f42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1100"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redmine = redminelib.Redmine('https://chanmine.bwh.harvard.edu/', key='bc92021bd829b2c07aa94b3bc2679c639204b902', requests={'verify': False})\n",
    "trackers = pd.DataFrame(data=[dict(d) for d in redmine.tracker.all()])\n",
    "\n",
    "projects = redmine.project.all()\n",
    "projects = pd.DataFrame([dict(d) for d in projects])\n",
    "projects = projects.drop(columns=['wiki_pages', 'memberships', 'issue_categories', 'time_entries', 'versions', 'news',\n",
    "                                  'issues', 'files', 'trackers', 'enabled_modules', 'time_entry_activities', 'issue_custom_fields',\n",
    "                                  'is_public', 'inherit_members', 'status', 'created_on', 'updated_on'])\n",
    "\n",
    "study_project_id = projects.loc[projects['name'] == REDMINE_STUDY, 'id'].iloc[0]\n",
    "study_project_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c3e3483b-0b4c-4ac0-972b-9f93079f7044",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relations</th>\n",
       "      <th>time_entries</th>\n",
       "      <th>children</th>\n",
       "      <th>attachments</th>\n",
       "      <th>changesets</th>\n",
       "      <th>journals</th>\n",
       "      <th>watchers</th>\n",
       "      <th>allowed_statuses</th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>...</th>\n",
       "      <th>category</th>\n",
       "      <th>parent</th>\n",
       "      <th>project</th>\n",
       "      <th>category</th>\n",
       "      <th>status</th>\n",
       "      <th>author_name</th>\n",
       "      <th>assignee</th>\n",
       "      <th>parent</th>\n",
       "      <th>url</th>\n",
       "      <th>omic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 463, 'name': 'dna/exome'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/exome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 462, 'name': 'dna/whole_genome'}</td>\n",
       "      <td>{'id': 22862}</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/whole_genome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>22862.0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 464, 'name': 'epigenetic/methylation'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>epigenetic/methylation</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 465, 'name': 'protein/olink'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>protein/olink</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 465, 'name': 'protein/olink'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>protein/olink</td>\n",
       "      <td>New</td>\n",
       "      <td>Brian Hobbs</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>17634</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 361, 'name': 'mirnaseq'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>mirnaseq</td>\n",
       "      <td>New</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18546</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 376, 'name': 'phenotype'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///udd/recpe/clock/code/cp_032018.sas7bdat</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>None</td>\n",
       "      <td>New</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    relations time_entries children attachments changesets journals watchers  \\\n",
       "82       None         None     None        None       None     None     None   \n",
       "18       None         None     None        None       None     None     None   \n",
       "99       None         None     None        None       None     None     None   \n",
       "15       None         None     None        None       None     None     None   \n",
       "57       None         None     None        None       None     None     None   \n",
       "..        ...          ...      ...         ...        ...      ...      ...   \n",
       "305      None         None     None        None       None     None     None   \n",
       "258      None         None     None        None       None     None     None   \n",
       "259      None         None     None        None       None     None     None   \n",
       "342      None         None     None        None       None     None     None   \n",
       "297      None         None     None        None       None     None     None   \n",
       "\n",
       "    allowed_statuses     id                        project  ...  \\\n",
       "82              None  22862      {'id': 28, 'name': 'A1A'}  ...   \n",
       "18              None  24096      {'id': 28, 'name': 'A1A'}  ...   \n",
       "99              None  22080      {'id': 28, 'name': 'A1A'}  ...   \n",
       "15              None  24110      {'id': 28, 'name': 'A1A'}  ...   \n",
       "57              None  23928      {'id': 28, 'name': 'A1A'}  ...   \n",
       "..               ...    ...                            ...  ...   \n",
       "305             None  17634    {'id': 235, 'name': 'VIVA'}  ...   \n",
       "258             None  18546    {'id': 235, 'name': 'VIVA'}  ...   \n",
       "259             None  18545    {'id': 235, 'name': 'VIVA'}  ...   \n",
       "342             None  16227  {'id': 952, 'name': 'eMERGE'}  ...   \n",
       "297             None  17846  {'id': 952, 'name': 'eMERGE'}  ...   \n",
       "\n",
       "                                          category         parent project  \\\n",
       "82                {'id': 463, 'name': 'dna/exome'}            NaN     A1A   \n",
       "18         {'id': 462, 'name': 'dna/whole_genome'}  {'id': 22862}     A1A   \n",
       "99   {'id': 464, 'name': 'epigenetic/methylation'}            NaN     A1A   \n",
       "15            {'id': 465, 'name': 'protein/olink'}            NaN     A1A   \n",
       "57            {'id': 465, 'name': 'protein/olink'}            NaN     A1A   \n",
       "..                                             ...            ...     ...   \n",
       "305                {'id': 361, 'name': 'mirnaseq'}            NaN    VIVA   \n",
       "258               {'id': 376, 'name': 'phenotype'}            NaN    VIVA   \n",
       "259                                            NaN            NaN    VIVA   \n",
       "342                                            NaN            NaN  eMERGE   \n",
       "297                                            NaN            NaN  eMERGE   \n",
       "\n",
       "                   category       status  author_name     assignee   parent  \\\n",
       "82                dna/exome  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "18         dna/whole_genome  In Progress  John Ziniti  John Ziniti  22862.0   \n",
       "99   epigenetic/methylation  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "15            protein/olink  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "57            protein/olink          New  Brian Hobbs  John Ziniti      NaN   \n",
       "..                      ...          ...          ...          ...      ...   \n",
       "305                mirnaseq          New  John Ziniti  John Ziniti      NaN   \n",
       "258               phenotype  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "259                    None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "342                    None          New  John Ziniti  John Ziniti      NaN   \n",
       "297                    None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "\n",
       "                                                   url  omic  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...     ?  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...     ?  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...     ?  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...     ?  \n",
       "57                                                         ?  \n",
       "..                                                 ...   ...  \n",
       "305  file:///proj/regeps/regep00/studies/VIVA/data/...     ?  \n",
       "258    file:///udd/recpe/clock/code/cp_032018.sas7bdat     ?  \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...     ?  \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...     ?  \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...     ?  \n",
       "\n",
       "[371 rows x 36 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def set_url(row):\n",
    "    cfs = row['custom_fields']\n",
    "    for cf in cfs:\n",
    "        cf_name = cf['name']\n",
    "        if cf_name == 'URL':\n",
    "            return cf['value']\n",
    "    return None\n",
    "\n",
    "dataset_tracker_id = trackers.loc[trackers['name'] == 'Dataset', 'id'].iloc[0]\n",
    "datasets = redmine.issue.filter(tracker_id=dataset_tracker_id)\n",
    "datasets = pd.DataFrame([dict(d) for d in datasets])\n",
    "\n",
    "datasets['project_name'] = datasets.apply(lambda w: w.project['name'], axis=1)\n",
    "datasets['category_name'] = datasets.apply(lambda w: w.category['name'] if pd.notna(w.category) else None, axis=1)\n",
    "datasets['status_name'] = datasets.apply(lambda w: w.status['name'] if pd.notna(w.status) else None, axis=1)\n",
    "datasets['author_name'] = datasets.apply(lambda w: w.author['name'] if pd.notna(w.author) else None, axis=1)\n",
    "datasets['assignee'] = datasets.apply(lambda w: w.assigned_to['name'] if pd.notna(w.assigned_to) else None, axis=1)\n",
    "datasets['parent_id'] = datasets.apply(lambda w: int(w.parent['id']) if pd.notna(w.parent) else None, axis=1)\n",
    "datasets['url'] = datasets.apply(set_url, axis=1)\n",
    "\n",
    "# datasets = datasets.drop(columns=['time_entries', 'changesets', 'watchers', 'allowed_statuses', 'due_date', 'done_ratio', 'tracker', 'priority', 'start_date', 'updated_on',\n",
    "#                                 'project', 'category', 'status', 'assigned_to', 'author', 'attachments', 'journals', 'is_private',\n",
    "##                                 'estimated_hours', 'created_on', 'closed_on', 'parent', 'description', 'custom_fields', 'relations',\n",
    "#                                 'children', 'author_name'])\n",
    "datasets = datasets.sort_values(['project_name', 'category_name', 'subject'])\n",
    "datasets = datasets.rename(columns={'project_name':'project','category_name':'category','status_name':'status','parent_id':'parent'})\n",
    "\n",
    "# datasets = datasets[datasets['project'] == REDMINE_STUDY]\n",
    "\n",
    "# datasets = datasets.drop(columns=['project', 'status'])\n",
    "\n",
    "datasets['omic'] = '?'\n",
    "\n",
    "### These will need to be fixed (made unique)\n",
    "## datasets[datasets['subject'] == 'TopMed WGS Data']\n",
    "datasets['subject'] = datasets.apply(lambda row: f\"(#{row['id']}) {row['subject']}\", axis=1)\n",
    "\n",
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "3a592d12-b0ea-4e38-9edd-e88f3d3c8d5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82                      (#22862) A1A_2013_HumanCoreExome\n",
       "18                    (#24096) A1A_HumanCoreExome_TOPMed\n",
       "99                                (#22080) A1A_850K_2021\n",
       "15                         (#24110) A1A olink Proteomics\n",
       "57         (#23928) Olink Proteomic Data Freeze 20221220\n",
       "                             ...                        \n",
       "305                        (#17634) VIVA Plasma miRNASeq\n",
       "258                         (#18546) VIVA Phenotype Data\n",
       "259                         (#18545) VivaThousandQC_BMIQ\n",
       "342                         (#16227) eMERGE GWAS Datsets\n",
       "297    (#17846) emERGE Principal Components Projected...\n",
       "Name: subject, Length: 371, dtype: object"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets.subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "17b3c67c-4e9a-4a82-abf7-6105fd1bbdb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['allowed_statuses',\n",
       " 'assigned_to',\n",
       " 'assignee',\n",
       " 'attachments',\n",
       " 'author',\n",
       " 'author_name',\n",
       " 'category',\n",
       " 'category',\n",
       " 'changesets',\n",
       " 'children',\n",
       " 'closed_on',\n",
       " 'created_on',\n",
       " 'custom_fields',\n",
       " 'description',\n",
       " 'done_ratio',\n",
       " 'due_date',\n",
       " 'estimated_hours',\n",
       " 'id',\n",
       " 'is_private',\n",
       " 'journals',\n",
       " 'omic',\n",
       " 'parent',\n",
       " 'parent',\n",
       " 'priority',\n",
       " 'project',\n",
       " 'project',\n",
       " 'relations',\n",
       " 'start_date',\n",
       " 'status',\n",
       " 'status',\n",
       " 'subject',\n",
       " 'time_entries',\n",
       " 'tracker',\n",
       " 'updated_on',\n",
       " 'url',\n",
       " 'watchers']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(datasets.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b411ffa7-9a59-45c6-b1ba-ba40a44f6504",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relations</th>\n",
       "      <th>time_entries</th>\n",
       "      <th>children</th>\n",
       "      <th>attachments</th>\n",
       "      <th>changesets</th>\n",
       "      <th>journals</th>\n",
       "      <th>watchers</th>\n",
       "      <th>allowed_statuses</th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>...</th>\n",
       "      <th>parent</th>\n",
       "      <th>project</th>\n",
       "      <th>category</th>\n",
       "      <th>status</th>\n",
       "      <th>author_name</th>\n",
       "      <th>assignee</th>\n",
       "      <th>parent</th>\n",
       "      <th>url</th>\n",
       "      <th>omic</th>\n",
       "      <th>projects.code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/exome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   relations time_entries children attachments changesets journals watchers  \\\n",
       "82      None         None     None        None       None     None     None   \n",
       "\n",
       "   allowed_statuses     id                    project  ... parent project  \\\n",
       "82             None  22862  {'id': 28, 'name': 'A1A'}  ...    NaN     A1A   \n",
       "\n",
       "     category       status  author_name     assignee parent  \\\n",
       "82  dna/exome  In Progress  John Ziniti  John Ziniti    NaN   \n",
       "\n",
       "                                                  url omic  projects.code  \n",
       "82  /proj/regeps/regep00/studies/A1A/data/dna/whol...    ?             p0  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a copy of the DataFrame with the renamed columns and selected columns\n",
    "datasets = datasets.rename(columns={'subject': 'submitter_id'})\n",
    "\n",
    "# change data set infor to 'po'\n",
    "# datasets['project_id'] = 'p0'\n",
    "datasets['projects.code'] = 'p0'\n",
    "\n",
    "datasets.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d37a1aec-b146-4953-a8af-8565824fbfd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['allowed_statuses',\n",
       " 'assigned_to',\n",
       " 'assignee',\n",
       " 'attachments',\n",
       " 'author',\n",
       " 'author_name',\n",
       " 'category',\n",
       " 'category',\n",
       " 'changesets',\n",
       " 'children',\n",
       " 'closed_on',\n",
       " 'created_on',\n",
       " 'custom_fields',\n",
       " 'description',\n",
       " 'done_ratio',\n",
       " 'due_date',\n",
       " 'estimated_hours',\n",
       " 'id',\n",
       " 'is_private',\n",
       " 'journals',\n",
       " 'omic',\n",
       " 'parent',\n",
       " 'parent',\n",
       " 'priority',\n",
       " 'project',\n",
       " 'project',\n",
       " 'projects.code',\n",
       " 'relations',\n",
       " 'start_date',\n",
       " 'status',\n",
       " 'status',\n",
       " 'submitter_id',\n",
       " 'time_entries',\n",
       " 'tracker',\n",
       " 'updated_on',\n",
       " 'url',\n",
       " 'watchers']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(datasets.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3f253308-2ffa-4048-987d-124f03c4436f",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = datasets.drop(columns =['allowed_statuses',\n",
    " 'assigned_to',\n",
    " 'assignee',\n",
    " 'attachments',\n",
    " 'author',\n",
    " 'author_name',\n",
    " 'category',\n",
    " 'category',\n",
    " 'changesets',\n",
    " 'children',\n",
    " 'closed_on',\n",
    " 'created_on',\n",
    " 'custom_fields',\n",
    " 'description',\n",
    " 'done_ratio',\n",
    " 'due_date',\n",
    " 'estimated_hours',\n",
    " 'is_private',\n",
    " 'journals',\n",
    " 'omic',\n",
    " 'parent',\n",
    " 'parent',\n",
    " 'priority',\n",
    " 'relations',\n",
    " 'start_date',\n",
    " 'status',\n",
    " 'status',\n",
    " 'time_entries',\n",
    " 'tracker',\n",
    " 'updated_on',\n",
    " 'watchers'])\n",
    "   \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "c3402f16-280c-48b7-af9a-c2b2f28ef960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24096) A1A_HumanCoreExome_TOPMed</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22080) A1A_850K_2021</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24110) A1A olink Proteomics</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#23928) Olink Proteomic Data Freeze 20221220</td>\n",
       "      <td>A1A</td>\n",
       "      <td></td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>17634</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#17634) VIVA Plasma miRNASeq</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>18546</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#18546) VIVA Phenotype Data</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///udd/recpe/clock/code/cp_032018.sas7bdat</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#18545) VivaThousandQC_BMIQ</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#16227) eMERGE GWAS Datsets</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#17846) emERGE Principal Components Projected...</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                        project  \\\n",
       "82   22862      {'id': 28, 'name': 'A1A'}   \n",
       "18   24096      {'id': 28, 'name': 'A1A'}   \n",
       "99   22080      {'id': 28, 'name': 'A1A'}   \n",
       "15   24110      {'id': 28, 'name': 'A1A'}   \n",
       "57   23928      {'id': 28, 'name': 'A1A'}   \n",
       "..     ...                            ...   \n",
       "305  17634    {'id': 235, 'name': 'VIVA'}   \n",
       "258  18546    {'id': 235, 'name': 'VIVA'}   \n",
       "259  18545    {'id': 235, 'name': 'VIVA'}   \n",
       "342  16227  {'id': 952, 'name': 'eMERGE'}   \n",
       "297  17846  {'id': 952, 'name': 'eMERGE'}   \n",
       "\n",
       "                                          submitter_id project  \\\n",
       "82                    (#22862) A1A_2013_HumanCoreExome     A1A   \n",
       "18                  (#24096) A1A_HumanCoreExome_TOPMed     A1A   \n",
       "99                              (#22080) A1A_850K_2021     A1A   \n",
       "15                       (#24110) A1A olink Proteomics     A1A   \n",
       "57       (#23928) Olink Proteomic Data Freeze 20221220     A1A   \n",
       "..                                                 ...     ...   \n",
       "305                      (#17634) VIVA Plasma miRNASeq    VIVA   \n",
       "258                       (#18546) VIVA Phenotype Data    VIVA   \n",
       "259                       (#18545) VivaThousandQC_BMIQ    VIVA   \n",
       "342                       (#16227) eMERGE GWAS Datsets  eMERGE   \n",
       "297  (#17846) emERGE Principal Components Projected...  eMERGE   \n",
       "\n",
       "                                                   url projects.code  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...            p0  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...            p0  \n",
       "57                                                                p0  \n",
       "..                                                 ...           ...  \n",
       "305  file:///proj/regeps/regep00/studies/VIVA/data/...            p0  \n",
       "258    file:///udd/recpe/clock/code/cp_032018.sas7bdat            p0  \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...            p0  \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0  \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0  \n",
       "\n",
       "[371 rows x 6 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d1b80fbc-eb67-453d-829f-1fe8f111d123",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "      <th>associated_experiment</th>\n",
       "      <th>copy_numbers_identified</th>\n",
       "      <th>data_description</th>\n",
       "      <th>experimental_description</th>\n",
       "      <th>experimental_intent</th>\n",
       "      <th>indels_identified</th>\n",
       "      <th>marker_panel_description</th>\n",
       "      <th>number_experimental_group</th>\n",
       "      <th>number_samples_per_experimental_group</th>\n",
       "      <th>somatic_mutations_identified</th>\n",
       "      <th>type_of_data</th>\n",
       "      <th>type_of_sample</th>\n",
       "      <th>type_of_specimen</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                    project                      submitter_id  \\\n",
       "82  22862  {'id': 28, 'name': 'A1A'}  (#22862) A1A_2013_HumanCoreExome   \n",
       "\n",
       "   project                                                url projects.code  \\\n",
       "82     A1A  /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "\n",
       "   associated_experiment copy_numbers_identified data_description  \\\n",
       "82                  None                    None             None   \n",
       "\n",
       "   experimental_description experimental_intent indels_identified  \\\n",
       "82                     None                None              None   \n",
       "\n",
       "   marker_panel_description number_experimental_group  \\\n",
       "82                     None                      None   \n",
       "\n",
       "   number_samples_per_experimental_group somatic_mutations_identified  \\\n",
       "82                                  None                         None   \n",
       "\n",
       "   type_of_data type_of_sample type_of_specimen        type  \n",
       "82         None           None             None  experiment  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the column names to be added\n",
    "new_columns = ['associated_experiment', 'copy_numbers_identified', 'data_description', 'experimental_description',\n",
    "               'experimental_intent', 'indels_identified', 'marker_panel_description', 'number_experimental_group',\n",
    "               'number_samples_per_experimental_group', 'number_samples_per_experimental_group',\n",
    "               'somatic_mutations_identified', 'type_of_data', 'type_of_sample', 'type_of_specimen']\n",
    "\n",
    "# Add the new columns with 'N/A' as the default value\n",
    "for column in new_columns:\n",
    "    datasets[column] =  None\n",
    "    \n",
    "# Add the new column named type    \n",
    "datasets['type'] = 'experiment'\n",
    "datasets.rename(columns={'project_id':'projects.code'})\n",
    "\n",
    "# Export the CSV file\n",
    "tsv_file_path = 'experiments.tsv'\n",
    "\n",
    "# Export the DataFrame as a TSV file\n",
    "datasets.to_csv(tsv_file_path, sep='\\t', index=False)\n",
    "datasets.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ec3353d0-9386-4ae1-ba17-1241fbc22272",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['associated_experiment',\n",
       " 'copy_numbers_identified',\n",
       " 'data_description',\n",
       " 'experimental_description',\n",
       " 'experimental_intent',\n",
       " 'id',\n",
       " 'indels_identified',\n",
       " 'marker_panel_description',\n",
       " 'number_experimental_group',\n",
       " 'number_samples_per_experimental_group',\n",
       " 'project',\n",
       " 'project',\n",
       " 'projects.code',\n",
       " 'somatic_mutations_identified',\n",
       " 'submitter_id',\n",
       " 'type',\n",
       " 'type_of_data',\n",
       " 'type_of_sample',\n",
       " 'type_of_specimen',\n",
       " 'url']"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(list(datasets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "eb17ea7a-a976-4e9a-88f2-b58edfe47d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the Cases TSV file\n",
    "# one row for each *subject* in a particular experiment\n",
    "# also group by associated_experiment??\n",
    "\n",
    "#cases = datasets[['associated_experiment', 'submitter_id']]\n",
    "cases = datasets['submitter_id']\n",
    "cases = cases.drop_duplicates()  # Remove duplicate cases\n",
    "cases.to_csv('cases.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "21c41398-9d3e-4a9d-86d7-9924149514d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cases Table\n",
    "# cases_columns = ['data_description', 'experimental_description',\n",
    "#                  'experimental_intent', 'type_of_data', 'type_of_sample', 'type_of_specimen']\n",
    "# cases = datasets.groupby(['associated_experiment', 'submitter_id'])[cases_columns].first().reset_index()\n",
    "# cases.rename(columns={'parent': 'case_id'}, inplace=True)\n",
    "# cases.to_csv('cases.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "d6867432-2220-44a4-8cbe-73be037d5ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the Sample TSV file\n",
    "# one row for each sample in a particular experiment\n",
    "samples_columns = ['submitter_id', 'associated_experiment', 'type_of_sample', 'type_of_specimen']\n",
    "samples = datasets[datasets['type_of_sample'].notnull()][samples_columns].copy()\n",
    "samples.rename(columns={'parent': 'sample_id'}, inplace=True)\n",
    "samples.to_csv('samples.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e7e2679a-623b-46ac-ab30-56a33852dce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export Aliquots (same as samples but different columns)\n",
    "aliquots = samples.rename(columns={'submitter_id': 'sample_id', 'type_of_sample': 'sample_type', 'type_of_specimen': 'specimen_type'})\n",
    "aliquots.to_csv('aliquots.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "0d3abb97-20f1-4b51-bdd9-00a40bfeb4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the Aliquotes TSV file\n",
    "# # same as samples, but different columns\n",
    "# aliquots_columns = ['submitter_id',  'associated_experiment', 'copy_numbers_identified', 'indels_identified',\n",
    "#                     'marker_panel_description', 'number_experimental_group', 'number_samples_per_experimental_group',\n",
    "#                     'somatic_mutations_identified']\n",
    "# aliquots = datasets[datasets['type_of_sample'].notnull()][aliquots_columns].copy()\n",
    "# aliquots.rename(columns={'parent': 'aliquot_id'}, inplace=True)\n",
    "# aliquots.to_csv('aliquots.tsv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dbc2ef3-b70a-4073-9a7c-cf3fc4f87d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(f'{len(datasets)=}')\n",
    "#datasets.value_counts('category_name', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "079bc12b-4e3a-46df-aca1-e394b4158c9c",
   "metadata": {},
   "source": [
    "### Annotated Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943032ea-6cee-40eb-8f5d-0568f9b9091f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "def annotate(ds):\n",
    "    dsid = ds['id']\n",
    "   # dscat = ds[\"category\"]\n",
    "    dsurl = ds[\"url\"]\n",
    "    if pd.notna(dsurl):\n",
    "        dsurl = dsurl.replace('file://', '')\n",
    "    dsname = ds[\"submitter_id\"]\n",
    "    \n",
    "    ds['redmine_url'] = f'https://chanmine.bwh.harvard.edu/issues/{dsid}'\n",
    "    \n",
    "    # new add stuff: I am trying to use it to store the paths and using the path function to subject mainfest files\n",
    "    # new add stuff:trying to find whether it exits, if the file exits, record S. If not, create a new file\n",
    "    try:\n",
    "        sample_manifest_path = Path(dsurl) / '.chammps/sample-manifest.csv'\n",
    "        ds_samples = pd.read_csv(sample_manifest_path)\n",
    "        S = len(ds_samples)\n",
    "    except:\n",
    "        sample_manifest_path = None\n",
    "        ds_samples = pd.DataFrame({'S_SAMPLEID':[]}) \n",
    "        S = 0            \n",
    "\n",
    "    try:\n",
    "        subject_manifest_path = Path(dsurl) / '.chammps/subject-manifest.csv'\n",
    "        ds_subjects = pd.read_csv(subject_manifest_path)\n",
    "        N = len(ds_subjects)\n",
    "    except:\n",
    "        subject_manifest_path = None\n",
    "        ds_subjects = pd.DataFrame()  # Create an empty DataFrame if subject manifest doesn't exist\n",
    "        N = 0\n",
    "        \n",
    "    # new add stuff: Perform merge with metadata from LIMS\n",
    "    # I assumed that we can create a file lims_metadata, which would contains the info can be combined\n",
    "    # not sure about combined base on 'submitter_id\" or other, but it should be unique \n",
    "    # print(f'{ds_samples}')\n",
    "    if 'S_SAMPLEID' in ds_samples.columns:\n",
    "        ds_samples = ds_samples.merge(samples, left_on='S_SAMPLEID', right_on='s_sampleid', how='left')\n",
    "    subject_batches.append(ds_samples)\n",
    "    \n",
    "    # Generate a set of participants across all experiments\n",
    "    # participants = set(merged_data['submitter_id'])\n",
    "\n",
    "    ds['S'] = S\n",
    "    ds['N'] = N\n",
    "    ds['sample-manifest'] = sample_manifest_path\n",
    "    #ds['participants'] = participants\n",
    "    # S = 0\n",
    "    # try:\n",
    "    #     smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "    #     # print(f'{smanifest=}')\n",
    "    #     exists = smanifest.exists()\n",
    "    # except:\n",
    "    #     exists = False\n",
    "    #     smanifest = None\n",
    "    # if exists:\n",
    "    #     ds_samples = pd.read_csv(smanifest)\n",
    "    #     S = len(ds_samples)\n",
    "    # else:\n",
    "    #     #print(f'NEED-MANIFEST: {smanifest=}')\n",
    "    #     pass\n",
    "    # ds['S'] = S\n",
    "\n",
    "    # N = 0\n",
    "    # try:\n",
    "    #     sbmanifest = Path(dsurl)/'.chammps/subject-manifest.csv'\n",
    "    #     exists = sbmanifest.exists()\n",
    "    # except:\n",
    "    #     exists = False\n",
    "    # if exists:\n",
    "    #     ds_subjects = pd.read_csv(sbmanifest)\n",
    "    #     N = len(ds_subjects)\n",
    "    # ds['N'] = N\n",
    "    # ds['sample-manifest'] = smanifest\n",
    "    # return(ds)\n",
    "\n",
    "subject_batches = []    \n",
    "datasets = datasets.apply(annotate, axis=1)\n",
    "if subject_batches:\n",
    "    subjects = pd.concat(subject_batches)\n",
    "else:\n",
    "    subjects = pd.DataFrame() ## FIXME: This needs \"fake\" columns\n",
    "\n",
    "#datasets = datasets.sort_values(['category',])\n",
    "# new add stuff: Save the DataFrame as a TSV file\n",
    "output_file = '../metadata/experiments.tsv'\n",
    "datasets.to_csv(output_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c34aa1a9-6b28-4da3-9ec0-14338733dba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# original code\n",
    "#def annotate(ds):\n",
    "#     dsid = ds['id']\n",
    "#    # dscat = ds[\"category\"]\n",
    "#     dsurl = ds[\"url\"]\n",
    "#     if pd.notna(dsurl):\n",
    "#         dsurl = dsurl.replace('file://', '')\n",
    "#     dsname = ds[\"submitter_id\"]\n",
    "    \n",
    "#     ds['redmine_url'] = f'https://chanmine.bwh.harvard.edu/issues/{dsid}'\n",
    "\n",
    "#     S = 0\n",
    "#     try:\n",
    "#         smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "#         # print(f'{smanifest=}')\n",
    "#         exists = smanifest.exists()\n",
    "#     except:\n",
    "#         exists = False\n",
    "#         smanifest = None\n",
    "#     if exists:\n",
    "#         ds_samples = pd.read_csv(smanifest)\n",
    "#         S = len(ds_samples)\n",
    "#     else:\n",
    "#         #print(f'NEED-MANIFEST: {smanifest=}')\n",
    "#         pass\n",
    "#     ds['S'] = S\n",
    "\n",
    "#     N = 0\n",
    "#     try:\n",
    "#         sbmanifest = Path(dsurl)/'.chammps/subject-manifest.csv'\n",
    "#         exists = sbmanifest.exists()\n",
    "#     except:\n",
    "#         exists = False\n",
    "#     if exists:\n",
    "#         ds_subjects = pd.read_csv(sbmanifest)\n",
    "#         N = len(ds_subjects)\n",
    "#     ds['N'] = N\n",
    "#     ds['sample-manifest'] = smanifest\n",
    "#     return(ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff73c6f4-4f5e-4486-bfff-4b06e35baa7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2db042a8-d8f2-497d-a0f2-86d102ce774c",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = datasets.drop(columns=['project', 'project', 'id', 'url', 'N', 'S', 'sample-manifest', 'redmine_url'])\n",
    "#datasets = datasets.rename(columns={'project_id':'projects.code'})\n",
    "datasets.to_csv('output.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0343a881-a9c9-4e3e-b3fb-0cbddf6437af",
   "metadata": {},
   "outputs": [],
   "source": [
    "subjects"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbd7677e-8961-48ad-8a9b-4b292d7cafe0",
   "metadata": {},
   "source": [
    "*type\tproject_id\t*submitter_id : S_follow 8 digits\t*cases.submitter_id\tdiagnoses.submitter_id\tbiospecimen_anatomic_site\tcomposition\tcurrent_weight\tdays_to_collection\tdays_to_sample_procurement\tdiagnosis_pathologically_confirmed\tfreezing_method\tinitial_weight\tintermediate_dimension\tis_ffpe\tlongest_dimension\tmethod_of_sample_procurement\toct_embedded\tpreservation_method\tsample_type\tsample_type_id\tsample_volume\tshortest_dimension\ttime_between_clamping_and_freezing\ttime_between_excision_and_freezing\ttissue_type\ttumor_code\ttumor_code_id\ttumor_descriptor\n",
    "sample"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6851d6bd-dc7f-43f4-85bf-73e1f86192b2",
   "metadata": {},
   "source": [
    "## Dataset Focus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d93009ec-1c47-4f90-8cca-fd63cb959402",
   "metadata": {},
   "source": [
    "### Storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f630c086-377a-4d91-ac14-78334fe92491",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Copy a Dataset id from the table above, and then execute this cell and those below to see information about it\n",
    "DATASET_ID = 20642\n",
    "\n",
    "ds = datasets.loc[datasets['id']==DATASET_ID]\n",
    "dsurl = ds.iloc[0]['url'].replace('file://', '')\n",
    "\n",
    "storage_manifests = sorted(glob.glob(str(Path(dsurl)/'results/rip/*/*-file-table-*.csv')))\n",
    "\n",
    "if storage_manifests:\n",
    "    print(f'{storage_manifests[-1]=}')\n",
    "    storage = pd.read_csv(storage_manifests[-1])\n",
    "else:\n",
    "    storage = pd.DataFrame()\n",
    "    print(f'NO-RIP-FILE-TABLES-PLEASE-RIP: {dsurl=}, going forward {storage=}')\n",
    "storage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4e8159-efb2-46c1-930f-c3d0b9f23c49",
   "metadata": {},
   "source": [
    "### Sample History"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5d7ee6-3e6d-4352-965e-089a77841f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'{SampleMap=}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b8f1c07-101a-4add-b2c7-7416a68ee74e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "def query_sample_history(row):\n",
    "    if 'S_SAMPLEID' in row:\n",
    "        destsampleid = sid = row['S_SAMPLEID']\n",
    "    else:\n",
    "        #print(f'{row=}')\n",
    "        aliasid = row['TOR_ID']\n",
    "        q_alias = session.query(Alias).filter(Alias.c.aliasid==aliasid)\n",
    "        aliases = pd.read_sql(q_alias.statement, session.bind)\n",
    "        destsampleid = sid = aliases.iloc[0]['keyid1']\n",
    "    sinfo = samples.loc[samples['s_sampleid'] == sid]\n",
    "    sample_history = [sid,]\n",
    "    sample_history_timestamps = [sinfo.iloc[0]['createdt']]\n",
    "    sample_history_types = [sinfo.iloc[0]['sampletypeid']]\n",
    "    while pd.notna(destsampleid):\n",
    "        q_sm = session.query(SampleMap).filter(SampleMap.c.destsampleid==destsampleid)\n",
    "        sm = pd.read_sql(q_sm.statement, session.bind)\n",
    "        # print(f'{sm=}')\n",
    "        if sm.empty:\n",
    "            destsampleid = None\n",
    "            break\n",
    "        else:\n",
    "            #if len(sm) > 1:\n",
    "            #    print(f'OUCH! Pooled? {sm=}')\n",
    "            sinfo = samples.loc[samples['s_sampleid'] == sm.iloc[0]['sourcesampleid']]\n",
    "            sample_history.append(sinfo.iloc[0]['s_sampleid'])\n",
    "            sample_history_types.append(sinfo.iloc[0]['sampletypeid'])\n",
    "            sample_history_timestamps.append(sinfo.iloc[0]['createdt'])\n",
    "            destsampleid = sm.iloc[0]['sourcesampleid']\n",
    "\n",
    "    row['sample_history_length'] = len(sample_history)\n",
    "    row['sample_history'] = tuple(sample_history)\n",
    "    row['sample_history_types'] = tuple(sample_history_types)\n",
    "    row['sample_history_timestamps'] = tuple(sample_history_timestamps)\n",
    "    return row\n",
    "\"\"\"\n",
    "\n",
    "def query_sample_history(sid):\n",
    "    data = { \"sampleid\": sid}\n",
    "\n",
    "    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "    return sample_history\n",
    "\n",
    "shistory = query_sample_history('S-001357960')\n",
    "shistory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d8c100a-1642-4e4e-afbe-d03ab971a686",
   "metadata": {},
   "source": [
    "### Example of plotting time interval data in Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e271bcc3-3663-4a13-95fd-7980073f3ef4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/50869690/how-to-plot-time-interval-data-in-python\n",
    "\n",
    "#df = pd.DataFrame({'Eng':['eng1','eng1','eng1','eng2','eng3','eng4','eng4','eng4','eng5','eng5'],\n",
    "#'Install date':['3/6/2010','10/25/2010','1/1/2014','1/1/2011','1/1/2014','1/1/2015','1/5/2016','5/8/2017','2/3/2010','1/1/2011'],\n",
    "#'Car':['car1','car1','car2','car3','car1','car1','car3','car2','car1','car4']})\n",
    "#df['Install date'] = pd.to_datetime(df['Install date'])\n",
    "#for car in df.Car.unique():\n",
    "#    most_recent_eng = df.loc[df.Car == car].groupby('Install date').max().tail(1).Eng.iloc[0]\n",
    "#    print(f'{car=},{most_recent_eng=}')\n",
    "#    new_df = df.loc[df.Car == car].append(pd.DataFrame({'Car':[car],'Eng':[most_recent_eng],'Install date':[pd.to_datetime('now')]}), sort=False)\n",
    "#    print(f'{new_df=}')\n",
    "#    pt = new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count')\n",
    "#    print(f'{pt=}')\n",
    "#    rs = pt.resample('1d')\n",
    "#    print(f'{rs=}')\n",
    "#    ff = rs.ffill()\n",
    "#    print(f'{ff=}')\n",
    "#    new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count').resample('1d').ffill().plot(kind='line',linewidth=10, title=car)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f5b65b4-1496-4797-9887-745c4a26b530",
   "metadata": {},
   "source": [
    "## Specific Datasets, With Sample History"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5c4aa68-28e6-4124-b81d-0e893a742453",
   "metadata": {},
   "source": [
    "### CAMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1888882c-9e6c-4a71-aa76-44b7852fbdc2",
   "metadata": {},
   "source": [
    "#### CAMP TOPMed WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f282a3b0-590a-4305-8ca2-3e7c866ab3c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TopMed WGS Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "# dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d9fdd6-77a2-4924-8c44-10dd8184b674",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (s, sample) in dataset_samples.iterrows():\n",
    "    sample_history = query_sample_history(sid)\n",
    "    sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "    sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "    sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "    most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "    df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "    df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "    df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "    df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "    pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "    pt = pt.resample('1d')\n",
    "    ff = pt.ffill()\n",
    "    _ = ff.plot(kind='line', linewidth=5, title=sid, marker='', figsize=(5,3))\n",
    "    if s > 20: break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6231b780-eedd-4d9b-8db9-eb3af6db612b",
   "metadata": {},
   "source": [
    "### LTRC TOPMed WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60e9230e-7ab8-4c75-aad8-903bb594ebf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC TOPMed WGS'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0afedca9-37ec-432f-bd05-31d0b639c088",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset_samples.value_counts('sample_history_types')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de0cfb1-1de9-4941-abdf-a67499841d2f",
   "metadata": {},
   "source": [
    "### TOPMed Blood Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb00a2b-cb45-4953-a0c3-9b574cd8ffe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TOPMed Blood Methylation Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da07c5f5-a6e0-415e-b531-1628c1d94e9a",
   "metadata": {},
   "source": [
    "### TOPMed Lung Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d99e44-063c-42b5-a19c-7f04793a5726",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TOPMed Lung Methylation Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1758f7b-e6f8-4b4e-89b9-640859596a59",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for (s, sample) in dataset_samples.iterrows():\n",
    "#    sample_df = datset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "#    print(f'{sample_df=}')\n",
    "#    sample_df sample_df['Job position'].apply(pd.Series).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')#\n",
    "\n",
    "#    most_recent_type = sample_df['sample_history_types'][-1]\n",
    "#    print(f'{sample=},{most_recent_type=}')\n",
    "#    new_df = df.loc[df.Car == car].append(pd.DataFrame({'Car':[car],'Eng':[most_recent_eng],'Install date':[pd.to_datetime('now')]}), sort=False)\n",
    "#    print(f'{new_df=}')\n",
    "#    pt = new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count')\n",
    "#    print(f'{pt=}')\n",
    "#    rs = pt.resample('1d')\n",
    "#    print(f'{rs=}')\n",
    "#    ff = rs.ffill()\n",
    "#    print(f'{ff=}')\n",
    "#    new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count').resample('1d').ffill().plot(kind='line',linewidth=10, title=car)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1e5b42-0c37-4a4e-aa3d-a5ec272b64e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 0\n",
    "sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "\n",
    "rs = pt.resample('1d')\n",
    "\n",
    "ff = rs.ffill()\n",
    "ff\n",
    "_ = ff.plot(kind='line', linewidth=8, title=sid, marker='', figsize=(5,3))\n",
    "#_ = plt.plot(ff, linewidth=2) #, title=sid, marker='')\n",
    "#_ = plt.plot(ff, linewidth=2) #, title=sid, marker='')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f695dcd6-9a69-4b93-943d-2260bab8e1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (s, sample) in dataset_samples.iterrows():\n",
    "    sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "    sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "    sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "    most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "    df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "    df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "    df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "    df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "    pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "\n",
    "    rs = pt.resample('1d')\n",
    "\n",
    "    ff = rs.ffill()\n",
    "    _ = ff.plot(kind='line', linewidth=8, title=sid, marker='', figsize=(5,3))\n",
    "    if s > 20: break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56b0b424-ed04-4bdb-8a8d-426f17c674df",
   "metadata": {},
   "source": [
    "### LTRC Phenotype Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f2f4b7-cb3e-48d7-87bb-d25eaffde0b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC Phenotype Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e79e441-e2bc-4438-b552-811faaef9c16",
   "metadata": {},
   "source": [
    "### LTRC RNASeq Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742dad9e-8aa9-478c-8fab-2dc9d6291e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC RNASeq Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ec022e3-7cdc-4968-85d8-6d2cf16499f8",
   "metadata": {},
   "source": [
    "## Upset Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808c9d96-e7f2-48b7-986b-71aeed7d257c",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASETS_FOR_UPSET_PLOT = ['LTRC TOPMed WGS', 'TOPMed Blood Methylation Data', 'TOPMed Lung Methylation Data', 'LTRC Phenotype Data', 'LTRC RNASeq Data']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a550e46-a190-4203-a5c2-b967820f9a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_for_upset = datasets[datasets['subject'].isin(DATASETS_FOR_UPSET_PLOT)]\n",
    "datasets_for_upset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257ce980-772f-496f-8ea9-130df1d2e4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = []\n",
    "for (d, dataset) in datasets_for_upset.iterrows():\n",
    "    smanifest = dataset['sample-manifest']\n",
    "    df = pd.read_csv(smanifest)\n",
    "    dataframes.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb80bafe-7243-489c-9e78-ae4d867fc258",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = dataframes[0]\n",
    "for df in dataframes[1:]:\n",
    "    merged = merged.merge(df, on='patid', how='left')\n",
    "\n",
    "#merged = phenotype.merge(rna, on='patid', how='left')\n",
    "#merged = merged.merge(lung_methylation, on='patid', how='left')\n",
    "#merged = merged.merge(blood_methylation, on='patid', how='left')\n",
    "#merged = merged.merge(wgs, on='patid', how='left')\n",
    "#merged = merged.merge(smoking, on='patid', how='left')\n",
    "merged = merged.fillna(False)\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71e034a4-0fe3-4769-a63b-aab3c8346fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "indicators = merged[['rna','lung_methylation', 'blood_methylation', 'wgs']]\n",
    "participants_by_omics = upsetplot.from_indicators(indicators, data=merged)\n",
    "#upsetplot.UpSet(participants_by_omics, orientation='vertical', show_counts=True)\n",
    "upset = upsetplot.UpSet(participants_by_omics, show_counts=True, sort_by='cardinality', sort_categories_by=None, intersection_plot_elements=0)\n",
    "upset.add_stacked_bars(by=\"smoking_status\", title=\"smoking\", elements=5)\n",
    "upset.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf56052-eb17-444c-bd12-3c9d08f4c7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "indicators = merged[['rna', 'lung_methylation', 'wgs']]\n",
    "participants_by_omics = upsetplot.from_indicators(indicators, data=merged)\n",
    "#upsetplot.UpSet(participants_by_omics, orientation='vertical', show_counts=True)\n",
    "upset = upsetplot.UpSet(participants_by_omics, show_counts=True, sort_by='cardinality', sort_categories_by=None, intersection_plot_elements=0)\n",
    "#upset.add_catplot(value='smoking_status', kind='bar', color='blue')\n",
    "upset.add_stacked_bars(by=\"smoking_status\", title=\"smoking\", elements=5)\n",
    "upset.plot()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cdnm-jupyter-python-3.7.6",
   "language": "python",
   "name": "cdnm-jupyter-python-3.7.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
