{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2e2b9e3-17a8-4cdb-beb2-d04b5e664682",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "STUDY = 'ESCAPE'\n",
    "REDMINE_STUDY = 'ESCAPE'\n",
    "\n",
    "#SPECIAL_SUBJECT_ID = 'patid'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19b945af-1835-4d01-aec3-bf6fdb999884",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "import redminelib\n",
    "import requests\n",
    "\n",
    "from urllib3.exceptions import InsecureRequestWarning\n",
    "requests.packages.urllib3.disable_warnings(category=InsecureRequestWarning)\n",
    "\n",
    "from sqlalchemy import or_\n",
    "from sqlalchemy import create_engine, MetaData, Table\n",
    "from sqlalchemy.orm import sessionmaker, aliased\n",
    "from sqlalchemy import text\n",
    "\n",
    "DBURL = 'oracle://sapphire_r:sapphire_r0@oar1.bwh.harvard.edu:1521/labvprd'\n",
    "\n",
    "sapphire8_md = MetaData(schema='SAPPHIRE8')\n",
    "engine = create_engine(DBURL)\n",
    "Session = sessionmaker(bind=engine)\n",
    "session = Session()\n",
    "\n",
    "Alias = Table('SDIALIAS',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Batch = Table('S_BATCH', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "Cohort = Table('U_COHORT',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Collection = Table('U_COLLECTION',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Package = Table('S_PACKAGE',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Sample = Table('S_SAMPLE',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "SampleFamily = Table('S_SAMPLEFAMILY',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "SampleMap = Table('S_SAMPLEMAP', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "SpecimenType = Table('U_SPECIMENTYPE', sapphire8_md, autoload=True, autoload_with=engine)\n",
    "Study = Table('S_STUDY',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Subject = Table('S_SUBJECT',sapphire8_md,autoload=True,autoload_with=engine)\n",
    "Tissue = Table('S_TISSUE', sapphire8_md, autoload=True, autoload_with=engine)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a1e5a6-7db2-45a0-9b6e-dc88a8d1cd60",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "534c0d28-c12e-49a1-be43-8844e1d6eb8b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_studyid</th>\n",
       "      <th>studydesc</th>\n",
       "      <th>studystatus</th>\n",
       "      <th>studytype</th>\n",
       "      <th>activeflag</th>\n",
       "      <th>defaultglpflag</th>\n",
       "      <th>studyalias</th>\n",
       "      <th>u_studydesign</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>ESCAPE clinical trial for lung cancer screenin...</td>\n",
       "      <td>Active</td>\n",
       "      <td>Human Genetics</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  s_studyid                                          studydesc studystatus  \\\n",
       "0    ESCAPE  ESCAPE clinical trial for lung cancer screenin...      Active   \n",
       "\n",
       "        studytype activeflag defaultglpflag studyalias u_studydesign  \n",
       "0  Human Genetics          Y              Y     ESCAPE          None  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_study = session.query(Study).filter(Study.c.s_studyid == STUDY)\n",
    "study = pd.read_sql(q_study.statement, session.bind)\n",
    "study = study.drop(columns=['estimatedsamples', 'usersequence', 'notes','collectionstartdt','collectionenddt','auditsequence',\n",
    "                            'primaryaffection', 'tracelogid', 'createdt', 'createby','createtool', 'moddt', 'completeddt','cancelledby','cancelleddt',\n",
    "                            'u_verifiedby','u_numberofsites','u_legacyid','securityuser','securitydepartment',\n",
    "                            'modby','modtool','projectid','proposedstartdt','proposedenddt','startdt','clinicalflag','plannedsites','plannedparticipants','completedby',\n",
    "                            'enddt','studyfocus','templateflag','subjectrequiredflag','priority','collectinforequiredflag','protocolname','hipaaflag',\n",
    "                            'departmentid','verifiedby','conservativecocflag','conservativerestrictionsflag',\n",
    "                            'verifiedbyrole'])\n",
    "study"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53dfd030-3e1b-4bf9-853c-627f346db2ce",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Tissue\n",
    "\n",
    "This is not a primary table, but can be merged to see additional details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8fcd94c4-f68a-4e5a-b518-6f1905769ccd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_tissueid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TS-00000001</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TS-00000002</td>\n",
       "      <td>CD4 + TCell - Unstim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TS-00000003</td>\n",
       "      <td>CD4 + TCell - Stim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TS-00000004</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TS-00000009</td>\n",
       "      <td>Dorsal Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>TS-00000218</td>\n",
       "      <td>Colo205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TS-00000219</td>\n",
       "      <td>H1299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TS-00000229</td>\n",
       "      <td>U937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TS-00000231</td>\n",
       "      <td>Adherent Cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TS-00000203</td>\n",
       "      <td>Beas-2B Cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     s_tissueid            tissuedesc\n",
       "0   TS-00000001                 Blood\n",
       "1   TS-00000002  CD4 + TCell - Unstim\n",
       "2   TS-00000003    CD4 + TCell - Stim\n",
       "3   TS-00000004                  Lung\n",
       "4   TS-00000009           Dorsal Lung\n",
       "..          ...                   ...\n",
       "58  TS-00000218               Colo205\n",
       "59  TS-00000219                 H1299\n",
       "60  TS-00000229                  U937\n",
       "61  TS-00000231        Adherent Cells\n",
       "62  TS-00000203         Beas-2B Cells\n",
       "\n",
       "[63 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_tissue = session.query(Tissue.c.s_tissueid, Tissue.c.tissuedesc)\n",
    "tissue = pd.read_sql(q_tissue.statement, session.bind)\n",
    "tissue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80415544-773c-4e09-a24c-edeb10b568b0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3135912-4bd6-448f-94e5-cd8fb9075698",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>u_cohortid</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>studyid</th>\n",
       "      <th>cohorttype</th>\n",
       "      <th>afflicted</th>\n",
       "      <th>s_subjectid</th>\n",
       "      <th>birthdt</th>\n",
       "      <th>genderflag</th>\n",
       "      <th>u_dummyflag</th>\n",
       "      <th>u_motherid</th>\n",
       "      <th>u_fatherid</th>\n",
       "      <th>u_pedigreeid</th>\n",
       "      <th>u_race</th>\n",
       "      <th>u_ethnicity</th>\n",
       "      <th>u_legacyid</th>\n",
       "      <th>u_fail</th>\n",
       "      <th>u_deceasedflag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CH-01031907</td>\n",
       "      <td>ST-01032389</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032389</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004010</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CH-01031908</td>\n",
       "      <td>ST-01032390</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032390</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004011</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CH-01031909</td>\n",
       "      <td>ST-01032391</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032391</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CH-01031910</td>\n",
       "      <td>ST-01032392</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>Proband</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032392</td>\n",
       "      <td>1946-07-03</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004013</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CH-01031911</td>\n",
       "      <td>ST-01032393</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>Proband</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032393</td>\n",
       "      <td>1953-05-19</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004014</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>CH-01032128</td>\n",
       "      <td>ST-01032610</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032610</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004231</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>CH-01032129</td>\n",
       "      <td>ST-01032611</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>Non-Proband</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032611</td>\n",
       "      <td>1957-10-18</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004232</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>CH-01032130</td>\n",
       "      <td>ST-01032612</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032612</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004233</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>CH-01032131</td>\n",
       "      <td>ST-01032613</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032613</td>\n",
       "      <td>NaT</td>\n",
       "      <td>U</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004234</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>CH-01032132</td>\n",
       "      <td>ST-01032614</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>Non-Proband</td>\n",
       "      <td>None</td>\n",
       "      <td>ST-01032614</td>\n",
       "      <td>2003-04-21</td>\n",
       "      <td>F</td>\n",
       "      <td>N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-100004235</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      u_cohortid    subjectid studyid   cohorttype afflicted  s_subjectid  \\\n",
       "0    CH-01031907  ST-01032389  ESCAPE         None      None  ST-01032389   \n",
       "1    CH-01031908  ST-01032390  ESCAPE         None      None  ST-01032390   \n",
       "2    CH-01031909  ST-01032391  ESCAPE         None      None  ST-01032391   \n",
       "3    CH-01031910  ST-01032392  ESCAPE      Proband      None  ST-01032392   \n",
       "4    CH-01031911  ST-01032393  ESCAPE      Proband      None  ST-01032393   \n",
       "..           ...          ...     ...          ...       ...          ...   \n",
       "245  CH-01032128  ST-01032610  ESCAPE         None      None  ST-01032610   \n",
       "246  CH-01032129  ST-01032611  ESCAPE  Non-Proband      None  ST-01032611   \n",
       "247  CH-01032130  ST-01032612  ESCAPE         None      None  ST-01032612   \n",
       "248  CH-01032131  ST-01032613  ESCAPE         None      None  ST-01032613   \n",
       "249  CH-01032132  ST-01032614  ESCAPE  Non-Proband      None  ST-01032614   \n",
       "\n",
       "       birthdt genderflag u_dummyflag u_motherid u_fatherid  u_pedigreeid  \\\n",
       "0          NaT          U        None       None       None  PG-100004010   \n",
       "1          NaT          U        None       None       None  PG-100004011   \n",
       "2          NaT          U        None       None       None  PG-100004012   \n",
       "3   1946-07-03          M           N       None       None  PG-100004013   \n",
       "4   1953-05-19          F           N       None       None  PG-100004014   \n",
       "..         ...        ...         ...        ...        ...           ...   \n",
       "245        NaT          U        None       None       None  PG-100004231   \n",
       "246 1957-10-18          M           N       None       None  PG-100004232   \n",
       "247        NaT          U        None       None       None  PG-100004233   \n",
       "248        NaT          U        None       None       None  PG-100004234   \n",
       "249 2003-04-21          F           N       None       None  PG-100004235   \n",
       "\n",
       "    u_race u_ethnicity u_legacyid u_fail u_deceasedflag  \n",
       "0     None        None       None   None           None  \n",
       "1     None        None       None   None           None  \n",
       "2     None        None       None   None           None  \n",
       "3     None        None       None   None           None  \n",
       "4     None        None       None   None           None  \n",
       "..     ...         ...        ...    ...            ...  \n",
       "245   None        None       None   None           None  \n",
       "246   None        None       None   None           None  \n",
       "247   None        None       None   None           None  \n",
       "248   None        None       None   None           None  \n",
       "249   None        None       None   None           None  \n",
       "\n",
       "[250 rows x 17 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_cohort = session.query(Cohort, Subject).filter(Cohort.c.studyid==STUDY, Subject.c.s_subjectid==Cohort.c.subjectid)\n",
    "cohort = pd.read_sql(q_cohort.statement, session.bind)\n",
    "cohort.drop(columns=['usersequence', 'auditsequence', 'tracelogid', 'templateflag', 'notes', 'createtool', 'modtool',\n",
    "                    'moddt', 'cohortdesc', 'createdt', 'createby', 'modby', 'activeflag', 'subjecttype', 'subjectdesc',\n",
    "                    'speciesid', 'strainid', 'securityuser', 'securitydepartment', 'activeflag_1', 'auditsequence_1',\n",
    "                    'createby_1', 'createdt_1', 'createtool_1', 'modby_1', 'moddt_1', 'modtool_1', 'templateflag_1',\n",
    "                     'tracelogid_1', 'usersequence_1', 'notes_1', 'u_cageid', 'u_country', 'affliction', 'severity',\n",
    "                    ], inplace=True)\n",
    "\n",
    "cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f3203d66-f5e3-43c9-87a7-cfa039f79a48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cohorttype\n",
       "NaN            172\n",
       "Non-Proband     51\n",
       "Proband         27\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('cohorttype', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e89e9b3-1ec3-444c-ba32-d60949ef69e8",
   "metadata": {},
   "source": [
    "### Distribution of Participant Date-of-Birth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d9204431-fc36-4d33-ae5c-1e5c44f13f02",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAEICAYAAAB25L6yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAbMklEQVR4nO3de3wV5b3v8c8PEgkYKAqxtWAInF0j1GCQcBNFrHartd6rchGweMrLCt6q7bG6ezbuc/C4xdbC1i2iCKgobC26ix7rtlYELIqEi4CAotxSUxUsCBSQwG//MZPlIuSyAlnJk/B9v17rxWTmmZnfs1b4ZtYzs2aZuyMiIuFq1tAFiIhI9RTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1BLrZjZTjPrUkOb3Lhd8/qqqzEzs5ZmNsfMtpvZc4ex/itmNqKW60wzs/9b231Jw1BQNzFmtsHMdsdB+amZTTWz7MPc1lwz+5/J89w9290/rm49d98Ut9t/OPutDTNzM/uHapZfZ2b74+djp5mtj5+Tk2uxj3SH2o+AbwLt3P2qSvbvZrYrrn+LmT1rZm3Ll7v7he4+vaqNx8/BgrRULvVCQd00Xezu2cDpQC/gn2qzskWa0u/Gwvj5+AZwHrAbKDazUxu2rIROwAfuXlZNm9PiPnQBjgPGprJhM8s48vKkoTWl/4xSgbv/BXgFONXMjjOzl8zsczP7WzzdsbxtfPQ8zszeAv4OPAWcBTwUH8k9FLdLHMHGb9l/bWYb47ftC+J5eXG7jKRt/z8zWxS3+08zOz5p38+Z2V/jZfPM7LtJy6aZ2cNm9rKZ7TCzd8zsf8TL5sXNlsc1XlPD87Hf3T9y9xuBN0kKu6pqMLNRwFDgF/E+5sTzv21mv4ufz/VmdnN1+zazrvHzsM3MVpnZJfH8e4D/DVwTb//6GvrwJfB7oFvSthPvfOKj57fM7EEz+wKYBUwC+sXb35a0ueMqe14lPArqJszMTgJ+ACwleq2nEh295RIdVT5UYZVhwCigNXAdMB8YEw9jjKlkFw8APYEzgOOBXwAHqihnODAS+DZQBkxMWvYK8B3gBGAJMKPCuoOBe4iOJNcB4wDcfUC8/LS4xllV7Lsys4n+EFVbg7tPjqfvj/dxcfxuYw6wHOgAnAvcambnV7YjM8uM2/9XvP2bgBlmlu/u/wzcC8yKtz+luqLN7DjgMuDtapr1AT6O93UtcAPxuwp3b5vUrtLnVcKjoG6aXoyPnBYQHTne6+5b3f137v53d99B9J/y7ArrTXP3Ve5e5u77qttBHFYjgVvc/S/x0eqf3X1vFas85e4r3X0X8CvgaotPNrr7E+6+I153LHCamX0jad3Z7r4oHhqYARTW4rmoyidEf1xIsYZkvYAcd/8Xd/8qHrN/DBhURfu+QDZwX9z+T8BLREGZqiXxa7qF6A/to9W0/cTd/y1+HXdX0y4dz6ukgYK6abrM3du6eyd3v9Hdd5tZKzN7NB6m+BKYB7S1g6/M2FyLfbQHsoCPUmyfvO2NQCbQ3syam9l9ZvZRXNeGpO2X+2vS9N+JQu9IdQC+AEixhmSdgG/Hwxjb4gC9i+iEYPmVMeWPXKJ3EZvdPfndxsa4hoOY2V1J605KWnR6fDScBTwCzDezrCrqS/V1TMfzKmmgoD563A7kA33cvQ1QPmxgSW0q3kqxulsrbgH2AKmOa56UNJ0L7Iu3MQS4lOgk3zeAvErqSofLiYZ2SKGGis/DZmB9/Mew/NHa3X8AiStjyh+biI7eT6pwgjYX+EvFotz93qR1b6hk+T7gcaAzUNXJ0Nq8jtIIKKiPHq2JxqW3xSfy/jmFdT4lusrgEPHR4RPAb+ITa83NrJ+ZtahiW9eaWTczawX8C/B8fPlea2AvsBVoRTReWxtV1lhRXGNnM/s3YCDR+Cwp1FBxH4uAL83sf8UnT5ub2alm1quKXb8D7CI6IZlpZgOBi4GZqdRdsQ/Aj4ley2ovk6xQf0czO6a2+5MwKKiPHr8FWhIdxb4N/CGFdSYAP7LoKpGJlSy/A1gBvEs0jPCvVP079RQwjejtdhZQfpXEk0TDAH8B3qf6k2SVGQtMj4cgrq6iTT8z2wl8CcwF2gC93H1FijVMAbrF+3gx/gNzMdGY7nqi5/RxoqPxQ7j7V8AlwIVx238Hhrv7mlr0c3nch78BI4DL3f2LFNf9E7AK+KuZbanFPiUQpi8OkHQzs7nA0+7+eEPXItIY6YhaRCRwCmoRkcBp6ENEJHA6ohYRCVxabtjSvn17z8vLS8emRUSapOLi4i3unlPZsrQEdV5eHosXL07HpkVEmiQz21jVMg19iIgETkEtIhI4BbWISOD07Q8iTcy+ffsoKSlhz549DV2KVCIrK4uOHTuSmZmZ8joKapEmpqSkhNatW5OXl4dZum9CKLXh7mzdupWSkhI6d+6c8noa+hBpYvbs2UO7du0U0gEyM9q1a1frdzspBbWZtTWz581sjZmtNrN+h1WliNQLhXS4Due1SXXoYwLwB3f/UXxP21a13pOIiByWGoPazMq/DeQ6SNxb96v0liUideaZOj66HlLz/YGaN29OQUEBZWVldO3alenTp9OqVWrHd9OmTWPx4sU89FDF714+ctnZ2ezcubPOt5tuqQx9dAE+B6aa2VIze9zMjq3YyMxGmdliM1v8+eef13mhR7vZa0sTD5HQtWzZkmXLlrFy5UqOOeYYJk2adNDy/fv3N1BljVMqQZ0BnA484u49iL5S6M6Kjdx9srsXuXtRTk6lH1cXkaPQWWedxbp165g7dy7nnHMOQ4YMoaCggD179vDjH/+YgoICevTowRtvvJFYZ/PmzVxwwQXk5+dzzz33JOZfdtll9OzZk+9+97tMnjw5MT87O5u7776b0047jb59+/Lpp58CsH79evr160evXr341a9+lWhfWlrKgAEDKCws5NRTT2X+/PmELJWgLgFK3P2d+OfniYJbRKRaZWVlvPLKKxQUFACwaNEixo0bx/vvv8/DDz8MwIoVK3j22WcZMWJE4mqIRYsWMWPGDJYtW8Zzzz2XuHfQE088QXFxMYsXL2bixIls3boVgF27dtG3b1+WL1/OgAEDeOyxxwC45ZZb+OlPf8q7777Lt771rURdzzzzDOeffz7Lli1j+fLlFBYW1tdTclhqDGp3/yuw2czy41nnEn2vnIhIpXbv3k1hYSFFRUXk5uZy/fXXA9C7d+/E9cMLFixg2LBhAJxyyil06tSJDz74AIDvf//7tGvXjpYtW3LFFVewYMECACZOnJg4at68eTMffvghAMcccww//OEPAejZsycbNmwA4K233mLw4MEAiX0B9OrVi6lTpzJ27FhWrFhB69at0/yMHJlUr/q4CZgRX/HxMdG3IIuIVKp8jLqiY4/9+vRWdV9aUvESNjNj7ty5/PGPf2ThwoW0atWKgQMHJo7AMzMzE+s0b96csrKyKrcFMGDAAObNm8fLL7/MsGHD+PnPf87w4cNr1cf6lNJ11O6+LB5/7u7ul7n739JdmIg0bQMGDGDGjBkAfPDBB2zatIn8/OiN+2uvvcYXX3zB7t27efHFF+nfvz/bt2/nuOOOo1WrVqxZs4a33675C+v79+/PzJkzARL7Ati4cSMnnHACP/nJT7j++utZsmRJGnpYd/QRcpGmLoXL6RrCjTfeyA033EBBQQEZGRlMmzaNFi1aAHDmmWcybNgw1q1bx5AhQygqKqKgoIBJkybRvXt38vPz6du3b437mDBhAkOGDGHChAlceeWViflz585l/PjxZGZmkp2dzZNPPpm2ftaFtHxnYlFRkeuLA+pW8mV5V+Sf2ICVSOhWr15N165dG7oMqUZlr5GZFbt7UWXtda8PEZHAKahFRAKnoBYRCZyCWkQkcApqEZHAKahFRAKn66hFmri6vuNiKpeHmhk/+9nP+PWvfw3AAw88wM6dOxk7duwR73/s2LE89thj5OTkUFZWxr333ssll1yS8vp5eXksXryY9u3bH3EtFevKzs7mjjvuqNPtgo6oRSQNWrRowezZs9myZUtatn/bbbclbtg0cuRIDhw4cNDy5I+QNwUKahGpcxkZGYwaNYoHH3zwkGUbN27k3HPPpXv37px77rls2rQJgOuuu46bb76ZM844gy5duvD888/XuJ+uXbuSkZHBli1bGDhwIHfddRdnn302EyZM4PXXX6dHjx4UFBQwcuRI9u7dm1hv/Pjx9O7dm969e7Nu3ToA5syZQ58+fejRowfnnXde4lapY8eOZeTIkQwcOJAuXbowceLExHbGjRtHfn4+5513HmvXrk3MnzhxIt26daN79+4MGjTo8J7EJApqEUmL0aNHM2PGDLZv337Q/DFjxjB8+HDee+89hg4dys0335xYVlpayoIFC3jppZe4885Dbnt/iHfeeYdmzZpRfg/8bdu28eabbzJ69Giuu+46Zs2axYoVKygrK+ORRx5JrNemTRsWLVrEmDFjuPXWW4HoY+tvv/02S5cuZdCgQdx///2J9mvWrOHVV19l0aJF3HPPPezbt4/i4mJmzpzJ0qVLmT17Nu+++26i/X333cfSpUt57733DvnShMOhoBaRtGjTpg3Dhw8/6AgUYOHChQwZMgSIbj1afgtTiL4YoFmzZnTr1i1xRFuZBx98kMLCQu644w5mzZqVuEPeNddcA8DatWvp3LkzJ598MgAjRoxg3rx5ifXLb306ePBgFi5cCEBJSQnnn38+BQUFjB8/nlWrViXaX3TRRbRo0YL27dtzwgkn8OmnnzJ//nwuv/xyWrVqRZs2bQ4aJ+/evTtDhw7l6aefJiPjyE8FKqhFJG1uvfVWpkyZwq5du6psk3wb0vKbMsHXt0G9++67KSwsPOjm/uVj1PPnz+ess85KzC+/jWpN9zBK3mf59E033cSYMWNYsWIFjz76aOIWqhXrSr6NalXfKP7yyy8zevRoiouL6dmz5xGPmSuoRSRtjj/+eK6++mqmTJmSmHfGGWccdOvRM888s9ptjBs3jmXLllV6f+uqnHLKKWzYsCEx/vzUU09x9tlnJ5bPmjUr8W+/fv0A2L59Ox06dABg+vTpNe5jwIABvPDCC+zevZsdO3YwZ84cAA4cOMDmzZs555xzuP/++9m2bdsRf6GuLs8TaeIa+m6Lt99++0HfKD5x4kRGjhzJ+PHjycnJYerUqXW+z6ysLKZOncpVV11FWVkZvXr14oYbbkgs37t3L3369OHAgQM8++yzQHTS8KqrrqJDhw707duX9evXV7uP008/nWuuuYbCwkI6deqUOLLfv38/1157Ldu3b8fdue2222jbtu0R9Ue3OW0kdJtTSZVucxo+3eZURKSJUVCLiAROQS3SBKVjSFPqxuG8NgpqkSYmKyuLrVu3KqwD5O5s3bqVrKysWq2nqz5EmpiOHTtSUlLC559/3tClSCWysrLo2LFjrdZRUIs0MZmZmXTu3Lmhy5A6pKEPEZHAKahFRAKX0tCHmW0AdgD7gbKqLsoWEZG6V5sx6nPcPT13ARcRkSpp6ENEJHCpBrUD/2VmxWY2qrIGZjbKzBab2WJdFiQiUndSDer+7n46cCEw2swGVGzg7pPdvcjdi8q/bUFERI5cSkHt7p/E/34GvAD0TmdRIiLytRqD2syONbPW5dPAPwIr012YiIhEUrnq45vAC/FXzmQAz7j7H9JalYiIJNQY1O7+MXBaPdQiIiKV0OV5IiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAQu5aA2s+ZmttTMXkpnQSIicrDaHFHfAqxOVyEiIlK5lILazDoCFwGPp7ccERGpKCPFdr8FfgG0rqqBmY0CRgHk5uYecWFStdlrSxPTV+SfGOw2RaRu1HhEbWY/BD5z9+Lq2rn7ZHcvcveinJycOitQRORol8rQR3/gEjPbAMwEvmdmT6e1KhERSagxqN39l+7e0d3zgEHAn9z92rRXJiIigK6jFhEJXqonEwFw97nA3LRUIiIildIRtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4BTUIiKBU1CLiAROQS0iEjgFtYhI4GoMajPLMrNFZrbczFaZ2T31UZiIiEQyUmizF/ieu+80s0xggZm94u5vp7k2EREhhaB2dwd2xj9mxg9PZ1EiIvK1VI6oMbPmQDHwD8DD7v5OJW1GAaMAcnNz67LGejd7bWli+or8Exvl/pK3WZXa7qu+nxcRiaR0MtHd97t7IdAR6G1mp1bSZrK7F7l7UU5OTh2XKSJy9KrVVR/uvg2YC1yQjmJERORQqVz1kWNmbePplsB5wJo01yUiIrFUxqhPBKbH49TNgP9w95fSW5aIiJRL5aqP94Ae9VCLiIhUQp9MFBEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwCmoRkcApqEVEAqegFhEJnIJaRCRwNQa1mZ1kZm+Y2WozW2Vmt9RHYSIiEslIoU0ZcLu7LzGz1kCxmb3m7u+nuTYRESGFI2p3L3X3JfH0DmA10CHdhYmISCSVI+oEM8sDegDvVLJsFDAKIDc3ty5qO+rNXlva4O1r2+aK/BNrVYOI1Czlk4lmlg38DrjV3b+suNzdJ7t7kbsX5eTk1GWNIiJHtZSC2swyiUJ6hrvPTm9JIiKSLJWrPgyYAqx299+kvyQREUmWyhF1f2AY8D0zWxY/fpDmukREJFbjyUR3XwBYPdQiIiKV0CcTRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCV2NQm9kTZvaZma2sj4JERORgqRxRTwMuSHMdIiJShRqD2t3nAV/UQy0iIlKJjLrakJmNAkYB5ObmHvZ2Zq8tTUxfkX/i1wueMRjiqf+cyjaPoLaqVNz+Ift+xqL5PT857H2kXTXPpYgcrK7ypTp1djLR3Se7e5G7F+Xk5NTVZkVEjnq66kNEJHAKahGRwKVyed6zwEIg38xKzOz69JclIiLlajyZ6O6D66MQERGpnIY+REQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQmcglpEJHAKahGRwCmoRUQCp6AWEQlcSkFtZheY2VozW2dmd6a7KBER+VqNQW1mzYGHgQuBbsBgM+uW7sJERCSSyhF1b2Cdu3/s7l8BM4FL01uWiIiUy0ihTQdgc9LPJUCfio3MbBQwKv5xp5mtPfLyKhhqtfs5Ne2BLYdZ0RH4dro2XDf9ObznMh0a6PVJi6bUF1B/6lqnqhakEtSV/Y/1Q2a4TwYm16KoIJjZYncvaug66or6E66m1BdQf+pTKkMfJcBJST93BD5JTzkiIlJRKkH9LvAdM+tsZscAg4Dfp7csEREpV+PQh7uXmdkY4FWgOfCEu69Ke2X1p9EN19RA/QlXU+oLqD/1xtwPGW4WEZGA6JOJIiKBU1CLiASuyQW1mT1hZp+Z2cqkeaeZ2UIzW2Fmc8ysTTw/z8x2m9my+DEpaZ2ecft1ZjbRzBrkwuLa9Cde1j1etipentVY+2NmQ5Nem2VmdsDMChtxfzLNbHo8f7WZ/TJpncbYn2PMbGo8f7mZDUxap8H7Y2Ynmdkb8XO9ysxuiecfb2avmdmH8b/HJa3zy7jmtWZ2fjD9cfcm9QAGAKcDK5PmvQucHU+PBP5PPJ2X3K7CdhYB/YiuI38FuLAR9CcDeA84Lf65HdC8sfanwnoFwMeN/PUZAsyMp1sBG4C8Rtyf0cDUePoEoBhoFkp/gBOB0+Pp1sAHRLfBuB+4M55/J/Cv8XQ3YDnQAugMfBTK/58md0Tt7vOALyrMzgfmxdOvAVdWtw0zOxFo4+4LPXqVngQuq+NSU1LL/vwj8J67L4/X3eru+xtxf5INBp6FRv36OHCsmWUALYGvgC8bcX+6Aa/H630GbAOKQumPu5e6+5J4egewmuiT1pcC0+Nm05Nqu5ToD+led18PrAN6h9CfJhfUVVgJXBJPX8XBH+DpbGZLzexNMzsrnteB6IM+5UrieaGoqj8nA25mr5rZEjP7RTy/sfYn2TXEQU3j7c/zwC6gFNgEPODuX9B4+7McuNTMMsysM9AzXhZcf8wsD+gBvAN8091LIQpzoncDUPntMjoQQH+OlqAeCYw2s2Kit0BfxfNLgVx37wH8DHgmHn9L6WPzDaiq/mQAZwJD438vN7Nzabz9AcDM+gB/d/fycdPG2p/ewH6iG710Bm43sy403v48QRRai4HfAn8GygisP2aWDfwOuNXdv6yuaSXzvJr59SaVe300eu6+hmhYADM7Gbgonr8X2BtPF5vZR0RHpSVEH5UvF9TH5qvqD1Hdb7r7lnjZ/ycab3yaxtmfcoP4+mgaGu/rMwT4g7vvAz4zs7eAImA+jbA/7l4G3Fbezsz+DHwI/I1A+mNmmUQhPcPdZ8ezPzWzE929NB7W+CyeX9XtMhr89+2oOKI2sxPif5sB/wRMin/Oseh+28RHNt8hOmFVCuwws77x2d3hwH82SPGVqKo/RJ8e7W5mreJx0LOB9xtxf8rnXUV0e10g8Xa1MfZnE/A9ixwL9AXWNNb+xL9nx8bT3wfK3D2Y37d431OA1e7+m6RFvwdGxNMjkmr7PTDIzFrEQznfARYF0Z/6PhOb7gfRkVcpsI/oL+H1wC1EZ3w/AO7j609kXgmsIhprWwJcnLSdIqKxuY+Ah8rXCbk/cftr4z6tBO5vAv0ZCLxdyXYaXX+AbOC5+PV5H/h5I+9PHrCW6CTdH4FOIfWHaPjPia6EWhY/fkB0NdTrREf/rwPHJ61zd1zzWpKu7Gjo/ugj5CIigTsqhj5ERBozBbWISOAU1CIigVNQi4gETkEtIhI4BbWISOAU1CIigftvUToe+EJSjHkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#colors = ['pink', 'orange', 'yellow', 'pink', 'red', 'green']\n",
    "\n",
    "#for (s, st) in enumerate(sample_types):\n",
    "#    x = samples.loc[samples.sampletypeid==st, 'createdt']\n",
    "#    plt.hist(x, alpha=0.5, bins=100, label=str(st), color=colors[s])  #density=True, stacked=True\n",
    "#x = samples.loc[~samples.sampletypeid.isin(sample_types), 'createdt']\n",
    "#plt.hist(x, alpha=0.5, bins=100, label=str('Other'), color='blue')\n",
    "\n",
    "#plt.suptitle('Probability Histogram of Diamond Depths', y=1.05, size=16)\n",
    "#ax.set_xlim(50, 70); ax.set_ylim(0, 1);\n",
    "#plt.tight_layout();\n",
    "\n",
    "#_ = plt.hist(samples.createdt, bins=100)\n",
    "#plt.gca().set(title='Frequency Histogram of Diamond Depths', ylabel='Frequency')\n",
    "#plt.xlim(50,75)\n",
    "#plt.legend()\n",
    "\n",
    "study_design = study.iloc[0]['u_studydesign']\n",
    "if study_design == 'Case Control':\n",
    "    pass\n",
    "else:\n",
    "    probands = cohort.loc[cohort['cohorttype'] == 'Proband', 'birthdt']\n",
    "    non_probands = cohort.loc[cohort['cohorttype'] != 'Proband', 'birthdt']\n",
    "    _ = plt.hist(probands, bins=100, color=\"orange\", label=\"Probands\")\n",
    "    _ = plt.hist(non_probands, bins=100, color=\"lightblue\", label=\"Non-Probands\")\n",
    "    plt.title('Participant Date-of-Birth')\n",
    "    plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "913d49e9-e719-43b8-ba8c-66864526de1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genderflag\n",
       "U    174\n",
       "F     55\n",
       "M     21\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('genderflag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e615bef-4dfa-4244-934b-e4fba865b5d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u_race\n",
       "NaN    250\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('u_race', dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "978f584b-2549-48a1-8a3b-f660bb72508d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u_ethnicity\n",
       "NaN    250\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort.value_counts('u_ethnicity', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd5c505-3466-4440-b690-0cb4a9c77165",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "712e3eb0-e477-49ce-be70-080f38848c91",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>collectiondesc</th>\n",
       "      <th>createdt</th>\n",
       "      <th>createby</th>\n",
       "      <th>moddt</th>\n",
       "      <th>modby</th>\n",
       "      <th>activeflag</th>\n",
       "      <th>studyid</th>\n",
       "      <th>samplesexpected</th>\n",
       "      <th>securityuser</th>\n",
       "      <th>securitydepartment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL-AACR-0266</td>\n",
       "      <td>ESCAPE Clinical Trial</td>\n",
       "      <td>2020-10-20 17:28:48</td>\n",
       "      <td>resat</td>\n",
       "      <td>2022-02-21 11:47:14</td>\n",
       "      <td>relef</td>\n",
       "      <td>Y</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>resat</td>\n",
       "      <td>SGG LAB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  u_collectionid         collectiondesc            createdt createby  \\\n",
       "0   CL-AACR-0266  ESCAPE Clinical Trial 2020-10-20 17:28:48    resat   \n",
       "\n",
       "                moddt  modby activeflag studyid samplesexpected securityuser  \\\n",
       "0 2022-02-21 11:47:14  relef          Y  ESCAPE            None        resat   \n",
       "\n",
       "  securitydepartment  \n",
       "0            SGG LAB  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_collections = session.query(Collection).filter(Collection.c.studyid==STUDY)\n",
    "collections = pd.read_sql(q_collections.statement, session.bind)\n",
    "collections.drop(columns=['usersequence', 'auditsequence', 'tracelogid', 'timepointname', 'timepointstart', 'timepointend', 'timepointtimeunits', \n",
    "                          'visitstart',  'visitname', 'visittimeunits', 'visitend', 'alternatename', 'receivableflag', 'restrictionclassid',\n",
    "                         'protocolname', 'startdt', 'enddt', 'templateflag', 'notes', 'createtool', 'modtool'], inplace=True)\n",
    "collections"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c27b4f0-42a8-4a2d-9bb9-c7c36cb459d6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sample Families (w/Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7bd36d82-cd46-4cf9-9b8b-8a2b9f6de997",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_samplefamilyid</th>\n",
       "      <th>studysiteid</th>\n",
       "      <th>collectiondt</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>tissueid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>u_initsampletypeid</th>\n",
       "      <th>u_inittreatmenttypeid</th>\n",
       "      <th>u_initcontainertypeid</th>\n",
       "      <th>u_initenvironmentid</th>\n",
       "      <th>u_specimentypeid</th>\n",
       "      <th>u_packageid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_samplefamilyid, studysiteid, collectiondt, subjectid, tissueid, createdt, u_collectionid, u_initsampletypeid, u_inittreatmenttypeid, u_initcontainertypeid, u_initenvironmentid, u_specimentypeid, u_packageid, tissuedesc]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_sf = session.query(SampleFamily, Tissue.c.tissuedesc).filter(SampleFamily.c.sstudyid==STUDY, Tissue.c.s_tissueid==SampleFamily.c.tissueid)\n",
    "sf = pd.read_sql(q_sf.statement, session.bind)\n",
    "sf.drop(columns=['samplefamilydesc', 'initialpackageid', 'initialdepartmentid', 'collectmethodid', 'verifiedby', 'verifieddt',\n",
    "                'u_visitname', 'u_timepointname', 'approvedby', 'approveddt', 'recieveddt', 'cocflag', 'restrictionsflag', 'age',\n",
    "                'ageunits', 'restrictclassid', 'externalsubject', 'animal', 'diseaseid', 'diseaseid', 'metastasisid', 'clinicaldiagid',\n",
    "                'conditionalapprovalreason', 'conditionalapprovalflag', 'specimendefid', 'activeflag', 'auditsequence', 'clinicalevent',\n",
    " 'clinicalprotocolid',\n",
    " 'clinicalprotocolrevision',\n",
    " 'clinicalprotocolversionid',\n",
    "                 'createby',\n",
    " 'createtool',\n",
    " 'deviationdesc',\n",
    " 'eventdefid',\n",
    " 'initialmass',\n",
    " 'initialmassunits',\n",
    " 'initialvolume',\n",
    " 'initialvolumeunits',\n",
    " 'kittrackitem',\n",
    " 'modby',\n",
    " 'moddt',\n",
    " 'modtool',\n",
    " 'notes', 'tracelogid',  'usersequence', 'templateflag', 'participantid', 'u_legacyfamilyid', 'u_collectiontypid', 'sampletypeid',\n",
    "                 'participanteventid', 'sstudyid',\n",
    "                ], inplace=True)\n",
    "sf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b536b713-5e25-4c91-aa45-423aaec7c57d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: int64)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf.value_counts(['u_collectionid', 'tissuedesc'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c395c8-1ba0-4224-929b-f61e8d0a9d00",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9345767c-e1ba-49ab-ac83-0b7abb17af2d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_sampleid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>batchid</th>\n",
       "      <th>samplefamilyid</th>\n",
       "      <th>sstudyid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-001923284</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2021-05-04 08:22:09</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210504-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-001983201</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:04</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-001926215</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-06-11 14:41:26</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-001926207</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2021-06-11 12:00:18</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00002</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-001983203</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:06</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-210611-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>S-001972944</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-08-30 12:50:19</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220830-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>S-001972945</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-08-30 12:50:19</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220830-00004</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>697</th>\n",
       "      <td>S-001970980</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-07-29 11:11:37</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>S-001983234</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2022-11-09 15:56:45</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>S-001969298</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>2022-07-12 09:07:41</td>\n",
       "      <td>None</td>\n",
       "      <td>SF-220712-00003</td>\n",
       "      <td>ESCAPE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>700 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      s_sampleid sampletypeid            createdt batchid   samplefamilyid  \\\n",
       "0    S-001923284       Saliva 2021-05-04 08:22:09    None  SF-210504-00002   \n",
       "1    S-001983201          DNA 2022-11-09 15:56:04    None  SF-210611-00002   \n",
       "2    S-001926215          DNA 2021-06-11 14:41:26    None  SF-210611-00002   \n",
       "3    S-001926207       Saliva 2021-06-11 12:00:18    None  SF-210611-00002   \n",
       "4    S-001983203          DNA 2022-11-09 15:56:06    None  SF-210611-00004   \n",
       "..           ...          ...                 ...     ...              ...   \n",
       "695  S-001972944          DNA 2022-08-30 12:50:19    None  SF-220830-00004   \n",
       "696  S-001972945          DNA 2022-08-30 12:50:19    None  SF-220830-00004   \n",
       "697  S-001970980          DNA 2022-07-29 11:11:37    None  SF-220712-00003   \n",
       "698  S-001983234          DNA 2022-11-09 15:56:45    None  SF-220712-00003   \n",
       "699  S-001969298       Saliva 2022-07-12 09:07:41    None  SF-220712-00003   \n",
       "\n",
       "    sstudyid  \n",
       "0     ESCAPE  \n",
       "1     ESCAPE  \n",
       "2     ESCAPE  \n",
       "3     ESCAPE  \n",
       "4     ESCAPE  \n",
       "..       ...  \n",
       "695   ESCAPE  \n",
       "696   ESCAPE  \n",
       "697   ESCAPE  \n",
       "698   ESCAPE  \n",
       "699   ESCAPE  \n",
       "\n",
       "[700 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_samples = session.query(Sample).filter(SampleFamily.c.sstudyid==STUDY, Sample.c.samplefamilyid==SampleFamily.c.s_samplefamilyid)\n",
    "samples = pd.read_sql(q_samples.statement, session.bind)\n",
    "samples.drop(columns=['sampledesc', 'createby', 'createtool', 'submitterid', 'modby', 'submitteddt', 'modtool', 'notes', 'auditsequence', 'tracelogid',\n",
    "                     'usersequence', 'activeflag', 'allocatedforaddressid', 'allocatedforaddresstype', 'allocatedfordepartmentid', 'auditsequence', 'autofinalrptflag',\n",
    "                     'autoreceiveflag', 'batchstageid',  'cancelledby', 'cancelleddt', 'classification', 'cocflag', 'cocrequiredflag', 'collectedby', 'collectiondt',\n",
    "                     'completedt', 'concentration', 'concentrationunits', 'conditionlabel', 'confirmedby', 'confirmeddt', 'controlsubstanceflag','deviations', 'disposaldt',\n",
    "                     'disposalstatus', 'disposaltargetdt', 'disposedby', 'duedt', 'duedtoffset', 'duedtoffsettimeunit', 'duedtoverrideflag', 'eventdt', 'eventnum', 'eventplan',\n",
    "                     'eventplanitem', 'glpflag', 'instrumentid', 'locationid', 'locationpath', 'moddt', 'monitorgroupid', 'numberlabels', 'physicalcondition', 'pooledflag',\n",
    "                     'preptypeid', 'previousstoragestatus', 'priority', 'processinstruction', 'processtype', 'productid', 'productversionid', 'projectid', 'qcsampletype',\n",
    "                     'reagentlotid', 'receivedby', 'receiveddt', 'receiverequiredflag', 'requestid', 'requestitemdetailid', 'requestitemid', 'restrictionsflag',\n",
    "                      'reviewdisposition', 'reviewedby', 'revieweddt', 'reviewremarks', 'reviewrequiredflag', 'samplepointid', 'samplepointinstance', 'samplestatus',\n",
    "                      'schedulerulelabel', 'scheduletemplateflag', 'sdiworkitemcompletionstatus', 'securitydepartment', 'securityuser', 'sourcesdiworkitemid',\n",
    "                      'sourcespid', 'sourcesplevelid', 'sourcespsourcelabel', 'sourcespversionid',    'starttestingdt',\n",
    "                      'storagedisposalstatus', 'studyid', 'samplesubtypeid', 'u_aliquotnumber', 'specimentype', 'treatments',\n",
    "                      'storagestatus',   'templateflag',  'workorderid', 'basedonsampleid', 'u_qcfailureflag', 'u_qctype'], inplace=True)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bd7888d7-0c83-44dd-b502-7d6f8cf11396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sampletypeid\n",
       "Saliva          390\n",
       "DNA             310\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.value_counts(['sampletypeid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cddfe305-a3f2-4e41-9070-db98da6d5ea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_samplefamilyid</th>\n",
       "      <th>studysiteid</th>\n",
       "      <th>collectiondt</th>\n",
       "      <th>subjectid</th>\n",
       "      <th>tissueid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>u_initsampletypeid</th>\n",
       "      <th>u_inittreatmenttypeid</th>\n",
       "      <th>u_initcontainertypeid</th>\n",
       "      <th>u_initenvironmentid</th>\n",
       "      <th>u_specimentypeid</th>\n",
       "      <th>u_packageid</th>\n",
       "      <th>tissuedesc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_samplefamilyid, studysiteid, collectiondt, subjectid, tissueid, createdt, u_collectionid, u_initsampletypeid, u_inittreatmenttypeid, u_initcontainertypeid, u_initenvironmentid, u_specimentypeid, u_packageid, tissuedesc]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf[sf['s_samplefamilyid'] == 'SFO-000000876']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "43b7590a-92ff-46c8-a89d-1827d8a08d78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_sampleid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>batchid</th>\n",
       "      <th>samplefamilyid</th>\n",
       "      <th>sstudyid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [s_sampleid, sampletypeid, createdt, batchid, samplefamilyid, sstudyid]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[samples['samplefamilyid'] == 'SFO-000000876']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5d41f03e-6752-4c2f-af06-9e65eb9449a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Column('sourcesampleid', VARCHAR(length=40), ForeignKey('sapphire8.s_sample.s_sampleid'), table=<S_SAMPLEMAP>, primary_key=True, nullable=False),\n",
       " Column('destsampleid', VARCHAR(length=40), ForeignKey('sapphire8.s_sample.s_sampleid'), table=<S_SAMPLEMAP>, primary_key=True, nullable=False),\n",
       " Column('auditsequence', NUMBER(asdecimal=False), table=<S_SAMPLEMAP>),\n",
       " Column('usersequence', NUMBER(asdecimal=False), table=<S_SAMPLEMAP>),\n",
       " Column('createdt', DATE(), table=<S_SAMPLEMAP>),\n",
       " Column('createby', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('createtool', VARCHAR(length=20), table=<S_SAMPLEMAP>),\n",
       " Column('tracelogid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('moddt', DATE(), table=<S_SAMPLEMAP>),\n",
       " Column('modby', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('modtool', VARCHAR(length=20), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanversionid', VARCHAR(length=40), table=<S_SAMPLEMAP>),\n",
       " Column('childsampleplanitemid', VARCHAR(length=20), table=<S_SAMPLEMAP>)]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(SampleMap.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4847e681-c8f6-4219-8ae5-3465308a56a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sourcesampleid</th>\n",
       "      <th>destsampleid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>createdt_source</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>createdt_dest</th>\n",
       "      <th>sampletypeid_dest</th>\n",
       "      <th>map_minus_sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160657</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160656</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160658</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160658</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160656</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S-000000876</td>\n",
       "      <td>S-000160657</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592463</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592468</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592464</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592465</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592466</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592467</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592467</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592466</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592465</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592464</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592463</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S-000160658</td>\n",
       "      <td>S-000592468</td>\n",
       "      <td>2010-06-22 09:30:12</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>S-000592467</td>\n",
       "      <td>S-000967661</td>\n",
       "      <td>2010-10-09 17:51:31</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>S-000592467</td>\n",
       "      <td>S-000967661</td>\n",
       "      <td>2010-10-09 17:51:31</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>S-000592464</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>S-000592464</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>S-000967661</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>S-000967661</td>\n",
       "      <td>S-001363124</td>\n",
       "      <td>2014-06-20 07:23:59</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sourcesampleid destsampleid            createdt createdt_source  \\\n",
       "0     S-000000876  S-000160657 2010-06-22 09:30:12             NaT   \n",
       "1     S-000000876  S-000160656 2010-06-22 09:30:12             NaT   \n",
       "2     S-000000876  S-000160658 2010-06-22 09:30:12             NaT   \n",
       "3     S-000000876  S-000160658 2010-06-22 09:30:12             NaT   \n",
       "4     S-000000876  S-000160656 2010-06-22 09:30:12             NaT   \n",
       "5     S-000000876  S-000160657 2010-06-22 09:30:12             NaT   \n",
       "6     S-000160658  S-000592463 2010-06-22 09:30:12             NaT   \n",
       "7     S-000160658  S-000592468 2010-06-22 09:30:12             NaT   \n",
       "8     S-000160658  S-000592464 2010-06-22 09:30:12             NaT   \n",
       "9     S-000160658  S-000592465 2010-06-22 09:30:12             NaT   \n",
       "10    S-000160658  S-000592466 2010-06-22 09:30:12             NaT   \n",
       "11    S-000160658  S-000592467 2010-06-22 09:30:12             NaT   \n",
       "12    S-000160658  S-000592467 2010-06-22 09:30:12             NaT   \n",
       "13    S-000160658  S-000592466 2010-06-22 09:30:12             NaT   \n",
       "14    S-000160658  S-000592465 2010-06-22 09:30:12             NaT   \n",
       "15    S-000160658  S-000592464 2010-06-22 09:30:12             NaT   \n",
       "16    S-000160658  S-000592463 2010-06-22 09:30:12             NaT   \n",
       "17    S-000160658  S-000592468 2010-06-22 09:30:12             NaT   \n",
       "18    S-000592467  S-000967661 2010-10-09 17:51:31             NaT   \n",
       "19    S-000592467  S-000967661 2010-10-09 17:51:31             NaT   \n",
       "20    S-000592464  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "21    S-000592464  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "22    S-000967661  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "23    S-000967661  S-001363124 2014-06-20 07:23:59             NaT   \n",
       "\n",
       "   sampletypeid createdt_dest sampletypeid_dest map_minus_sample  \n",
       "0           NaN           NaT               NaN              NaT  \n",
       "1           NaN           NaT               NaN              NaT  \n",
       "2           NaN           NaT               NaN              NaT  \n",
       "3           NaN           NaT               NaN              NaT  \n",
       "4           NaN           NaT               NaN              NaT  \n",
       "5           NaN           NaT               NaN              NaT  \n",
       "6           NaN           NaT               NaN              NaT  \n",
       "7           NaN           NaT               NaN              NaT  \n",
       "8           NaN           NaT               NaN              NaT  \n",
       "9           NaN           NaT               NaN              NaT  \n",
       "10          NaN           NaT               NaN              NaT  \n",
       "11          NaN           NaT               NaN              NaT  \n",
       "12          NaN           NaT               NaN              NaT  \n",
       "13          NaN           NaT               NaN              NaT  \n",
       "14          NaN           NaT               NaN              NaT  \n",
       "15          NaN           NaT               NaN              NaT  \n",
       "16          NaN           NaT               NaN              NaT  \n",
       "17          NaN           NaT               NaN              NaT  \n",
       "18          NaN           NaT               NaN              NaT  \n",
       "19          NaN           NaT               NaN              NaT  \n",
       "20          NaN           NaT               NaN              NaT  \n",
       "21          NaN           NaT               NaN              NaT  \n",
       "22          NaN           NaT               NaN              NaT  \n",
       "23          NaN           NaT               NaN              NaT  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q_smap = session.query(SampleMap)\n",
    "q_smap = q_smap.filter(or_(SampleMap.c.destsampleid==Sample.c.s_sampleid, SampleMap.c.sourcesampleid==Sample.c.s_sampleid)).filter(Sample.c.samplefamilyid=='SFO-000000876')\n",
    "q_smap = q_smap.order_by(SampleMap.c.createdt, SampleMap.c.sourcesampleid)\n",
    "\n",
    "sample_map = pd.read_sql(q_smap.statement, session.bind)\n",
    "sample_map.drop(columns=['createby','createtool','tracelogid','moddt','modby','modtool','childsampleplanid','childsampleplanversionid','childsampleplanitemid', \n",
    "                         'auditsequence','usersequence'], inplace=True)\n",
    "sample_info = samples[['s_sampleid', 'createdt', 'sampletypeid']]\n",
    "sample_map = sample_map.merge(sample_info, left_on='sourcesampleid', right_on='s_sampleid', how='left', suffixes=(None, '_source'))\n",
    "sample_map = sample_map.merge(sample_info, left_on='destsampleid', right_on='s_sampleid', how='left', suffixes=(None, '_dest'))\n",
    "sample_map.drop(columns=['s_sampleid', 's_sampleid_dest'], inplace=True)\n",
    "sample_map['map_minus_sample'] = sample_map['createdt'] - sample_map['createdt_dest']\n",
    "sample_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fe6d13af-c051-4623-9072-f48c20e733b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAD4CAYAAADhNOGaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAONklEQVR4nO3cUWyd513H8e+PZJGYNpGOpF2WpDiMAIvQJsohVAzQWFdIwtQMaRcpsFYFKaqgqEigNVsluOCmCAmmaqVVVCpaMRFNrNCAMkLXAUMaGXVGmy6ErCawxiSsbkEdWi+irH8u/Fa45qQ+znlt13m+H8mK3/d9js//kaPz9Tk5TqoKSVK7vm2lB5AkrSxDIEmNMwSS1DhDIEmNMwSS1Li1Kz3A5diwYUNNTEys9BiStKocP378haraOP/8qgzBxMQEk5OTKz2GJK0qSb427LwvDUlS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS4wyBJDXOEEhS43oJQZJdSU4nmUpyYMj1JLm3u34iyXXzrq9J8k9J/rKPeSRJoxs7BEnWAPcBu4EdwM1JdsxbthvY3n3sB+6fd/1O4NS4s0iSFq+PZwQ7gamqOlNVF4BDwN55a/YCj9SsY8D6JJsAkmwBfgZ4sIdZJEmL1EcINgNn5xxPd+dGXfMJ4KPAK693J0n2J5lMMjkzMzPWwJKk/9NHCDLkXI2yJskHgeer6vhCd1JVB6tqUFWDjRs3Xs6ckqQh+gjBNLB1zvEW4NyIa94L3JTk35l9Sen9Sf64h5kkSSPqIwRPAtuTbEuyDtgHHJ635jBwS/fuoeuBl6rqfFV9rKq2VNVEd7vPV9Uv9DCTJGlEa8f9AlV1MckdwFFgDfBQVZ1Mcnt3/QHgCLAHmAJeBm4b934lSf1I1fyX89/4BoNBTU5OrvQYkrSqJDleVYP55/3NYklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMb1EoIku5KcTjKV5MCQ60lyb3f9RJLruvNbk/xNklNJTia5s495JEmjGzsESdYA9wG7gR3AzUl2zFu2G9jefewH7u/OXwR+vareBVwP/MqQ20qSllAfzwh2AlNVdaaqLgCHgL3z1uwFHqlZx4D1STZV1fmq+jJAVf0PcArY3MNMkqQR9RGCzcDZOcfT/P8H8wXXJJkAfhD4Ug8zSZJG1EcIMuRcLWZNkrcAnwF+raq+MfROkv1JJpNMzszMXPawkqTX6iME08DWOcdbgHOjrknyJmYj8KmqevRSd1JVB6tqUFWDjRs39jC2JAn6CcGTwPYk25KsA/YBh+etOQzc0r176Hrgpao6nyTAHwKnqur3ephFkrRIa8f9AlV1MckdwFFgDfBQVZ1Mcnt3/QHgCLAHmAJeBm7rbv5e4CPAM0me6s59vKqOjDuXJGk0qZr/cv4b32AwqMnJyZUeQ5JWlSTHq2ow/7y/WSxJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjTMEktQ4QyBJjeslBEl2JTmdZCrJgSHXk+Te7vqJJNeNeltJ0tIaOwRJ1gD3AbuBHcDNSXbMW7Yb2N597AfuX8RtJUlLqI9nBDuBqao6U1UXgEPA3nlr9gKP1KxjwPokm0a8rSRpCfURgs3A2TnH0925UdaMclsAkuxPMplkcmZmZuyhJUmz+ghBhpyrEdeMctvZk1UHq2pQVYONGzcuckRJ0qWs7eFrTANb5xxvAc6NuGbdCLeVJC2hPp4RPAlsT7ItyTpgH3B43prDwC3du4euB16qqvMj3laStITGfkZQVReT3AEcBdYAD1XVySS3d9cfAI4Ae4Ap4GXgtte77bgzSZJGl6qhL8m/oQ0Gg5qcnFzpMSRpVUlyvKoG88/7m8WS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNMwSS1DhDIEmNGysESd6W5PEkz3Z/XnWJdbuSnE4yleTAnPO/m+RfkpxI8mdJ1o8zjyRp8cZ9RnAAeKKqtgNPdMevkWQNcB+wG9gB3JxkR3f5ceAHqurdwFeBj405jyRpkcYNwV7g4e7zh4EPDVmzE5iqqjNVdQE41N2OqvrrqrrYrTsGbBlzHknSIo0bgmuq6jxA9+fVQ9ZsBs7OOZ7uzs33i8Bnx5xHkrRIaxdakORzwNuHXLp7xPvIkHM17z7uBi4Cn3qdOfYD+wGuvfbaEe9akrSQBUNQVR+41LUkX0+yqarOJ9kEPD9k2TSwdc7xFuDcnK9xK/BB4IaqKi6hqg4CBwEGg8El10mSFmfcl4YOA7d2n98KPDZkzZPA9iTbkqwD9nW3I8ku4C7gpqp6ecxZJEmXYdwQ3APcmORZ4MbumCTvSHIEoPvH4DuAo8Ap4NNVdbK7/SeBtwKPJ3kqyQNjziNJWqQFXxp6PVX1InDDkPPngD1zjo8AR4as+55x7l+SND5/s1iSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGmcIJKlxhkCSGjdWCJK8LcnjSZ7t/rzqEut2JTmdZCrJgSHXfyNJJdkwzjySpMUb9xnBAeCJqtoOPNEdv0aSNcB9wG5gB3Bzkh1zrm8FbgSeG3MWSdJlGDcEe4GHu88fBj40ZM1OYKqqzlTVBeBQd7tX/T7wUaDGnEWSdBnGDcE1VXUeoPvz6iFrNgNn5xxPd+dIchPwH1X19EJ3lGR/kskkkzMzM2OOLUl61dqFFiT5HPD2IZfuHvE+MuRcJXlz9zV+apQvUlUHgYMAg8HAZw+S1JMFQ1BVH7jUtSRfT7Kpqs4n2QQ8P2TZNLB1zvEW4BzwTmAb8HSSV89/OcnOqvrPRexBkjSGcV8aOgzc2n1+K/DYkDVPAtuTbEuyDtgHHK6qZ6rq6qqaqKoJZoNxnRGQpOU1bgjuAW5M8iyz7/y5ByDJO5IcAaiqi8AdwFHgFPDpqjo55v1Kknqy4EtDr6eqXgRuGHL+HLBnzvER4MgCX2tinFkkSZfH3yyWpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqnCGQpMYZAklqXKpqpWdYtCQzwNcu8+YbgBd6HGc1cM9tcM9tGGfP31VVG+efXJUhGEeSyaoarPQcy8k9t8E9t2Ep9uxLQ5LUOEMgSY1rMQQHV3qAFeCe2+Ce29D7npv7NwJJ0mu1+IxAkjSHIZCkxl2xIUiyK8npJFNJDgy5niT3dtdPJLluJebs0wh7/vluryeSfDHJe1Zizj4ttOc56344ybeSfHg55+vbKPtN8r4kTyU5meTvlnvGvo3w9/o7kvxFkqe7Pd+2EnP2KclDSZ5P8pVLXO/38auqrrgPYA3wr8B3A+uAp4Ed89bsAT4LBLge+NJKz70Me/5R4Kru890t7HnOus8DR4APr/TcS/w9Xg/8M3Btd3z1Ss+9DHv+OPA73ecbgf8C1q307GPu+yeA64CvXOJ6r49fV+ozgp3AVFWdqaoLwCFg77w1e4FHatYxYH2STcs9aI8W3HNVfbGq/rs7PAZsWeYZ+zbK9xngV4HPAM8v53BLYJT9/hzwaFU9B1BVLey5gLcmCfAWZkNwcXnH7FdVfYHZfVxKr49fV2oINgNn5xxPd+cWu2Y1Wex+fonZnyhWswX3nGQz8LPAA8s411IZ5Xv8vcBVSf42yfEktyzbdEtjlD1/EngXcA54Brizql5ZnvFWTK+PX2vHHueNKUPOzX+f7ChrVpOR95PkJ5kNwY8t6URLb5Q9fwK4q6q+NfsD46o2yn7XAj8E3AB8O/APSY5V1VeXerglMsqefxp4Cng/8E7g8SR/X1XfWOLZVlKvj19Xagimga1zjrcw+9PCYtesJiPtJ8m7gQeB3VX14jLNtlRG2fMAONRFYAOwJ8nFqvrzZZmwX6P+vX6hqr4JfDPJF4D3AKs1BKPs+Tbgnpp98Xwqyb8B3w/84/KMuCJ6ffy6Ul8aehLYnmRbknXAPuDwvDWHgVu6f32/Hnipqs4v96A9WnDPSa4FHgU+sop/QpxrwT1X1baqmqiqCeBPgV9epRGA0f5ePwb8eJK1Sd4M/Ahwapnn7NMoe36O2WdAJLkG+D7gzLJOufx6ffy6Ip8RVNXFJHcAR5l918FDVXUyye3d9QeYfQfJHmAKeJnZnypWrRH3/JvAdwJ/0P2EfLFW8f/cOOKerxij7LeqTiX5K+AE8ArwYFUNfQviajDi9/i3gT9K8gyzL5ncVVWr+r+mTvInwPuADUmmgd8C3gRL8/jlfzEhSY27Ul8akiSNyBBIUuMMgSQ1zhBIUuMMgSQ1zhBIUuMMgSQ17n8BHVr56ueUpskAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "_ = plt.hist(sf.collectiondt, bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0180fd51-e2b0-4159-ac07-f781332386a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x7ffa054c2610>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAc6klEQVR4nO3de3RU5bnH8e+TSUxQ0KNAEIVlkEJVCAQbES/YnFLRaioqVmKrFbUq9S4XS8RaXBWl1mptVSqn9YAtBqzWSy2KiqDIoSLRcAl3NUoAIaIoQbAEnvPHbOKEXMg9k83vs9as2fPuy7zPiL/seWfPO+buiIhIuCS0dAdERKTxKdxFREJI4S4iEkIKdxGREFK4i4iEUGJLdwCgQ4cOnpaW1tLdEBFpVfLz8z91945VrYuLcE9LS2PRokUt3Q0RkVbFzD6qbp2GZUREQkjhLiISQgp3EZEQiosx96rs2rWL4uJidu7c2dJdiTspKSl06dKFpKSklu6KiMSpuA334uJi2rVrR1paGmbW0t2JG+7Oli1bKC4uplu3bi3dHRGJU3E7LLNz507at2+vYN+HmdG+fXu9oxGRGsVtuAMK9mrodRGR/YnrcBcRkfrZb7ibWVczm2NmK8ys0MxuDtrHm9l6MysIbufE7JNrZmvNbJWZndUoPS1a37i3WohEImRkZNCrVy/69u3LAw88wJ49ewCYO3cuZsY///nP8u2zs7OZO3du+eOSkhKSkpJ47LHHGuUlEBGprdqcuZcBo9z9eGAAcL2ZnRCse9DdM4LbTIBgXQ7QCzgbeNTMIk3Q9ybXpk0bCgoKKCws5NVXX2XmzJncdddd5eu7dOnChAkTqt3/73//OwMGDCAvL685uisiUm6/4e7uG9393WB5G7ACOLqGXYYA0939a3f/EFgL9G+Mzrak1NRUJk+ezMMPP8zeX6/q27cvhx12GK+++mqV++Tl5fG73/2O4uJi1q+v3bsFEZHGUKcxdzNLA/oBbwdNN5jZEjN73MwOD9qOBtbF7FZMFX8MzOwaM1tkZotKSkrq3vMWcOyxx7Jnzx42b95c3nbHHXdw9913V9p23bp1fPLJJ/Tv35+LL76YGTNmNGdXReQAV+twN7O2wDPALe7+JTAJ6A5kABuB3+3dtIrdK/1Qq7tPdvdMd8/s2LHKSc3i0r6/OTtw4EAA5s2bV6F9+vTpXHzxxQDk5ORoaEZEmlWtvsRkZklEg32au/8DwN03xaz/H+DF4GEx0DVm9y7AhkbpbQv74IMPiEQipKamsmLFivL2cePGMWHCBBITv3k58/Ly2LRpE9OmTQNgw4YNrFmzhh49ejR7v0XkwFObq2UM+Auwwt0fiGnvHLPZBcCyYPkFIMfMks2sG9ADWNh4XW4ZJSUljBgxghtuuKHSdeaDBw/m888/Z/HixQCsWrWK7du3s379eoqKiigqKiI3N5fp06e3RNdF5ABUmzP304DLgKVmVhC03Q5cYmYZRIdcioBrAdy90MyeApYTvdLmenff3eCeptX0GW7T2LFjBxkZGezatYvExEQuu+wyRo4cWeW248aNY8iQIUD0rP2CCy6osH7o0KHk5OTwy1/+ssn7LSJi+44ht4TMzEzf98c6VqxYwfHHH99CPYp/en1ExMzy3T2zqnX6hqqISAgp3EVEQkjhLiISQgp3EZEQUriLiISQwl1EJITi9mf2KlkyvnGP16d2x5swYQJPPvkkkUiEhIQEHnvsMU4++eTG7YuISCNrPeHeAhYsWMCLL77Iu+++S3JyMp9++in/+c9/arVvWVlZhekIRESak9KnBhs3bqRDhw4kJycD0KFDBwDy8/MZOXIkpaWldOjQgSlTptC5c2eysrI49dRTmT9/Pueddx5Lly4lOzubiy66CIC2bdtSWlrK3Llz+dWvfkWnTp0oKCjgwgsvJD09nYceeogdO3bw3HPP0b179xarW0RaP42512Dw4MGsW7eOnj17ct111/HGG2+wa9cubrzxRp5++mny8/O58sorGTduXPk+W7du5Y033mDUqFE1Hnvx4sU89NBDLF26lL/+9a+sXr2ahQsX8rOf/Yw//vGPTV2aiIScztxr0LZtW/Lz85k3bx5z5sxh2LBh3HHHHSxbtowzzzwTgN27d9O58zdzqA0bNqxWxz7ppJPK9+vevTuDBw8GID09nTlz5jRyJSJyoFG470ckEiErK4usrCzS09N55JFH6NWrFwsWLKhy+0MOOaR8OTExsfw3V929wnj93qEegISEhPLHCQkJlJWVNUUpInIA0bBMDVatWsWaNWvKHxcUFHD88cdTUlJSHu67du2isLCwyv3T0tLIz88H4Pnnn2fXrl1N32kREVrTmXstL11sTKWlpdx4441s3bqVxMREvvWtbzF58mSuueYabrrpJr744gvKysq45ZZb6NWrV6X9r776aoYMGUL//v0ZNGhQhbN6EZGmpCl/Wym9PiKiKX9FRA4wCncRkRBSuIuIhJDCXUQkhBTuIiIhpHAXEQmh1nOdO+Ob/XiRSIT09HTKyso4/vjjmTp1KgcffHD5BGAiIvFKZ+41aNOmDQUFBSxbtoyDDjqIP/3pTy3dJRGRWlG419LAgQNZu3ZthbbS0lIGDRrEiSeeSHp6Os8//zwA27dv59xzz6Vv37707t2bGTNmANHpCG6//XZOOeUUMjMzeffddznrrLPo3r17+R+O6o4pIlIXrWhYpuWUlZXx0ksvcfbZZ1doT0lJ4dlnn+XQQw/l008/ZcCAAZx33nm8/PLLHHXUUfzrX/8C4Isvvijfp2vXrixYsIBbb72V4cOHM3/+fHbu3EmvXr0YMWJEtcc0s2atWURaN4V7DXbs2EFGRgYQPXO/6qqrKqx3d26//XbefPNNEhISWL9+PZs2bSI9PZ3Ro0fzi1/8guzsbAYOHFi+z3nnnQdEp/YtLS2lXbt2tGvXjpSUFLZu3cohhxxS5TGPPPLIZqtbRFo/hXsN9o65V2fatGmUlJSQn59PUlISaWlp7Ny5k549e5Kfn8/MmTPJzc1l8ODB3HnnnQAVpvbdd9rfsrKyao8pIlIXCvcG+OKLL0hNTSUpKYk5c+bw0UcfAbBhwwaOOOIILr30Utq2bcuUKVMafEwRkbpoReE+vqU7UMlPfvITfvjDH5KZmUlGRgbHHXccAEuXLmXMmDEkJCSQlJTEpEmTGnxMEZG60JS/rZReHxHRlL8iIgcYhbuISAjtN9zNrKuZzTGzFWZWaGY3B+1HmNmrZrYmuD88Zp9cM1trZqvM7KymLEBERCqrzZl7GTDK3Y8HBgDXm9kJwFhgtrv3AGYHjwnW5QC9gLOBR80s0hSdFxGRqu033N19o7u/GyxvA1YARwNDgKnBZlOB84PlIcB0d//a3T8E1gL9G7nfIiJSgzqNuZtZGtAPeBvo5O4bIfoHAEgNNjsaWBezW3HQtu+xrjGzRWa2qKSkpB5dFxGR6tQ63M2sLfAMcIu7f1nTplW0Vbre0t0nu3umu2d27Nhx/x0oWt+4t1rYunUrjz76aM3dKiriySefLH88d+5cDjvsMPr168dxxx3H6NGjy9dNmTKFhIQElixZUt7Wu3dvioqKyh+/9957mBmzZs2qVR9FRKpSq3A3sySiwT7N3f8RNG8ys87B+s7A5qC9GOgas3sXYEPjdLd51SfcIToPzXvvvcd7773Hiy++yPz588vXdenShQkTJlR7vLy8PE4//XTy8vIa1nkROaDV5moZA/4CrHD3B2JWvQBcHixfDjwf055jZslm1g3oASxsvC43n7Fjx/L++++TkZHBmDFjGDNmDL179yY9Pb18Gt+xY8cyb948MjIyePDBByvs36ZNGzIyMli//pt3CtnZ2RQWFrJq1apKz+fuPP3000yZMoVXXnlFc8qISL3V5sz9NOAy4HtmVhDczgEmAmea2RrgzOAx7l4IPAUsB14Grnf33U3S+yY2ceJEunfvTkFBAQMGDKCgoIDFixfz2muvMWbMGDZu3MjEiRMZOHAgBQUF3HrrrRX2//zzz1mzZg1nnHFGeVtCQgK33XYb99xzT6Xnmz9/Pt26daN79+5kZWUxc+bMJq9RRMKpNlfLvOXu5u593D0juM109y3uPsjdewT3n8XsM8Hdu7v7t939paYtoXm89dZbXHLJJUQiETp16sR3v/td3nnnnSq3nTdvHn369OHII48kOzu70nS9P/7xj/n3v//Nhx9+WKE9Ly+PnJwcAHJycjQ0IyL11oomDmtZdZmDZ+DAgbz44ousXr2a008/nQsuuKB8XniAxMRERo0axW9+85vytt27d/PMM8/wwgsvMGHCBNydLVu2sG3bNtq1a9eYpYjIAUDTD9SgXbt2bNu2DYAzzjiDGTNmsHv3bkpKSnjzzTfp379/hW321bNnT3JzcyuE+F7Dhw/ntddeY+9loK+99hp9+/Zl3bp1FBUV8dFHHzF06FCee+65JqtPRMKr9Zy5p1W6VL7JtW/fntNOO43evXvzgx/8gD59+tC3b1/MjPvuu48jjzyS9u3bk5iYSN++fRk+fDj9+vWrcIwRI0Zw//33VxqCOeigg7jpppu4+eabgeiQzAUXXFBhm6FDhzJp0iQuu+yypi1UREJHU/62Unp9RERT/oqIHGAU7iIiIaRwFxEJIYW7iEgIKdxFREJI4S4iEkKt5zr38eOb/XiRSIT09HTcnUgkwsMPP8ypp55KUVER2dnZLFu2rMHdyMrK4v777yczs8qrmURE6qX1hHsLaNOmDQUFBQDMmjWL3Nxc3njjjZbtlIhILWhYppa+/PJLDj/88ErtO3fu5IorriA9PZ1+/foxZ86cGtt37NhBTk4Offr0YdiwYezYsaNZ6xCRA4PO3GuwY8cOMjIy2LlzJxs3buT111+vtM0jjzwCwNKlS1m5ciWDBw9m9erV1bZPmjSJgw8+mCVLlrBkyRJOPPHEZq1JRA4MOnOvwd5hmZUrV/Lyyy/z05/+tNLskG+99Vb53C/HHXccxxxzDKtXr662/c033+TSSy8FoE+fPvTp06d5ixKRA4LCvZZOOeUUPv30U/b9Me/q5uapac6e6I9biYg0HYV7La1cuZLdu3fTvn37Cu1nnHEG06ZNA2D16tV8/PHHfPvb365V+7Jlyyr8WLaISGNpPWPujX0pZC3sHXOH6Jn41KlTiUQiFba57rrrGDFiBOnp6SQmJjJlyhSSk5Orbf/5z3/OFVdcQZ8+fcjIyKB///7NXpeIhJ+m/G2l9PqIiKb8FRE5wCjcRURCSOEuIhJCCncRkRBSuIuIhJDCXUQkhFrNde7j545v3ONl7f94bdu2pbS0tFL7E088wX333Ye74+5ceeWVjB49muHDh5Odnc1FF13UqH0VEakrnbnX0UsvvcTvf/97XnnlFQoLC3n33Xc57LDDWrpbIiIVKNzr6N577+X+++/nqKOOAiAlJYWrr766hXslIlKRwr2Oli1bxne+852W7oaISI0U7iIiIaRwr6NevXqRn5/f0t0QEamRwr2OcnNzue222/jkk08A+Prrr/nDH/7Qwr0SEalov5dCmtnjQDaw2d17B23jgauBvb9ccbu7zwzW5QJXAbuBm9x9VmN0tDaXLja2r776ii5dupQ/HjlyJCNHjmTTpk18//vfx90xM6688sryba699lpuueUWALp27cqCBQuau9siIvuf8tfMzgBKgSf2CfdSd79/n21PAPKA/sBRwGtAT3ffXdNzaMrfutPrIyINmvLX3d8EPqvlcw0Bprv71+7+IbCWaNCLiEgzasiY+w1mtsTMHjezw4O2o4F1MdsUB22VmNk1ZrbIzBbt+7ukIiLSMPUN90lAdyAD2Aj8Lmiv6pefqxz3cffJ7p7p7pkdO3asZzdERKQq9Qp3d9/k7rvdfQ/wP3wz9FIMdI3ZtAuwoWFdFBGRuqpXuJtZ55iHFwDLguUXgBwzSzazbkAPYGHDuigiInVVm0sh84AsoIOZFQO/ArLMLIPokEsRcC2Auxea2VPAcqAMuH5/V8qIiEjj22+4u/slVTT/pYbtJwATGtKpqowf3zLHKy4u5vrrr2f58uXs2bOH7Oxsfvvb37J8+XI2bNjAOeecExxvPG3btmX06NGN21ERkXrQN1Rr4O5ceOGFnH/++axZs4bVq1dTWlrKuHHjKCgoYObMmY32XLt36w2OiDQehXsNXn/9dVJSUrjiiisAiEQiPPjgg/z5z3/mtttuY8aMGWRkZDBjxgwAli9fTlZWFscee2yFKQn+9re/0b9/fzIyMrj22mvLg7xt27bceeednHzyyfomq4g0KoV7DQoLCytN73vooYeSlpbGHXfcwbBhwygoKGDYsGEArFy5klmzZrFw4ULuuusudu3axYoVK5gxYwbz58+noKCASCTCtGnTANi+fTu9e/fm7bff5vTTT2/2+kQkvFrNz+y1hL1zx9S2/dxzzyU5OZnk5GRSU1PZtGkTs2fPJj8/n5NOOgmAHTt2kJqaCkTfCQwdOrRpixCRA5LCvQa9evXimWeeqdD25Zdfsm7dOiKRSKXtk5OTy5cjkQhlZWW4O5dffjn33ntvpe1TUlKqPI6ISENpWKYGgwYN4quvvuKJJ54Aoh96jho1iuHDh9OpUye2bdtWq2M8/fTTbN68GYDPPvuMjz76qEn7LSLSas7cG/tSyNowM5599lmuu+46fv3rX7Nnzx7OOecc7rnnHrZv387EiRPJyMggNze32mOccMIJ3H333QwePJg9e/aQlJTEI488wjHHHNOMlYjIgWa/U/42B035W3d6fUSkQVP+iohI66NwFxEJobgO93gYMopHel1EZH/iNtxTUlLYsmWLgmwf7s6WLVtISUlp6a6ISByL26tlunTpQnFxMfqVpspSUlIq/HC3iMi+4jbck5KS6NatW0t3Q0SkVYrbYRkREak/hbuISAgp3EVEQkjhLiISQgp3EZEQUriLiISQwl1EJIQU7iIiIaRwFxEJIYW7iEgIKdxFREJI4S4iEkIKdxGREFK4i4iEkMJdRCSEFO4iIiGkcBcRCSGFu4hICO033M3scTPbbGbLYtqOMLNXzWxNcH94zLpcM1trZqvM7Kym6riIiFSvNmfuU4Cz92kbC8x29x7A7OAxZnYCkAP0CvZ51MwijdZbERGplf2Gu7u/CXy2T/MQYGqwPBU4P6Z9urt/7e4fAmuB/o3TVRERqa36jrl3cveNAMF9atB+NLAuZrvioK0SM7vGzBaZ2aKSkpJ6dkNERKrS2B+oWhVtXtWG7j7Z3TPdPbNjx46N3A0RkQNbfcN9k5l1BgjuNwftxUDXmO26ABvq3z0REamP+ob7C8DlwfLlwPMx7Tlmlmxm3YAewMKGdVFEROoqcX8bmFkekAV0MLNi4FfAROApM7sK+Bj4EYC7F5rZU8ByoAy43t13N1HfRUSkGvsNd3e/pJpVg6rZfgIwoSGdEhGRhtE3VEVEQkjhLiISQgp3EZEQUriLiISQwl1EJIQU7iIiIaRwFxEJIYW7iEgIKdxFREJI4S4iEkIKdxGREFK4i4iEkMJdRCSEFO4iIiGkcBcRCSGFu4hICCncRURCSOEuIhJCCncRkRBSuIuIhJDCXUQkhBTuIiIhpHAXEQkhhbuISAgp3EVEQkjhLiISQgp3EZEQUriLiISQwl1EJIQU7iIiIaRwFxEJIYW7iEgIJTZkZzMrArYBu4Eyd880syOAGUAaUARc7O6fN6ybIiJSF41x5v7f7p7h7pnB47HAbHfvAcwOHouISDNqimGZIcDUYHkqcH4TPIeIiNSgoeHuwCtmlm9m1wRtndx9I0Bwn1rVjmZ2jZktMrNFJSUlDeyGiIjEatCYO3Cau28ws1TgVTNbWdsd3X0yMBkgMzPTG9gPERGJ0aAzd3ffENxvBp4F+gObzKwzQHC/uaGdFBGRuql3uJvZIWbWbu8yMBhYBrwAXB5sdjnwfEM7KSIiddOQYZlOwLNmtvc4T7r7y2b2DvCUmV0FfAz8qOHdFBGRuqh3uLv7B0DfKtq3AIMa0ikREWkYfUNVRCSEFO4iIiGkcBcRCSGFu4hICCncRURCSOEuIhJCCncRkRBSuIuIhJDCXUQkhBTuIiIhpHAXEQkhhbuISAgp3EVEQkjhLiISQgp3EZEQUriLiISQwl1EJIQU7iIiIaRwFxEJIYW7iEgIKdxFREJI4S4iEkIKdxGREFK4i4iEkMJdRCSEFO4iIiGkcBcRCSGFu4hICCncRURCSOEuIhJCCncRkcZUtL6lewAo3EVEmk5s0Ddz6DdZuJvZ2Wa2yszWmtnYpnoeEZF4Nn58yzxvk4S7mUWAR4AfACcAl5jZCU3xXCIiUllTnbn3B9a6+wfu/h9gOjCkiZ5LRCQujB8P43/frtqz9eY8izd3b/yDml0EnO3uPwseXwac7O43xGxzDXBN8PDbwKoGPGUH4NMG7N8aqMZwUI3hEC81HuPuHatakdhET2hVtFX4K+Luk4HJjfJkZovcPbMxjhWvVGM4qMZwaA01NtWwTDHQNeZxF2BDEz2XiIjso6nC/R2gh5l1M7ODgBzghSZ6LhER2UeTDMu4e5mZ3QDMAiLA4+5e2BTPFWiU4Z04pxrDQTWGQ9zX2CQfqIqISMvSN1RFREJI4S4iEkJxGe5m9riZbTazZTFtfc1sgZktNbN/mtmhQftPzKwg5rbHzDKCdd8Jtl9rZn8ws6ou0WwRdawxycymBu0rzCw3Zp+w1HiQmf1v0L7YzLJi9onnGrua2Zzgv0uhmd0ctB9hZq+a2Zrg/vCYfXKDWlaZ2Vkx7XFZZ11rNLP2wfalZvbwPscKS41nmll+UEu+mX0v5ljxUaO7x90NOAM4EVgW0/YO8N1g+Urg11Xslw58EPN4IXAK0evuXwJ+0NK11adG4MfA9GD5YKAISAtZjdcD/xsspwL5QEIrqLEzcGKw3A5YTXTKjfuAsUH7WOA3wfIJwGIgGegGvA9E4rnOetR4CHA6MAJ4eJ9jhaXGfsBRwXJvYH281djiL2oNL3baPqHwJd98ANwVWF7FPvcAE2L+Y62MWXcJ8FhL11WfGoO+/5Po1U3tg394R4SsxkeAS2O2m010Gou4r3Gfep8HziT6jevOMf8WVwXLuUBuzPazgiBoNXXur8aY7YbHhnsYawzaDdhC9A923NQYl8My1VgGnBcs/4iKX5LaaxiQFywfTfTLVHsVB23xrLoanwa2AxuBj4H73f0zwlXjYmCImSWaWTfgO8G6VlOjmaURPaN7G+jk7hsBgvvUYLOjgXUxu+2tp1XUWcsaqxPWGocC77n718RRja0p3K8ErjezfKJvm/4Tu9LMTga+cve947v7nQIhDlVXY39gN3AU0bfyo8zsWMJV4+NE/0dYBPwe+D+gjFZSo5m1BZ4BbnH3L2vatIo2r6E9btShxmoPUUVbq67RzHoBvwGu3dtUxWYtUmNTzS3T6Nx9JTAYwMx6Aufus0kO35y1QzQousQ8jvspEGqo8cfAy+6+C9hsZvOBTGAeIanR3cuAW/duZ2b/B6wBPifOazSzJKKBMM3d/xE0bzKzzu6+0cw6A5uD9uqm5ojrf691rLE6oarRzLoAzwI/dff3g+a4qbHVnLmbWWpwnwDcAfwpZl0C0bf40/e2BW+htpnZgODT6p8SHUeLWzXU+DHwPYs6BBhAdFwvNDWa2cFBbZjZmUCZuy+P9xqDPv0FWOHuD8SsegG4PFi+nG/6/AKQY2bJwfBTD2BhPNdZjxqrFKYazey/gH8R/fxk/t6N46rGlv7gopoPM/KIji/vIvqX8CrgZqIfJK4GJhJ8KBdsnwX8u4rjZBId430feDh2n5a+1aVGoC3wd6AQWA6MCWGNaUQ/vFoBvEZ0KtPWUOPpRN92LwEKgts5RD/4nk303cds4IiYfcYFtawi5kqKeK2znjUWAZ8BpcF/+xPCVCPRE5PtMdsWAKnxVKOmHxARCaFWMywjIiK1p3AXEQkhhbuISAgp3EVEQkjhLiISQgp3EZEQUriLiITQ/wNBzeW/gGyZFQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_types = ['DNA', 'Serum', 'Plasma', 'totRNA', 'Blood', 'LCL']\n",
    "\n",
    "colors = ['pink', 'orange', 'yellow', 'pink', 'red', 'green']\n",
    "\n",
    "for (s, st) in enumerate(sample_types):\n",
    "    x = samples.loc[samples.sampletypeid==st, 'createdt']\n",
    "    plt.hist(x, alpha=0.5, bins=100, label=str(st), color=colors[s])  #density=True, stacked=True\n",
    "x = samples.loc[~samples.sampletypeid.isin(sample_types), 'createdt']\n",
    "plt.hist(x, alpha=0.5, bins=100, label=str('Other'), color='blue')\n",
    "\n",
    "#plt.suptitle('Probability Histogram of Diamond Depths', y=1.05, size=16)\n",
    "#ax.set_xlim(50, 70); ax.set_ylim(0, 1);\n",
    "#plt.tight_layout();\n",
    "\n",
    "#_ = plt.hist(samples.createdt, bins=100)\n",
    "#plt.gca().set(title='Frequency Histogram of Diamond Depths', ylabel='Frequency')\n",
    "#plt.xlim(50,75)\n",
    "plt.legend()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24d0c0b-e3aa-4026-8a86-02db8109a5a0",
   "metadata": {},
   "source": [
    "### Sankey Diagram of Sample Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a82f5299-0dba-4d43-b2c9-845d7aece17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample=Table('S_SAMPLE', MetaData(), Column('s_sampleid', VARCHAR(length=40), table=<S_SAMPLE>, primary_key=True, nullable=False), Column('sampledesc', VARCHAR(length=80), table=<S_SAMPLE>), Column('sampletypeid', VARCHAR(length=40), ForeignKey('sapphire8.s_sampletype.s_sampletypeid'), table=<S_SAMPLE>), Column('createdt', DATE(), table=<S_SAMPLE>), Column('createby', VARCHAR(length=40), table=<S_SAMPLE>), Column('createtool', VARCHAR(length=20), table=<S_SAMPLE>), Column('moddt', DATE(), table=<S_SAMPLE>), Column('submitterid', VARCHAR(length=40), table=<S_SAMPLE>), Column('modby', VARCHAR(length=40), table=<S_SAMPLE>), Column('submitteddt', DATE(), table=<S_SAMPLE>), Column('modtool', VARCHAR(length=20), table=<S_SAMPLE>), Column('notes', VARCHAR(length=2000), table=<S_SAMPLE>), Column('auditsequence', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('tracelogid', VARCHAR(length=40), table=<S_SAMPLE>), Column('usersequence', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('reviewrequiredflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('priority', VARCHAR(length=20), table=<S_SAMPLE>), Column('disposalstatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('samplestatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('collectiondt', DATE(), table=<S_SAMPLE>), Column('receiveddt', DATE(), table=<S_SAMPLE>), Column('batchid', VARCHAR(length=40), table=<S_SAMPLE>), Column('physicalcondition', VARCHAR(length=80), table=<S_SAMPLE>), Column('productid', VARCHAR(length=40), table=<S_SAMPLE>), Column('projectid', VARCHAR(length=40), table=<S_SAMPLE>), Column('samplepointid', VARCHAR(length=40), table=<S_SAMPLE>), Column('locationid', VARCHAR(length=40), table=<S_SAMPLE>), Column('requestid', VARCHAR(length=40), table=<S_SAMPLE>), Column('starttestingdt', DATE(), table=<S_SAMPLE>), Column('completedt', DATE(), table=<S_SAMPLE>), Column('disposaldt', DATE(), table=<S_SAMPLE>), Column('duedt', DATE(), table=<S_SAMPLE>), Column('basedonsampleid', VARCHAR(length=40), ForeignKey('sapphire8.s_sample.s_sampleid'), table=<S_SAMPLE>), Column('collectedby', VARCHAR(length=20), table=<S_SAMPLE>), Column('templateflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('receivedby', VARCHAR(length=40), table=<S_SAMPLE>), Column('disposedby', VARCHAR(length=20), table=<S_SAMPLE>), Column('reviewedby', VARCHAR(length=40), table=<S_SAMPLE>), Column('revieweddt', DATE(), table=<S_SAMPLE>), Column('reviewdisposition', VARCHAR(length=80), table=<S_SAMPLE>), Column('reviewremarks', VARCHAR(length=255), table=<S_SAMPLE>), Column('numberlabels', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('controlsubstanceflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('receiverequiredflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('cocrequiredflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('autoreceiveflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('eventdt', DATE(), table=<S_SAMPLE>), Column('autofinalrptflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('studyid', VARCHAR(length=40), table=<S_SAMPLE>), Column('eventplan', VARCHAR(length=40), table=<S_SAMPLE>), Column('conditionlabel', VARCHAR(length=80), table=<S_SAMPLE>), Column('eventplanitem', VARCHAR(length=20), table=<S_SAMPLE>), Column('samplefamilyid', VARCHAR(length=40), ForeignKey('sapphire8.s_samplefamily.s_samplefamilyid'), table=<S_SAMPLE>), Column('schedulerulelabel', VARCHAR(length=80), table=<S_SAMPLE>), Column('confirmedby', VARCHAR(length=40), table=<S_SAMPLE>), Column('confirmeddt', DATE(), table=<S_SAMPLE>), Column('sstudyid', VARCHAR(length=40), ForeignKey('sapphire8.s_study.s_studyid'), table=<S_SAMPLE>), Column('pooledflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('glpflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('cocflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('restrictionsflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('concentration', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('concentrationunits', VARCHAR(length=40), table=<S_SAMPLE>), Column('storagestatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('previousstoragestatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('deviations', VARCHAR(length=2000), table=<S_SAMPLE>), Column('treatments', VARCHAR(length=2000), table=<S_SAMPLE>), Column('allocatedforaddressid', VARCHAR(length=40), ForeignKey('sapphire8.address.addressid'), table=<S_SAMPLE>), Column('allocatedforaddresstype', VARCHAR(length=80), ForeignKey('sapphire8.address.addresstype'), table=<S_SAMPLE>), Column('allocatedfordepartmentid', VARCHAR(length=40), ForeignKey('sapphire8.department.departmentid'), table=<S_SAMPLE>), Column('preptypeid', VARCHAR(length=40), ForeignKey('sapphire8.s_preptype.s_preptypeid'), table=<S_SAMPLE>), Column('storagedisposalstatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('reagentlotid', VARCHAR(length=40), table=<S_SAMPLE>), Column('qcsampletype', VARCHAR(length=80), table=<S_SAMPLE>), Column('specimentype', VARCHAR(length=80), table=<S_SAMPLE>), Column('disposaltargetdt', DATE(), table=<S_SAMPLE>), Column('sourcespid', VARCHAR(length=40), table=<S_SAMPLE>), Column('sourcespversionid', VARCHAR(length=40), table=<S_SAMPLE>), Column('sourcesplevelid', VARCHAR(length=20), table=<S_SAMPLE>), Column('sourcespsourcelabel', VARCHAR(length=80), table=<S_SAMPLE>), Column('batchstageid', VARCHAR(length=40), table=<S_SAMPLE>), Column('activeflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('productversionid', VARCHAR(length=40), table=<S_SAMPLE>), Column('instrumentid', VARCHAR(length=40), table=<S_SAMPLE>), Column('workorderid', VARCHAR(length=40), table=<S_SAMPLE>), Column('samplesubtypeid', VARCHAR(length=40), ForeignKey('sapphire8.s_sampletype.s_sampletypeid'), table=<S_SAMPLE>), Column('classification', VARCHAR(length=40), table=<S_SAMPLE>), Column('duedtoffset', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('duedtoffsettimeunit', VARCHAR(length=80), table=<S_SAMPLE>), Column('duedtoverrideflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('sourcesdiworkitemid', VARCHAR(length=40), table=<S_SAMPLE>), Column('requestitemid', VARCHAR(length=40), table=<S_SAMPLE>), Column('requestitemdetailid', VARCHAR(length=40), table=<S_SAMPLE>), Column('cancelleddt', DATE(), table=<S_SAMPLE>), Column('cancelledby', VARCHAR(length=40), table=<S_SAMPLE>), Column('sdiworkitemcompletionstatus', VARCHAR(length=20), table=<S_SAMPLE>), Column('scheduletemplateflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('eventnum', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('processinstruction', VARCHAR(length=2000), table=<S_SAMPLE>), Column('processtype', VARCHAR(length=80), table=<S_SAMPLE>), Column('monitorgroupid', VARCHAR(length=40), table=<S_SAMPLE>), Column('locationpath', VARCHAR(length=2000), table=<S_SAMPLE>), Column('samplepointinstance', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('u_aliquotnumber', NUMBER(asdecimal=False), table=<S_SAMPLE>), Column('u_qcfailureflag', VARCHAR(length=1), table=<S_SAMPLE>), Column('u_qctype', VARCHAR(length=20), table=<S_SAMPLE>), Column('securityuser', VARCHAR(length=40), ForeignKey('sapphire8.sysuser.sysuserid'), table=<S_SAMPLE>), Column('securitydepartment', VARCHAR(length=40), ForeignKey('sapphire8.department.departmentid'), table=<S_SAMPLE>), schema='SAPPHIRE8')\n"
     ]
    }
   ],
   "source": [
    "print(f'{Sample=}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "20c380aa-8e7d-4b95-b336-c0f2fa41800f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">s_sampleid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>list</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u_collectionid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [(s_sampleid, list), (s_sampleid, count)]\n",
       "Index: []"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build the data structure -- source,dest,count\n",
    "\n",
    "#1: Find the bases of the tree (samples with no parents) and then group them by collection (count)\n",
    "\n",
    "data = {'studyid': STUDY}\n",
    "statement = text(\"\"\"select s.S_SAMPLEID from sapphire8.s_sample s where s.SSTUDYID='CAMP' AND S_SAMPLEID NOT IN (SELECT DESTSAMPLEID from sapphire8.s_samplemap)\"\"\")\n",
    "base_samples = pd.read_sql(statement, session.bind, params=data)\n",
    "base_samples = base_samples.merge(samples, on='s_sampleid', how='left')\n",
    "base_samples = base_samples.merge(sf, left_on='samplefamilyid', right_on='s_samplefamilyid', how='left')\n",
    "# base_samples\n",
    "\n",
    "base_samples_grouped = base_samples.groupby(['u_collectionid', 'sampletypeid']).agg({'s_sampleid': [list, 'count'],})\n",
    "base_samples_grouped\n",
    "\n",
    "# base_samples.value_counts(['u_collectionid', 'sampletypeid'])\n",
    "\n",
    "# https://plainenglish.io/blog/create-a-sankey-diagram-in-python-e09e23cb1a75\n",
    "#def query_sample_tree():\n",
    "#    data = { \"sampleid\": sid}\n",
    "#\n",
    "#    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "#    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "#    return sample_history\n",
    "#\n",
    "#shistory = query_sample_history('S-001357960')\n",
    "#shistory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5856735d-7da0-4d5c-a4f3-fd3be1bdb48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_next_level(row):\n",
    "    cid = row['u_collectionid']\n",
    "    statement = text(\"\"\"select s.S_SAMPLEID from sapphire8.s_sample s where s.SSTUDYID='CAMP' AND S_SAMPLEID NOT IN (SELECT DESTSAMPLEID from sapphire8.s_samplemap)\"\"\")\n",
    "    base_samples = pd.read_sql(statement, session.bind, params=data)\n",
    "    base_samples = base_samples.merge(samples, on='s_sampleid', how='left')\n",
    "    base_samples = base_samples.merge(sf, left_on='samplefamilyid', right_on='s_samplefamilyid', how='left')\n",
    "    # base_samples\n",
    "\n",
    "    base_samples_grouped = base_samples.groupby(['u_collectionid', 'sampletypeid']).agg({'s_sampleid': [list, 'count'],})\n",
    "    base_samples_grouped\n",
    "\n",
    "    # base_samples.value_counts(['u_collectionid', 'sampletypeid'])\n",
    "\n",
    "    # https://plainenglish.io/blog/create-a-sankey-diagram-in-python-e09e23cb1a75\n",
    "    #def query_sample_tree():\n",
    "    #    data = { \"sampleid\": sid}\n",
    "    #\n",
    "    #    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "    #    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "    #    return sample_history\n",
    "    #\n",
    "    #shistory = query_sample_history('S-001357960')\n",
    "    #shistory    \n",
    "base_samples_grouped = base_samples_grouped.apply(annotate_next_level, axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa115812-e299-4dd9-8cac-b53d64627df1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5825f826-f503-4736-9fcb-d9de9fc37704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_packageid</th>\n",
       "      <th>packagedesc</th>\n",
       "      <th>packagestatus</th>\n",
       "      <th>contenttype</th>\n",
       "      <th>sstudyid</th>\n",
       "      <th>sampletypeid</th>\n",
       "      <th>expectedsamplecnt</th>\n",
       "      <th>recipientdepartmentid</th>\n",
       "      <th>createdt</th>\n",
       "      <th>activeflag</th>\n",
       "      <th>u_shippeddate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKG-0321-01005270</td>\n",
       "      <td>AACR DNA Pilot to PPM</td>\n",
       "      <td>Created</td>\n",
       "      <td>Sample</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PPM</td>\n",
       "      <td>2021-03-31 12:38:10</td>\n",
       "      <td>Y</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PKG-1222-01005626</td>\n",
       "      <td>ESCAPE (old AACR) DNA plates to BROAD -23703</td>\n",
       "      <td>Shipped</td>\n",
       "      <td>Sample</td>\n",
       "      <td>ESCAPE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>2022-12-05 10:45:34</td>\n",
       "      <td>Y</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         s_packageid                                   packagedesc  \\\n",
       "0  PKG-0321-01005270                         AACR DNA Pilot to PPM   \n",
       "1  PKG-1222-01005626  ESCAPE (old AACR) DNA plates to BROAD -23703   \n",
       "\n",
       "  packagestatus contenttype sstudyid sampletypeid expectedsamplecnt  \\\n",
       "0       Created      Sample   ESCAPE         None              None   \n",
       "1       Shipped      Sample   ESCAPE         None              None   \n",
       "\n",
       "  recipientdepartmentid            createdt activeflag u_shippeddate  \n",
       "0                   PPM 2021-03-31 12:38:10          Y          None  \n",
       "1                 BROAD 2022-12-05 10:45:34          Y          None  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_packages = session.query(Package).filter(Package.c.sstudyid==STUDY)\n",
    "packages = pd.read_sql(q_packages.statement, session.bind)\n",
    "packages.drop(columns=['carrier', 'carriertype', 'trackingnumber', 'createby', 'createtool', 'moddt', 'modby', 'modtool', 'securityuser', 'securitydepartment',\n",
    "                       'usersequence', 'notes', 'auditsequence', 'condition', 'senderaddressid', 'senderaddresstype', 'recipientaddressid', 'recipientaddresstype',\n",
    "                       'tracelogid', 'templateflag', 'packagetype', 'senderdepartmentid', 'expecteddt', ], inplace=True)\n",
    "packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1271e44b-c324-4f2f-ad5c-ad9c41824b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open(DATASETS_PATH) as fh:\n",
    "#    raw = json.load(fh)#\n",
    "#\n",
    "#for ds in raw['datasets']:\n",
    "#    manifest_path = ds['manifest_path']\n",
    "#    manifest_delim = ds.get('manifest_delimiter', ',')\n",
    "#    manifest_columns = ds.get('manifest_columns', None)\n",
    "#    if manifest_columns:\n",
    "#        sample_table = pd.read_csv(manifest_path, delimiter=manifest_delim, names=manifest_columns)\n",
    "#    else:\n",
    "#        sample_table = pd.read_csv(manifest_path, delimiter=manifest_delim)\n",
    "#    print(sample_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3a4667-42d0-4b03-8591-e07ccd0f445e",
   "metadata": {},
   "source": [
    "## Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "76a5748f-362e-407a-a56f-bbfeb4df7f42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1100"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redmine = redminelib.Redmine('https://chanmine.bwh.harvard.edu/', key='bc92021bd829b2c07aa94b3bc2679c639204b902', requests={'verify': False})\n",
    "trackers = pd.DataFrame(data=[dict(d) for d in redmine.tracker.all()])\n",
    "\n",
    "projects = redmine.project.all()\n",
    "projects = pd.DataFrame([dict(d) for d in projects])\n",
    "projects = projects.drop(columns=['wiki_pages', 'memberships', 'issue_categories', 'time_entries', 'versions', 'news',\n",
    "                                  'issues', 'files', 'trackers', 'enabled_modules', 'time_entry_activities', 'issue_custom_fields',\n",
    "                                  'is_public', 'inherit_members', 'status', 'created_on', 'updated_on'])\n",
    "\n",
    "study_project_id = projects.loc[projects['name'] == REDMINE_STUDY, 'id'].iloc[0]\n",
    "study_project_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "c3e3483b-0b4c-4ac0-972b-9f93079f7044",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relations</th>\n",
       "      <th>time_entries</th>\n",
       "      <th>children</th>\n",
       "      <th>attachments</th>\n",
       "      <th>changesets</th>\n",
       "      <th>journals</th>\n",
       "      <th>watchers</th>\n",
       "      <th>allowed_statuses</th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>...</th>\n",
       "      <th>category</th>\n",
       "      <th>parent</th>\n",
       "      <th>project</th>\n",
       "      <th>category</th>\n",
       "      <th>status</th>\n",
       "      <th>author_name</th>\n",
       "      <th>assignee</th>\n",
       "      <th>parent</th>\n",
       "      <th>url</th>\n",
       "      <th>omic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 463, 'name': 'dna/exome'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/exome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 462, 'name': 'dna/whole_genome'}</td>\n",
       "      <td>{'id': 22862}</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/whole_genome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>22862.0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 464, 'name': 'epigenetic/methylation'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>epigenetic/methylation</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 465, 'name': 'protein/olink'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>protein/olink</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>{'id': 465, 'name': 'protein/olink'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>protein/olink</td>\n",
       "      <td>New</td>\n",
       "      <td>Brian Hobbs</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>None</td>\n",
       "      <td>New</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23145</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rebays</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22764</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rebays</td>\n",
       "      <td>None</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    relations time_entries children attachments changesets journals watchers  \\\n",
       "82       None         None     None        None       None     None     None   \n",
       "18       None         None     None        None       None     None     None   \n",
       "99       None         None     None        None       None     None     None   \n",
       "15       None         None     None        None       None     None     None   \n",
       "57       None         None     None        None       None     None     None   \n",
       "..        ...          ...      ...         ...        ...      ...      ...   \n",
       "259      None         None     None        None       None     None     None   \n",
       "342      None         None     None        None       None     None     None   \n",
       "297      None         None     None        None       None     None     None   \n",
       "73       None         None     None        None       None     None     None   \n",
       "85       None         None     None        None       None     None     None   \n",
       "\n",
       "    allowed_statuses     id                         project  ...  \\\n",
       "82              None  22862       {'id': 28, 'name': 'A1A'}  ...   \n",
       "18              None  24096       {'id': 28, 'name': 'A1A'}  ...   \n",
       "99              None  22080       {'id': 28, 'name': 'A1A'}  ...   \n",
       "15              None  24110       {'id': 28, 'name': 'A1A'}  ...   \n",
       "57              None  23928       {'id': 28, 'name': 'A1A'}  ...   \n",
       "..               ...    ...                             ...  ...   \n",
       "259             None  18545     {'id': 235, 'name': 'VIVA'}  ...   \n",
       "342             None  16227   {'id': 952, 'name': 'eMERGE'}  ...   \n",
       "297             None  17846   {'id': 952, 'name': 'eMERGE'}  ...   \n",
       "73              None  23145  {'id': 1135, 'name': 'rebays'}  ...   \n",
       "85              None  22764  {'id': 1135, 'name': 'rebays'}  ...   \n",
       "\n",
       "                                          category         parent project  \\\n",
       "82                {'id': 463, 'name': 'dna/exome'}            NaN     A1A   \n",
       "18         {'id': 462, 'name': 'dna/whole_genome'}  {'id': 22862}     A1A   \n",
       "99   {'id': 464, 'name': 'epigenetic/methylation'}            NaN     A1A   \n",
       "15            {'id': 465, 'name': 'protein/olink'}            NaN     A1A   \n",
       "57            {'id': 465, 'name': 'protein/olink'}            NaN     A1A   \n",
       "..                                             ...            ...     ...   \n",
       "259                                            NaN            NaN    VIVA   \n",
       "342                                            NaN            NaN  eMERGE   \n",
       "297                                            NaN            NaN  eMERGE   \n",
       "73                                             NaN            NaN  rebays   \n",
       "85                                             NaN            NaN  rebays   \n",
       "\n",
       "                   category       status  author_name     assignee   parent  \\\n",
       "82                dna/exome  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "18         dna/whole_genome  In Progress  John Ziniti  John Ziniti  22862.0   \n",
       "99   epigenetic/methylation  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "15            protein/olink  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "57            protein/olink          New  Brian Hobbs  John Ziniti      NaN   \n",
       "..                      ...          ...          ...          ...      ...   \n",
       "259                    None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "342                    None          New  John Ziniti  John Ziniti      NaN   \n",
       "297                    None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "73                     None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "85                     None  In Progress  John Ziniti  John Ziniti      NaN   \n",
       "\n",
       "                                                   url  omic  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...     ?  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...     ?  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...     ?  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...     ?  \n",
       "57                                                         ?  \n",
       "..                                                 ...   ...  \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...     ?  \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...     ?  \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...     ?  \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...     ?  \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...     ?  \n",
       "\n",
       "[371 rows x 36 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def set_url(row):\n",
    "    cfs = row['custom_fields']\n",
    "    for cf in cfs:\n",
    "        cf_name = cf['name']\n",
    "        if cf_name == 'URL':\n",
    "            return cf['value']\n",
    "    return None\n",
    "\n",
    "dataset_tracker_id = trackers.loc[trackers['name'] == 'Dataset', 'id'].iloc[0]\n",
    "datasets = redmine.issue.filter(tracker_id=dataset_tracker_id)\n",
    "datasets = pd.DataFrame([dict(d) for d in datasets])\n",
    "\n",
    "datasets['project_name'] = datasets.apply(lambda w: w.project['name'], axis=1)\n",
    "datasets['category_name'] = datasets.apply(lambda w: w.category['name'] if pd.notna(w.category) else None, axis=1)\n",
    "datasets['status_name'] = datasets.apply(lambda w: w.status['name'] if pd.notna(w.status) else None, axis=1)\n",
    "datasets['author_name'] = datasets.apply(lambda w: w.author['name'] if pd.notna(w.author) else None, axis=1)\n",
    "datasets['assignee'] = datasets.apply(lambda w: w.assigned_to['name'] if pd.notna(w.assigned_to) else None, axis=1)\n",
    "datasets['parent_id'] = datasets.apply(lambda w: int(w.parent['id']) if pd.notna(w.parent) else None, axis=1)\n",
    "datasets['url'] = datasets.apply(set_url, axis=1)\n",
    "\n",
    "# datasets = datasets.drop(columns=['time_entries', 'changesets', 'watchers', 'allowed_statuses', 'due_date', 'done_ratio', 'tracker', 'priority', 'start_date', 'updated_on',\n",
    "#                                 'project', 'category', 'status', 'assigned_to', 'author', 'attachments', 'journals', 'is_private',\n",
    "##                                 'estimated_hours', 'created_on', 'closed_on', 'parent', 'description', 'custom_fields', 'relations',\n",
    "#                                 'children', 'author_name'])\n",
    "datasets = datasets.sort_values(['project_name', 'category_name', 'subject'])\n",
    "datasets = datasets.rename(columns={'project_name':'project','category_name':'category','status_name':'status','parent_id':'parent'})\n",
    "\n",
    "# datasets = datasets[datasets['project'] == REDMINE_STUDY]\n",
    "\n",
    "# datasets = datasets.drop(columns=['project', 'status'])\n",
    "\n",
    "datasets['omic'] = '?'\n",
    "\n",
    "### These will need to be fixed (made unique)\n",
    "## datasets[datasets['subject'] == 'TopMed WGS Data']\n",
    "datasets['subject'] = datasets.apply(lambda row: f\"(#{row['id']}) {row['subject']}\", axis=1)\n",
    "\n",
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "3a592d12-b0ea-4e38-9edd-e88f3d3c8d5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82                      (#22862) A1A_2013_HumanCoreExome\n",
       "18                    (#24096) A1A_HumanCoreExome_TOPMed\n",
       "99                                (#22080) A1A_850K_2021\n",
       "15                         (#24110) A1A olink Proteomics\n",
       "57         (#23928) Olink Proteomic Data Freeze 20221220\n",
       "                             ...                        \n",
       "259                         (#18545) VivaThousandQC_BMIQ\n",
       "342                         (#16227) eMERGE GWAS Datsets\n",
       "297    (#17846) emERGE Principal Components Projected...\n",
       "73                    (#23145) Graz_cohort_clinical_data\n",
       "85                 (#22764) graz_copd: BSA_0503_LBI_COPD\n",
       "Name: subject, Length: 371, dtype: object"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets.subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "17b3c67c-4e9a-4a82-abf7-6105fd1bbdb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['allowed_statuses',\n",
       " 'assigned_to',\n",
       " 'assignee',\n",
       " 'attachments',\n",
       " 'author',\n",
       " 'author_name',\n",
       " 'category',\n",
       " 'category',\n",
       " 'changesets',\n",
       " 'children',\n",
       " 'closed_on',\n",
       " 'created_on',\n",
       " 'custom_fields',\n",
       " 'description',\n",
       " 'done_ratio',\n",
       " 'due_date',\n",
       " 'estimated_hours',\n",
       " 'id',\n",
       " 'is_private',\n",
       " 'journals',\n",
       " 'omic',\n",
       " 'parent',\n",
       " 'parent',\n",
       " 'priority',\n",
       " 'project',\n",
       " 'project',\n",
       " 'relations',\n",
       " 'start_date',\n",
       " 'status',\n",
       " 'status',\n",
       " 'subject',\n",
       " 'time_entries',\n",
       " 'tracker',\n",
       " 'updated_on',\n",
       " 'url',\n",
       " 'watchers']"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(datasets.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "b411ffa7-9a59-45c6-b1ba-ba40a44f6504",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relations</th>\n",
       "      <th>time_entries</th>\n",
       "      <th>children</th>\n",
       "      <th>attachments</th>\n",
       "      <th>changesets</th>\n",
       "      <th>journals</th>\n",
       "      <th>watchers</th>\n",
       "      <th>allowed_statuses</th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>...</th>\n",
       "      <th>parent</th>\n",
       "      <th>project</th>\n",
       "      <th>category</th>\n",
       "      <th>status</th>\n",
       "      <th>author_name</th>\n",
       "      <th>assignee</th>\n",
       "      <th>parent</th>\n",
       "      <th>url</th>\n",
       "      <th>omic</th>\n",
       "      <th>projects.code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1A</td>\n",
       "      <td>dna/exome</td>\n",
       "      <td>In Progress</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>John Ziniti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>?</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   relations time_entries children attachments changesets journals watchers  \\\n",
       "82      None         None     None        None       None     None     None   \n",
       "\n",
       "   allowed_statuses     id                    project  ... parent project  \\\n",
       "82             None  22862  {'id': 28, 'name': 'A1A'}  ...    NaN     A1A   \n",
       "\n",
       "     category       status  author_name     assignee parent  \\\n",
       "82  dna/exome  In Progress  John Ziniti  John Ziniti    NaN   \n",
       "\n",
       "                                                  url omic  projects.code  \n",
       "82  /proj/regeps/regep00/studies/A1A/data/dna/whol...    ?             p0  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a copy of the DataFrame with the renamed columns and selected columns\n",
    "datasets = datasets.rename(columns={'subject': 'submitter_id'})\n",
    "\n",
    "# change data set infor to 'po'\n",
    "# datasets['project_id'] = 'p0'\n",
    "datasets['projects.code'] = 'p0'\n",
    "\n",
    "datasets.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "d37a1aec-b146-4953-a8af-8565824fbfd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['allowed_statuses',\n",
       " 'assigned_to',\n",
       " 'assignee',\n",
       " 'attachments',\n",
       " 'author',\n",
       " 'author_name',\n",
       " 'category',\n",
       " 'category',\n",
       " 'changesets',\n",
       " 'children',\n",
       " 'closed_on',\n",
       " 'created_on',\n",
       " 'custom_fields',\n",
       " 'description',\n",
       " 'done_ratio',\n",
       " 'due_date',\n",
       " 'estimated_hours',\n",
       " 'id',\n",
       " 'is_private',\n",
       " 'journals',\n",
       " 'omic',\n",
       " 'parent',\n",
       " 'parent',\n",
       " 'priority',\n",
       " 'project',\n",
       " 'project',\n",
       " 'projects.code',\n",
       " 'relations',\n",
       " 'start_date',\n",
       " 'status',\n",
       " 'status',\n",
       " 'submitter_id',\n",
       " 'time_entries',\n",
       " 'tracker',\n",
       " 'updated_on',\n",
       " 'url',\n",
       " 'watchers']"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(datasets.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "3f253308-2ffa-4048-987d-124f03c4436f",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = datasets.drop(columns =['allowed_statuses',\n",
    " 'assigned_to',\n",
    " 'assignee',\n",
    " 'attachments',\n",
    " 'author',\n",
    " 'author_name',\n",
    " 'category',\n",
    " 'category',\n",
    " 'changesets',\n",
    " 'children',\n",
    " 'closed_on',\n",
    " 'created_on',\n",
    " 'custom_fields',\n",
    " 'description',\n",
    " 'done_ratio',\n",
    " 'due_date',\n",
    " 'estimated_hours',\n",
    " 'is_private',\n",
    " 'journals',\n",
    " 'omic',\n",
    " 'parent',\n",
    " 'parent',\n",
    " 'priority',\n",
    " 'relations',\n",
    " 'start_date',\n",
    " 'status',\n",
    " 'status',\n",
    " 'time_entries',\n",
    " 'tracker',\n",
    " 'updated_on',\n",
    " 'watchers'])\n",
    "   \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "c3402f16-280c-48b7-af9a-c2b2f28ef960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24096) A1A_HumanCoreExome_TOPMed</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22080) A1A_850K_2021</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24110) A1A olink Proteomics</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#23928) Olink Proteomic Data Freeze 20221220</td>\n",
       "      <td>A1A</td>\n",
       "      <td></td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#18545) VivaThousandQC_BMIQ</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#16227) eMERGE GWAS Datsets</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#17846) emERGE Principal Components Projected...</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>23145</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#23145) Graz_cohort_clinical_data</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>22764</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#22764) graz_copd: BSA_0503_LBI_COPD</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "      <td>p0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                         project  \\\n",
       "82   22862       {'id': 28, 'name': 'A1A'}   \n",
       "18   24096       {'id': 28, 'name': 'A1A'}   \n",
       "99   22080       {'id': 28, 'name': 'A1A'}   \n",
       "15   24110       {'id': 28, 'name': 'A1A'}   \n",
       "57   23928       {'id': 28, 'name': 'A1A'}   \n",
       "..     ...                             ...   \n",
       "259  18545     {'id': 235, 'name': 'VIVA'}   \n",
       "342  16227   {'id': 952, 'name': 'eMERGE'}   \n",
       "297  17846   {'id': 952, 'name': 'eMERGE'}   \n",
       "73   23145  {'id': 1135, 'name': 'rebays'}   \n",
       "85   22764  {'id': 1135, 'name': 'rebays'}   \n",
       "\n",
       "                                          submitter_id project  \\\n",
       "82                    (#22862) A1A_2013_HumanCoreExome     A1A   \n",
       "18                  (#24096) A1A_HumanCoreExome_TOPMed     A1A   \n",
       "99                              (#22080) A1A_850K_2021     A1A   \n",
       "15                       (#24110) A1A olink Proteomics     A1A   \n",
       "57       (#23928) Olink Proteomic Data Freeze 20221220     A1A   \n",
       "..                                                 ...     ...   \n",
       "259                       (#18545) VivaThousandQC_BMIQ    VIVA   \n",
       "342                       (#16227) eMERGE GWAS Datsets  eMERGE   \n",
       "297  (#17846) emERGE Principal Components Projected...  eMERGE   \n",
       "73                  (#23145) Graz_cohort_clinical_data  rebays   \n",
       "85               (#22764) graz_copd: BSA_0503_LBI_COPD  rebays   \n",
       "\n",
       "                                                   url projects.code  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...            p0  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...            p0  \n",
       "57                                                                p0  \n",
       "..                                                 ...           ...  \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...            p0  \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0  \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0  \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...            p0  \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...            p0  \n",
       "\n",
       "[371 rows x 6 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "d1b80fbc-eb67-453d-829f-1fe8f111d123",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "      <th>associated_experiment</th>\n",
       "      <th>copy_numbers_identified</th>\n",
       "      <th>data_description</th>\n",
       "      <th>experimental_description</th>\n",
       "      <th>experimental_intent</th>\n",
       "      <th>indels_identified</th>\n",
       "      <th>marker_panel_description</th>\n",
       "      <th>number_experimental_group</th>\n",
       "      <th>number_samples_per_experimental_group</th>\n",
       "      <th>somatic_mutations_identified</th>\n",
       "      <th>type_of_data</th>\n",
       "      <th>type_of_sample</th>\n",
       "      <th>type_of_specimen</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                    project                      submitter_id  \\\n",
       "82  22862  {'id': 28, 'name': 'A1A'}  (#22862) A1A_2013_HumanCoreExome   \n",
       "\n",
       "   project                                                url projects.code  \\\n",
       "82     A1A  /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "\n",
       "   associated_experiment copy_numbers_identified data_description  \\\n",
       "82                  None                    None             None   \n",
       "\n",
       "   experimental_description experimental_intent indels_identified  \\\n",
       "82                     None                None              None   \n",
       "\n",
       "   marker_panel_description number_experimental_group  \\\n",
       "82                     None                      None   \n",
       "\n",
       "   number_samples_per_experimental_group somatic_mutations_identified  \\\n",
       "82                                  None                         None   \n",
       "\n",
       "   type_of_data type_of_sample type_of_specimen        type  \n",
       "82         None           None             None  experiment  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the column names to be added\n",
    "new_columns = ['associated_experiment', 'copy_numbers_identified', 'data_description', 'experimental_description',\n",
    "               'experimental_intent', 'indels_identified', 'marker_panel_description', 'number_experimental_group',\n",
    "               'number_samples_per_experimental_group', 'number_samples_per_experimental_group',\n",
    "               'somatic_mutations_identified', 'type_of_data', 'type_of_sample', 'type_of_specimen']\n",
    "\n",
    "# Add the new columns with 'N/A' as the default value\n",
    "for column in new_columns:\n",
    "    datasets[column] =  None\n",
    "    \n",
    "# Add the new column named type    \n",
    "datasets['type'] = 'experiment'\n",
    "datasets.rename(columns={'project_id':'projects.code'})\n",
    "\n",
    "# Export the CSV file\n",
    "tsv_file_path = 'experiments.tsv'\n",
    "\n",
    "# Export the DataFrame as a TSV file\n",
    "datasets.to_csv(tsv_file_path, sep='\\t', index=False)\n",
    "datasets.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "4dbc2ef3-b70a-4073-9a7c-cf3fc4f87d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(f'{len(datasets)=}')\n",
    "#datasets.value_counts('category_name', dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "079bc12b-4e3a-46df-aca1-e394b4158c9c",
   "metadata": {},
   "source": [
    "### Annotated Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "943032ea-6cee-40eb-8f5d-0568f9b9091f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "      <th>associated_experiment</th>\n",
       "      <th>copy_numbers_identified</th>\n",
       "      <th>data_description</th>\n",
       "      <th>experimental_description</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples_per_experimental_group</th>\n",
       "      <th>somatic_mutations_identified</th>\n",
       "      <th>type_of_data</th>\n",
       "      <th>type_of_sample</th>\n",
       "      <th>type_of_specimen</th>\n",
       "      <th>type</th>\n",
       "      <th>redmine_url</th>\n",
       "      <th>S</th>\n",
       "      <th>N</th>\n",
       "      <th>sample-manifest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22862</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24096) A1A_HumanCoreExome_TOPMed</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/24096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22080) A1A_850K_2021</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22080</td>\n",
       "      <td>302</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24110) A1A olink Proteomics</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/24110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#23928) Olink Proteomic Data Freeze 20221220</td>\n",
       "      <td>A1A</td>\n",
       "      <td></td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/23928</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>.chammps/sample-manifest.csv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#18545) VivaThousandQC_BMIQ</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/18545</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/VIVA/data/methyla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#16227) eMERGE GWAS Datsets</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/16227</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/eMERGE/data/gwas/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#17846) emERGE Principal Components Projected...</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/17846</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/eMERGE/data/pcs/....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>23145</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#23145) Graz_cohort_clinical_data</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/23145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>22764</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#22764) graz_copd: BSA_0503_LBI_COPD</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22764</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                         project  \\\n",
       "82   22862       {'id': 28, 'name': 'A1A'}   \n",
       "18   24096       {'id': 28, 'name': 'A1A'}   \n",
       "99   22080       {'id': 28, 'name': 'A1A'}   \n",
       "15   24110       {'id': 28, 'name': 'A1A'}   \n",
       "57   23928       {'id': 28, 'name': 'A1A'}   \n",
       "..     ...                             ...   \n",
       "259  18545     {'id': 235, 'name': 'VIVA'}   \n",
       "342  16227   {'id': 952, 'name': 'eMERGE'}   \n",
       "297  17846   {'id': 952, 'name': 'eMERGE'}   \n",
       "73   23145  {'id': 1135, 'name': 'rebays'}   \n",
       "85   22764  {'id': 1135, 'name': 'rebays'}   \n",
       "\n",
       "                                          submitter_id project  \\\n",
       "82                    (#22862) A1A_2013_HumanCoreExome     A1A   \n",
       "18                  (#24096) A1A_HumanCoreExome_TOPMed     A1A   \n",
       "99                              (#22080) A1A_850K_2021     A1A   \n",
       "15                       (#24110) A1A olink Proteomics     A1A   \n",
       "57       (#23928) Olink Proteomic Data Freeze 20221220     A1A   \n",
       "..                                                 ...     ...   \n",
       "259                       (#18545) VivaThousandQC_BMIQ    VIVA   \n",
       "342                       (#16227) eMERGE GWAS Datsets  eMERGE   \n",
       "297  (#17846) emERGE Principal Components Projected...  eMERGE   \n",
       "73                  (#23145) Graz_cohort_clinical_data  rebays   \n",
       "85               (#22764) graz_copd: BSA_0503_LBI_COPD  rebays   \n",
       "\n",
       "                                                   url projects.code  \\\n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...            p0   \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...            p0   \n",
       "57                                                                p0   \n",
       "..                                                 ...           ...   \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...            p0   \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0   \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0   \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...            p0   \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...            p0   \n",
       "\n",
       "    associated_experiment copy_numbers_identified data_description  \\\n",
       "82                   None                    None             None   \n",
       "18                   None                    None             None   \n",
       "99                   None                    None             None   \n",
       "15                   None                    None             None   \n",
       "57                   None                    None             None   \n",
       "..                    ...                     ...              ...   \n",
       "259                  None                    None             None   \n",
       "342                  None                    None             None   \n",
       "297                  None                    None             None   \n",
       "73                   None                    None             None   \n",
       "85                   None                    None             None   \n",
       "\n",
       "    experimental_description  ... number_samples_per_experimental_group  \\\n",
       "82                      None  ...                                  None   \n",
       "18                      None  ...                                  None   \n",
       "99                      None  ...                                  None   \n",
       "15                      None  ...                                  None   \n",
       "57                      None  ...                                  None   \n",
       "..                       ...  ...                                   ...   \n",
       "259                     None  ...                                  None   \n",
       "342                     None  ...                                  None   \n",
       "297                     None  ...                                  None   \n",
       "73                      None  ...                                  None   \n",
       "85                      None  ...                                  None   \n",
       "\n",
       "    somatic_mutations_identified type_of_data type_of_sample type_of_specimen  \\\n",
       "82                          None         None           None             None   \n",
       "18                          None         None           None             None   \n",
       "99                          None         None           None             None   \n",
       "15                          None         None           None             None   \n",
       "57                          None         None           None             None   \n",
       "..                           ...          ...            ...              ...   \n",
       "259                         None         None           None             None   \n",
       "342                         None         None           None             None   \n",
       "297                         None         None           None             None   \n",
       "73                          None         None           None             None   \n",
       "85                          None         None           None             None   \n",
       "\n",
       "           type                                    redmine_url    S  N  \\\n",
       "82   experiment  https://chanmine.bwh.harvard.edu/issues/22862    0  0   \n",
       "18   experiment  https://chanmine.bwh.harvard.edu/issues/24096    0  0   \n",
       "99   experiment  https://chanmine.bwh.harvard.edu/issues/22080  302  0   \n",
       "15   experiment  https://chanmine.bwh.harvard.edu/issues/24110    0  0   \n",
       "57   experiment  https://chanmine.bwh.harvard.edu/issues/23928    0  0   \n",
       "..          ...                                            ...  ... ..   \n",
       "259  experiment  https://chanmine.bwh.harvard.edu/issues/18545    0  0   \n",
       "342  experiment  https://chanmine.bwh.harvard.edu/issues/16227    0  0   \n",
       "297  experiment  https://chanmine.bwh.harvard.edu/issues/17846    0  0   \n",
       "73   experiment  https://chanmine.bwh.harvard.edu/issues/23145    0  0   \n",
       "85   experiment  https://chanmine.bwh.harvard.edu/issues/22764    0  0   \n",
       "\n",
       "                                       sample-manifest  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...  \n",
       "57                        .chammps/sample-manifest.csv  \n",
       "..                                                 ...  \n",
       "259  /proj/regeps/regep00/studies/VIVA/data/methyla...  \n",
       "342  /proj/regeps/regep00/studies/eMERGE/data/gwas/...  \n",
       "297  /proj/regeps/regep00/studies/eMERGE/data/pcs/....  \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...  \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...  \n",
       "\n",
       "[371 rows x 24 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def annotate(ds):\n",
    "    dsid = ds['id']\n",
    "   # dscat = ds[\"category\"]\n",
    "    dsurl = ds[\"url\"]\n",
    "    if pd.notna(dsurl):\n",
    "        dsurl = dsurl.replace('file://', '')\n",
    "    dsname = ds[\"submitter_id\"]\n",
    "    \n",
    "    ds['redmine_url'] = f'https://chanmine.bwh.harvard.edu/issues/{dsid}'\n",
    "\n",
    "    S = 0\n",
    "    try:\n",
    "        smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "        # print(f'{smanifest=}')\n",
    "        exists = smanifest.exists()\n",
    "    except:\n",
    "        exists = False\n",
    "        smanifest = None\n",
    "    if exists:\n",
    "        ds_samples = pd.read_csv(smanifest)\n",
    "        S = len(ds_samples)\n",
    "    else:\n",
    "        #print(f'NEED-MANIFEST: {smanifest=}')\n",
    "        pass\n",
    "    ds['S'] = S\n",
    "\n",
    "    N = 0\n",
    "    try:\n",
    "        sbmanifest = Path(dsurl)/'.chammps/subject-manifest.csv'\n",
    "        exists = sbmanifest.exists()\n",
    "    except:\n",
    "        exists = False\n",
    "    if exists:\n",
    "        ds_subjects = pd.read_csv(sbmanifest)\n",
    "        N = len(ds_subjects)\n",
    "    ds['N'] = N\n",
    "    ds['sample-manifest'] = smanifest\n",
    "    return(ds)\n",
    "\n",
    "datasets = datasets.apply(annotate, axis=1)\n",
    "#datasets = datasets.sort_values(['category',])\n",
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "ff73c6f4-4f5e-4486-bfff-4b06e35baa7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>project</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>project</th>\n",
       "      <th>url</th>\n",
       "      <th>projects.code</th>\n",
       "      <th>associated_experiment</th>\n",
       "      <th>copy_numbers_identified</th>\n",
       "      <th>data_description</th>\n",
       "      <th>experimental_description</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples_per_experimental_group</th>\n",
       "      <th>somatic_mutations_identified</th>\n",
       "      <th>type_of_data</th>\n",
       "      <th>type_of_sample</th>\n",
       "      <th>type_of_specimen</th>\n",
       "      <th>type</th>\n",
       "      <th>redmine_url</th>\n",
       "      <th>S</th>\n",
       "      <th>N</th>\n",
       "      <th>sample-manifest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>22862</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22862) A1A_2013_HumanCoreExome</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22862</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24096</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24096) A1A_HumanCoreExome_TOPMed</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/24096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/dna/whol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>22080</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#22080) A1A_850K_2021</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22080</td>\n",
       "      <td>302</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/epigenet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24110</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#24110) A1A olink Proteomics</td>\n",
       "      <td>A1A</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/24110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/A1A/data/protein/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>23928</td>\n",
       "      <td>{'id': 28, 'name': 'A1A'}</td>\n",
       "      <td>(#23928) Olink Proteomic Data Freeze 20221220</td>\n",
       "      <td>A1A</td>\n",
       "      <td></td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/23928</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>.chammps/sample-manifest.csv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>18545</td>\n",
       "      <td>{'id': 235, 'name': 'VIVA'}</td>\n",
       "      <td>(#18545) VivaThousandQC_BMIQ</td>\n",
       "      <td>VIVA</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/VIVA/data/...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/18545</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/VIVA/data/methyla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>16227</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#16227) eMERGE GWAS Datsets</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/16227</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/eMERGE/data/gwas/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>17846</td>\n",
       "      <td>{'id': 952, 'name': 'eMERGE'}</td>\n",
       "      <td>(#17846) emERGE Principal Components Projected...</td>\n",
       "      <td>eMERGE</td>\n",
       "      <td>file:///proj/regeps/regep00/studies/eMERGE/dat...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/17846</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/regeps/regep00/studies/eMERGE/data/pcs/....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>23145</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#23145) Graz_cohort_clinical_data</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/23145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/rebays/rebay00/data/pheno/COPD Graz_Coho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>22764</td>\n",
       "      <td>{'id': 1135, 'name': 'rebays'}</td>\n",
       "      <td>(#22764) graz_copd: BSA_0503_LBI_COPD</td>\n",
       "      <td>rebays</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "      <td>p0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>experiment</td>\n",
       "      <td>https://chanmine.bwh.harvard.edu/issues/22764</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>/proj/rebays/rebay00/data/dna/whole_genome/gra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>371 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                         project  \\\n",
       "82   22862       {'id': 28, 'name': 'A1A'}   \n",
       "18   24096       {'id': 28, 'name': 'A1A'}   \n",
       "99   22080       {'id': 28, 'name': 'A1A'}   \n",
       "15   24110       {'id': 28, 'name': 'A1A'}   \n",
       "57   23928       {'id': 28, 'name': 'A1A'}   \n",
       "..     ...                             ...   \n",
       "259  18545     {'id': 235, 'name': 'VIVA'}   \n",
       "342  16227   {'id': 952, 'name': 'eMERGE'}   \n",
       "297  17846   {'id': 952, 'name': 'eMERGE'}   \n",
       "73   23145  {'id': 1135, 'name': 'rebays'}   \n",
       "85   22764  {'id': 1135, 'name': 'rebays'}   \n",
       "\n",
       "                                          submitter_id project  \\\n",
       "82                    (#22862) A1A_2013_HumanCoreExome     A1A   \n",
       "18                  (#24096) A1A_HumanCoreExome_TOPMed     A1A   \n",
       "99                              (#22080) A1A_850K_2021     A1A   \n",
       "15                       (#24110) A1A olink Proteomics     A1A   \n",
       "57       (#23928) Olink Proteomic Data Freeze 20221220     A1A   \n",
       "..                                                 ...     ...   \n",
       "259                       (#18545) VivaThousandQC_BMIQ    VIVA   \n",
       "342                       (#16227) eMERGE GWAS Datsets  eMERGE   \n",
       "297  (#17846) emERGE Principal Components Projected...  eMERGE   \n",
       "73                  (#23145) Graz_cohort_clinical_data  rebays   \n",
       "85               (#22764) graz_copd: BSA_0503_LBI_COPD  rebays   \n",
       "\n",
       "                                                   url projects.code  \\\n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...            p0   \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...            p0   \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...            p0   \n",
       "57                                                                p0   \n",
       "..                                                 ...           ...   \n",
       "259  file:///proj/regeps/regep00/studies/VIVA/data/...            p0   \n",
       "342  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0   \n",
       "297  file:///proj/regeps/regep00/studies/eMERGE/dat...            p0   \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...            p0   \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...            p0   \n",
       "\n",
       "    associated_experiment copy_numbers_identified data_description  \\\n",
       "82                   None                    None             None   \n",
       "18                   None                    None             None   \n",
       "99                   None                    None             None   \n",
       "15                   None                    None             None   \n",
       "57                   None                    None             None   \n",
       "..                    ...                     ...              ...   \n",
       "259                  None                    None             None   \n",
       "342                  None                    None             None   \n",
       "297                  None                    None             None   \n",
       "73                   None                    None             None   \n",
       "85                   None                    None             None   \n",
       "\n",
       "    experimental_description  ... number_samples_per_experimental_group  \\\n",
       "82                      None  ...                                  None   \n",
       "18                      None  ...                                  None   \n",
       "99                      None  ...                                  None   \n",
       "15                      None  ...                                  None   \n",
       "57                      None  ...                                  None   \n",
       "..                       ...  ...                                   ...   \n",
       "259                     None  ...                                  None   \n",
       "342                     None  ...                                  None   \n",
       "297                     None  ...                                  None   \n",
       "73                      None  ...                                  None   \n",
       "85                      None  ...                                  None   \n",
       "\n",
       "    somatic_mutations_identified type_of_data type_of_sample type_of_specimen  \\\n",
       "82                          None         None           None             None   \n",
       "18                          None         None           None             None   \n",
       "99                          None         None           None             None   \n",
       "15                          None         None           None             None   \n",
       "57                          None         None           None             None   \n",
       "..                           ...          ...            ...              ...   \n",
       "259                         None         None           None             None   \n",
       "342                         None         None           None             None   \n",
       "297                         None         None           None             None   \n",
       "73                          None         None           None             None   \n",
       "85                          None         None           None             None   \n",
       "\n",
       "           type                                    redmine_url    S  N  \\\n",
       "82   experiment  https://chanmine.bwh.harvard.edu/issues/22862    0  0   \n",
       "18   experiment  https://chanmine.bwh.harvard.edu/issues/24096    0  0   \n",
       "99   experiment  https://chanmine.bwh.harvard.edu/issues/22080  302  0   \n",
       "15   experiment  https://chanmine.bwh.harvard.edu/issues/24110    0  0   \n",
       "57   experiment  https://chanmine.bwh.harvard.edu/issues/23928    0  0   \n",
       "..          ...                                            ...  ... ..   \n",
       "259  experiment  https://chanmine.bwh.harvard.edu/issues/18545    0  0   \n",
       "342  experiment  https://chanmine.bwh.harvard.edu/issues/16227    0  0   \n",
       "297  experiment  https://chanmine.bwh.harvard.edu/issues/17846    0  0   \n",
       "73   experiment  https://chanmine.bwh.harvard.edu/issues/23145    0  0   \n",
       "85   experiment  https://chanmine.bwh.harvard.edu/issues/22764    0  0   \n",
       "\n",
       "                                       sample-manifest  \n",
       "82   /proj/regeps/regep00/studies/A1A/data/dna/whol...  \n",
       "18   /proj/regeps/regep00/studies/A1A/data/dna/whol...  \n",
       "99   /proj/regeps/regep00/studies/A1A/data/epigenet...  \n",
       "15   /proj/regeps/regep00/studies/A1A/data/protein/...  \n",
       "57                        .chammps/sample-manifest.csv  \n",
       "..                                                 ...  \n",
       "259  /proj/regeps/regep00/studies/VIVA/data/methyla...  \n",
       "342  /proj/regeps/regep00/studies/eMERGE/data/gwas/...  \n",
       "297  /proj/regeps/regep00/studies/eMERGE/data/pcs/....  \n",
       "73   /proj/rebays/rebay00/data/pheno/COPD Graz_Coho...  \n",
       "85   /proj/rebays/rebay00/data/dna/whole_genome/gra...  \n",
       "\n",
       "[371 rows x 24 columns]"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "2db042a8-d8f2-497d-a0f2-86d102ce774c",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = datasets.drop(columns=['project', 'project', 'id', 'url', 'N', 'S', 'sample-manifest', 'redmine_url'])\n",
    "#datasets = datasets.rename(columns={'project_id':'projects.code'})\n",
    "datasets.to_csv('output.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "0343a881-a9c9-4e3e-b3fb-0cbddf6437af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>projects.code</th>\n",
       "      <th>associated_experiment</th>\n",
       "      <th>copy_numbers_identified</th>\n",
       "      <th>data_description</th>\n",
       "      <th>experimental_description</th>\n",
       "      <th>experimental_intent</th>\n",
       "      <th>indels_identified</th>\n",
       "      <th>marker_panel_description</th>\n",
       "      <th>number_experimental_group</th>\n",
       "      <th>number_samples_per_experimental_group</th>\n",
       "      <th>somatic_mutations_identified</th>\n",
       "      <th>type_of_data</th>\n",
       "      <th>type_of_sample</th>\n",
       "      <th>type_of_specimen</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [submitter_id, projects.code, associated_experiment, copy_numbers_identified, data_description, experimental_description, experimental_intent, indels_identified, marker_panel_description, number_experimental_group, number_samples_per_experimental_group, somatic_mutations_identified, type_of_data, type_of_sample, type_of_specimen, type]\n",
       "Index: []"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets[datasets['submitter_id'] == 'TopMed WGS Data']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6851d6bd-dc7f-43f4-85bf-73e1f86192b2",
   "metadata": {},
   "source": [
    "## Dataset Focus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d93009ec-1c47-4f90-8cca-fd63cb959402",
   "metadata": {},
   "source": [
    "### Storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "f630c086-377a-4d91-ac14-78334fe92491",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/.conda/envs/cdnm-jupyter-python-3.7.6/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3360\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3361\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/cdnm-jupyter-python-3.7.6/lib/python3.9/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/cdnm-jupyter-python-3.7.6/lib/python3.9/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/d/tmp/rejpz/ipykernel_28756/1205867693.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mDATASET_ID\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m20642\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdatasets\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdatasets\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mDATASET_ID\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mdsurl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'url'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'file://'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/cdnm-jupyter-python-3.7.6/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3456\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3457\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3458\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3459\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/cdnm-jupyter-python-3.7.6/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3361\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3363\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3365\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'id'"
     ]
    }
   ],
   "source": [
    "## Copy a Dataset id from the table above, and then execute this cell and those below to see information about it\n",
    "DATASET_ID = 20642\n",
    "\n",
    "ds = datasets.loc[datasets['id']==DATASET_ID]\n",
    "dsurl = ds.iloc[0]['url'].replace('file://', '')\n",
    "\n",
    "storage_manifests = sorted(glob.glob(str(Path(dsurl)/'results/rip/*/*-file-table-*.csv')))\n",
    "\n",
    "if storage_manifests:\n",
    "    print(f'{storage_manifests[-1]=}')\n",
    "    storage = pd.read_csv(storage_manifests[-1])\n",
    "else:\n",
    "    storage = pd.DataFrame()\n",
    "    print(f'NO-RIP-FILE-TABLES-PLEASE-RIP: {dsurl=}, going forward {storage=}')\n",
    "storage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4e8159-efb2-46c1-930f-c3d0b9f23c49",
   "metadata": {},
   "source": [
    "### Sample History"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5d7ee6-3e6d-4352-965e-089a77841f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'{SampleMap=}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b8f1c07-101a-4add-b2c7-7416a68ee74e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "def query_sample_history(row):\n",
    "    if 'S_SAMPLEID' in row:\n",
    "        destsampleid = sid = row['S_SAMPLEID']\n",
    "    else:\n",
    "        #print(f'{row=}')\n",
    "        aliasid = row['TOR_ID']\n",
    "        q_alias = session.query(Alias).filter(Alias.c.aliasid==aliasid)\n",
    "        aliases = pd.read_sql(q_alias.statement, session.bind)\n",
    "        destsampleid = sid = aliases.iloc[0]['keyid1']\n",
    "    sinfo = samples.loc[samples['s_sampleid'] == sid]\n",
    "    sample_history = [sid,]\n",
    "    sample_history_timestamps = [sinfo.iloc[0]['createdt']]\n",
    "    sample_history_types = [sinfo.iloc[0]['sampletypeid']]\n",
    "    while pd.notna(destsampleid):\n",
    "        q_sm = session.query(SampleMap).filter(SampleMap.c.destsampleid==destsampleid)\n",
    "        sm = pd.read_sql(q_sm.statement, session.bind)\n",
    "        # print(f'{sm=}')\n",
    "        if sm.empty:\n",
    "            destsampleid = None\n",
    "            break\n",
    "        else:\n",
    "            #if len(sm) > 1:\n",
    "            #    print(f'OUCH! Pooled? {sm=}')\n",
    "            sinfo = samples.loc[samples['s_sampleid'] == sm.iloc[0]['sourcesampleid']]\n",
    "            sample_history.append(sinfo.iloc[0]['s_sampleid'])\n",
    "            sample_history_types.append(sinfo.iloc[0]['sampletypeid'])\n",
    "            sample_history_timestamps.append(sinfo.iloc[0]['createdt'])\n",
    "            destsampleid = sm.iloc[0]['sourcesampleid']\n",
    "\n",
    "    row['sample_history_length'] = len(sample_history)\n",
    "    row['sample_history'] = tuple(sample_history)\n",
    "    row['sample_history_types'] = tuple(sample_history_types)\n",
    "    row['sample_history_timestamps'] = tuple(sample_history_timestamps)\n",
    "    return row\n",
    "\"\"\"\n",
    "\n",
    "def query_sample_history(sid):\n",
    "    data = { \"sampleid\": sid}\n",
    "\n",
    "    statement = text(\"\"\"select SOURCESAMPLEID, DESTSAMPLEID, level lvl from sapphire8.s_samplemap start with DESTSAMPLEID=:sampleid connect by DESTSAMPLEID= prior SOURCESAMPLEID\"\"\")\n",
    "    sample_history = pd.read_sql(statement, session.bind, params=data)\n",
    "    return sample_history\n",
    "\n",
    "shistory = query_sample_history('S-001357960')\n",
    "shistory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d8c100a-1642-4e4e-afbe-d03ab971a686",
   "metadata": {},
   "source": [
    "### Example of plotting time interval data in Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e271bcc3-3663-4a13-95fd-7980073f3ef4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/50869690/how-to-plot-time-interval-data-in-python\n",
    "\n",
    "#df = pd.DataFrame({'Eng':['eng1','eng1','eng1','eng2','eng3','eng4','eng4','eng4','eng5','eng5'],\n",
    "#'Install date':['3/6/2010','10/25/2010','1/1/2014','1/1/2011','1/1/2014','1/1/2015','1/5/2016','5/8/2017','2/3/2010','1/1/2011'],\n",
    "#'Car':['car1','car1','car2','car3','car1','car1','car3','car2','car1','car4']})\n",
    "#df['Install date'] = pd.to_datetime(df['Install date'])\n",
    "#for car in df.Car.unique():\n",
    "#    most_recent_eng = df.loc[df.Car == car].groupby('Install date').max().tail(1).Eng.iloc[0]\n",
    "#    print(f'{car=},{most_recent_eng=}')\n",
    "#    new_df = df.loc[df.Car == car].append(pd.DataFrame({'Car':[car],'Eng':[most_recent_eng],'Install date':[pd.to_datetime('now')]}), sort=False)\n",
    "#    print(f'{new_df=}')\n",
    "#    pt = new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count')\n",
    "#    print(f'{pt=}')\n",
    "#    rs = pt.resample('1d')\n",
    "#    print(f'{rs=}')\n",
    "#    ff = rs.ffill()\n",
    "#    print(f'{ff=}')\n",
    "#    new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count').resample('1d').ffill().plot(kind='line',linewidth=10, title=car)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f5b65b4-1496-4797-9887-745c4a26b530",
   "metadata": {},
   "source": [
    "## Specific Datasets, With Sample History"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5c4aa68-28e6-4124-b81d-0e893a742453",
   "metadata": {},
   "source": [
    "### CAMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1888882c-9e6c-4a71-aa76-44b7852fbdc2",
   "metadata": {},
   "source": [
    "#### CAMP TOPMed WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f282a3b0-590a-4305-8ca2-3e7c866ab3c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TopMed WGS Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "# dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d9fdd6-77a2-4924-8c44-10dd8184b674",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (s, sample) in dataset_samples.iterrows():\n",
    "    sample_history = query_sample_history(sid)\n",
    "    sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "    sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "    sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "    most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "    df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "    df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "    df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "    df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "    pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "    pt = pt.resample('1d')\n",
    "    ff = pt.ffill()\n",
    "    _ = ff.plot(kind='line', linewidth=5, title=sid, marker='', figsize=(5,3))\n",
    "    if s > 20: break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6231b780-eedd-4d9b-8db9-eb3af6db612b",
   "metadata": {},
   "source": [
    "### LTRC TOPMed WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60e9230e-7ab8-4c75-aad8-903bb594ebf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC TOPMed WGS'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0afedca9-37ec-432f-bd05-31d0b639c088",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset_samples.value_counts('sample_history_types')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de0cfb1-1de9-4941-abdf-a67499841d2f",
   "metadata": {},
   "source": [
    "### TOPMed Blood Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb00a2b-cb45-4953-a0c3-9b574cd8ffe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TOPMed Blood Methylation Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da07c5f5-a6e0-415e-b531-1628c1d94e9a",
   "metadata": {},
   "source": [
    "### TOPMed Lung Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d99e44-063c-42b5-a19c-7f04793a5726",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'TOPMed Lung Methylation Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1758f7b-e6f8-4b4e-89b9-640859596a59",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for (s, sample) in dataset_samples.iterrows():\n",
    "#    sample_df = datset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "#    print(f'{sample_df=}')\n",
    "#    sample_df sample_df['Job position'].apply(pd.Series).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')#\n",
    "\n",
    "#    most_recent_type = sample_df['sample_history_types'][-1]\n",
    "#    print(f'{sample=},{most_recent_type=}')\n",
    "#    new_df = df.loc[df.Car == car].append(pd.DataFrame({'Car':[car],'Eng':[most_recent_eng],'Install date':[pd.to_datetime('now')]}), sort=False)\n",
    "#    print(f'{new_df=}')\n",
    "#    pt = new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count')\n",
    "#    print(f'{pt=}')\n",
    "#    rs = pt.resample('1d')\n",
    "#    print(f'{rs=}')\n",
    "#    ff = rs.ffill()\n",
    "#    print(f'{ff=}')\n",
    "#    new_df.pivot_table(values='Car',columns='Eng',index='Install date', aggfunc='count').resample('1d').ffill().plot(kind='line',linewidth=10, title=car)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1e5b42-0c37-4a4e-aa3d-a5ec272b64e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 0\n",
    "sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "\n",
    "rs = pt.resample('1d')\n",
    "\n",
    "ff = rs.ffill()\n",
    "ff\n",
    "_ = ff.plot(kind='line', linewidth=8, title=sid, marker='', figsize=(5,3))\n",
    "#_ = plt.plot(ff, linewidth=2) #, title=sid, marker='')\n",
    "#_ = plt.plot(ff, linewidth=2) #, title=sid, marker='')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f695dcd6-9a69-4b93-943d-2260bab8e1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (s, sample) in dataset_samples.iterrows():\n",
    "    sid = dataset_samples.iloc[s]['S_SAMPLEID']\n",
    "    sample_df = dataset_samples.iloc[s][['sample_history', 'sample_history_types', 'sample_history_timestamps']]\n",
    "\n",
    "    sids = pd.Series(sample_df['sample_history']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    types = pd.Series(sample_df['sample_history_types']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "    timestamps = pd.Series(sample_df['sample_history_timestamps']).reset_index().melt(id_vars='index').dropna()[['index', 'value']].set_index('index')\n",
    "\n",
    "    most_recent_type = types.iloc[0]['value']\n",
    "\n",
    "    df = sids.merge(types, left_index=True, right_index=True).merge(timestamps, left_index=True, right_index=True)\n",
    "    df = df.append(pd.DataFrame({'value_x':[sid], 'value_y': [most_recent_type], 'value':[pd.to_datetime('now')]}))\n",
    "    df = df.rename(columns={'value_x':'sampleid','value_y':'sampletype','value':'timestamp'})\n",
    "    df = df.sort_values('timestamp', ascending=True).reset_index().drop(columns=['index',])\n",
    "\n",
    "    pt = df.pivot_table(values='sampleid', columns='sampletype', index='timestamp', aggfunc='count')\n",
    "\n",
    "    rs = pt.resample('1d')\n",
    "\n",
    "    ff = rs.ffill()\n",
    "    _ = ff.plot(kind='line', linewidth=8, title=sid, marker='', figsize=(5,3))\n",
    "    if s > 20: break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56b0b424-ed04-4bdb-8a8d-426f17c674df",
   "metadata": {},
   "source": [
    "### LTRC Phenotype Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f2f4b7-cb3e-48d7-87bb-d25eaffde0b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC Phenotype Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e79e441-e2bc-4438-b552-811faaef9c16",
   "metadata": {},
   "source": [
    "### LTRC RNASeq Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742dad9e-8aa9-478c-8fab-2dc9d6291e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_NAME = 'LTRC RNASeq Data'\n",
    "ds = datasets.loc[datasets['subject']==DATASET_NAME]\n",
    "dsname = ds.iloc[0][\"subject\"]\n",
    "dsurl = ds.iloc[0][\"url\"]\n",
    "dsurl = dsurl.replace('file://', '')\n",
    "    \n",
    "smanifest = Path(dsurl)/'.chammps/sample-manifest.csv'\n",
    "dataset_samples = pd.read_csv(smanifest)\n",
    "dataset_samples = dataset_samples.apply(query_sample_history, axis=1)\n",
    "dataset_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ec022e3-7cdc-4968-85d8-6d2cf16499f8",
   "metadata": {},
   "source": [
    "## Upset Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808c9d96-e7f2-48b7-986b-71aeed7d257c",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASETS_FOR_UPSET_PLOT = ['LTRC TOPMed WGS', 'TOPMed Blood Methylation Data', 'TOPMed Lung Methylation Data', 'LTRC Phenotype Data', 'LTRC RNASeq Data']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a550e46-a190-4203-a5c2-b967820f9a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_for_upset = datasets[datasets['subject'].isin(DATASETS_FOR_UPSET_PLOT)]\n",
    "datasets_for_upset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257ce980-772f-496f-8ea9-130df1d2e4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = []\n",
    "for (d, dataset) in datasets_for_upset.iterrows():\n",
    "    smanifest = dataset['sample-manifest']\n",
    "    df = pd.read_csv(smanifest)\n",
    "    dataframes.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb80bafe-7243-489c-9e78-ae4d867fc258",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = dataframes[0]\n",
    "for df in dataframes[1:]:\n",
    "    merged = merged.merge(df, on='patid', how='left')\n",
    "\n",
    "#merged = phenotype.merge(rna, on='patid', how='left')\n",
    "#merged = merged.merge(lung_methylation, on='patid', how='left')\n",
    "#merged = merged.merge(blood_methylation, on='patid', how='left')\n",
    "#merged = merged.merge(wgs, on='patid', how='left')\n",
    "#merged = merged.merge(smoking, on='patid', how='left')\n",
    "merged = merged.fillna(False)\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71e034a4-0fe3-4769-a63b-aab3c8346fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "indicators = merged[['rna','lung_methylation', 'blood_methylation', 'wgs']]\n",
    "participants_by_omics = upsetplot.from_indicators(indicators, data=merged)\n",
    "#upsetplot.UpSet(participants_by_omics, orientation='vertical', show_counts=True)\n",
    "upset = upsetplot.UpSet(participants_by_omics, show_counts=True, sort_by='cardinality', sort_categories_by=None, intersection_plot_elements=0)\n",
    "upset.add_stacked_bars(by=\"smoking_status\", title=\"smoking\", elements=5)\n",
    "upset.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf56052-eb17-444c-bd12-3c9d08f4c7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "indicators = merged[['rna', 'lung_methylation', 'wgs']]\n",
    "participants_by_omics = upsetplot.from_indicators(indicators, data=merged)\n",
    "#upsetplot.UpSet(participants_by_omics, orientation='vertical', show_counts=True)\n",
    "upset = upsetplot.UpSet(participants_by_omics, show_counts=True, sort_by='cardinality', sort_categories_by=None, intersection_plot_elements=0)\n",
    "#upset.add_catplot(value='smoking_status', kind='bar', color='blue')\n",
    "upset.add_stacked_bars(by=\"smoking_status\", title=\"smoking\", elements=5)\n",
    "upset.plot()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cdnm-jupyter-python-3.7.6",
   "language": "python",
   "name": "cdnm-jupyter-python-3.7.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
